
### Tier 1 - Variants of strong clinical significance {.tabset}

- Considering evidence items with strongest evidence levels (A & B) for the query tumor type (_`r pcg_report[['metadata']][['config']][["t_props"]][['tumor_type']]`_)  in the [database for clinical interpretations of variants in cancer, CIViC](https://civic.genome.wustl.edu/#/home) or [Cancer Biomarkers database](https://www.cancergenomeinterpreter.org/biomarkers), a total of __`r NROW(pcg_report[['content']][['snv_indel']][["disp"]][['tier1']])`__ unique, somatic variants were found, with the following number of evidence items:
    * Tier 1 - Predictive/Therapeutic: __`r NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['query_ttype']][['predictive']][['A_B']])`__ evidence items
    * Tier 1 - Prognostic: __`r NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['query_ttype']][['prognostic']][['A_B']])`__ evidence items
    * Tier 1 - Diagnostic: __`r NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['query_ttype']][['diagnostic']][['A_B']])`__ evidence items

<br><br>


```{r biomarker_note1, echo=F, results = "asis", include = pcg_report$content$snv_indel$v_stat$n_tier1 > 0}

cat('<b>NOTE:</b> Reported biomarkers in CIViC/CGI are mapped at different resolutions (i.e. filter <b>Biomarker mapping</b>). The accuracy of a match between variants in the tumor sample and the reported biomarkers will vary accordingly (highlighted by gene symbols with different color backgrounds):\n\n')

cat('<ul><li>Biomarker match at the <mark style="background-color:black; font-weight:bold; color:white">exact variant/codon level</mark></li>')
cat(paste0('<br><li>Biomarker match at the <mark style="background-color:', pcg_report[['metadata']][['color_palette']][['warning']][['values']][1],'; color:white; font-weight:bold">exon/gene level</mark></li></ul>\n'))

htmltools::br()
```



```{r prepare_tier1_data, echo=F, results='asis'}

show_tier1_filters <- list()
missing_tier1_items <- list()
for(type in c('diagnostic','prognostic','predictive')){
  show_tier1_filters[[type]] <- F
  missing_tier1_items[[type]] <- T
  if(NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['query_ttype']][[type]][['A_B']]) > 0){
    show_tier1_filters[[type]] <- T
    missing_tier1_items[[type]] <- F
  }
}

```


```{r active_tab_snv_predictive, echo = F, results = "asis"}
if(missing_tier1_items[['predictive']] == F | 
   (missing_tier1_items[['diagnostic']] == T & 
    missing_tier1_items[['prognostic']] == T)){
  cat("")
  cat("#### Predictive biomarkers {.active}")
  cat("")
}else{
  cat("")
  cat("#### Predictive biomarkers")
  cat("")
}
```

<br>

```{r tier1_predictive, echo=F, results = 'asis', eval = show_tier1_filters[['predictive']]}

variants_tier1_predictive_shared <- crosstalk::SharedData$new(pcg_report[['content']][['snv_indel']][['clin_eitem']][['query_ttype']][['predictive']][['A_B']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_tier1_predictive_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical significance", variants_tier1_predictive_shared, ~CLINICAL_SIGNIFICANCE),
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_tier1_predictive_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_slider("RATING", "Rating", variants_tier1_predictive_shared, ~RATING, min = 0, max = 5, step = 1, ticks = T)

  ),
  list(
    crosstalk::filter_select("CONSEQUENCE", "Consequence", variants_tier1_predictive_shared, ~CONSEQUENCE),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_tier1_predictive_shared, ~BIOMARKER_MAPPING),
    crosstalk::filter_select("THERAPEUTIC_CONTEXT", "Therapeutic context", variants_tier1_predictive_shared, ~THERAPEUTIC_CONTEXT)
  )
)

cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_tier1_predictive_shared,  escape=F,extensions=c("Buttons","Responsive"), 
              options=list(buttons = c('csv','excel'),  scrollCollapse=T, dom = 'Bfrtip')) |>
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated','A: FDA/NCCN/ELN guidelines',
                                                     'B: Clinical evidence','B1: Clinical evidence: late trials',
                                                     'B2: Clinical evidence: early trials','C: Case study',
                                                     'D: Preclinical evidence','E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", "#56B4E9","#56B4E9","#0072B2","#E69F00", "#F0E442"))) |>
  DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',pcg_report[['metadata']][['color_palette']][['warning']][['values']][1], pcg_report[['metadata']][['color_palette']][['warning']][['values']][1])))



```

```{r tier1_predictive_missing, echo=F, results = 'asis', eval = missing_tier1_items[['predictive']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br>


```{r active_tab_snv_prognostic, echo = F, results = "asis"}
if(missing_tier1_items[['prognostic']] == F & 
   missing_tier1_items[['predictive']] == T){
  cat("")
  cat("#### Prognostic biomarkers {.active}")
  cat("")
}else{
  cat("")
  cat("#### Prognostic biomarkers")
  cat("")
}
```

<br>

```{r tier1_prognostic, echo=F, results='asis', eval = show_tier1_filters[['prognostic']]}

variants_tier1_prognostic_shared <- crosstalk::SharedData$new(pcg_report[['content']][['snv_indel']][['clin_eitem']][['query_ttype']][['prognostic']][['A_B']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("CONSEQUENCE", "Consequence", variants_tier1_prognostic_shared, ~CONSEQUENCE),
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_tier1_prognostic_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical significance", variants_tier1_prognostic_shared, ~CLINICAL_SIGNIFICANCE)
  ),
  list(
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_tier1_prognostic_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_tier1_prognostic_shared, ~BIOMARKER_MAPPING),
    crosstalk::filter_slider("RATING", "Rating", variants_tier1_prognostic_shared, ~RATING, min = 0, max = 5, step = 1, ticks = T)

  )
)
cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_tier1_prognostic_shared, escape=F,extensions=c("Buttons","Responsive"), 
              options=list(buttons = c('csv','excel'), scrollCollapse=T, dom = 'Bfrtip')) |>
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated','A: FDA/NCCN/ELN guidelines',
                                                     'B: Clinical evidence','B1: Clinical evidence: late trials',
                                                     'B2: Clinical evidence: early trials','C: Case study',
                                                     'D: Preclinical evidence','E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", "#56B4E9","#56B4E9","#0072B2","#E69F00", "#F0E442"))) |>
   DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',pcg_report[['metadata']][['color_palette']][['warning']][['values']][1], pcg_report[['metadata']][['color_palette']][['warning']][['values']][1])))



```

```{r tier1_prognostic_missing, echo=F, results = 'asis', eval = missing_tier1_items[['prognostic']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br>

```{r active_tab_snv_diagnostic, echo = F, results = "asis"}
if(missing_tier1_items[['diagnostic']] == F & 
   missing_tier1_items[['predictive']] == T & 
   missing_tier1_items[['prognostic']] == T){
  cat("")
  cat("#### Diagnostic biomarkers {.active}")
  cat("")
}else{
  cat("")
  cat("#### Diagnostic biomarkers")
  cat("")
}
```

<br>

```{r tier1_diagnostic, echo=F, results='asis', eval = show_tier1_filters[['diagnostic']]}

variants_tier1_diagnostic_shared <- crosstalk::SharedData$new(pcg_report[['content']][['snv_indel']][['clin_eitem']][['query_ttype']][['diagnostic']][['A_B']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("CONSEQUENCE", "Consequence", variants_tier1_diagnostic_shared, ~CONSEQUENCE),
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_tier1_diagnostic_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical association", variants_tier1_diagnostic_shared, ~CLINICAL_SIGNIFICANCE)

  ),
  list(
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_tier1_diagnostic_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_tier1_diagnostic_shared, ~BIOMARKER_MAPPING),
    crosstalk::filter_slider("RATING", "Rating", variants_tier1_diagnostic_shared, ~RATING, min = 0, max = 5, step = 1, ticks = T)

  )
)
cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_tier1_diagnostic_shared, escape=F, extensions=c("Buttons","Responsive"), 
              options=list(buttons = c('csv','excel'), scrollCollapse=T, dom = 'Bfrtip')) |>
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated','A: FDA/NCCN/ELN guidelines',
                                                     'B: Clinical evidence','B1: Clinical evidence: late trials',
                                                     'B2: Clinical evidence: early trials','C: Case study',
                                                     'D: Preclinical evidence','E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", "#56B4E9","#56B4E9","#0072B2","#E69F00", "#F0E442"))) |>
   DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',pcg_report[['metadata']][['color_palette']][['warning']][['values']][1], pcg_report[['metadata']][['color_palette']][['warning']][['values']][1])))



```

```{r tier1_diagnostic_missing, echo=F, results = 'asis', eval = missing_tier1_items[['diagnostic']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br>


