---
title: "Cancer Predisposition Sequencing Report - `r cps_report$metadata$sample_name`"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: bibliography.bib
nocite: |
  @Huang2018-ah, @Richards2015-kj, @Maxwell2016-mr, @Amendola2016-pu
output: 
  html_document:
    toc: true
    theme: default
    number_sections: false
    toc_float: true
    toc_depth: 3
    fig_width: 5
    fig_height: 4
    includes:
      after_body: disclaimer_predisposition.md
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo = F,warning=FALSE, dpi=72, error=F, eval=T)
options(scipen=999)

width_panel <- 14
height_panel <- 3 + as.integer((max(0,(NROW(cps_report[['metadata']][['gene_panel']][['genes']]) - 24)) / 16))

width_donut <- 10
height_donut <- 7

set_clinvar <- NROW(cps_report$content$snv_indel$variant_set$tsv %>% dplyr::filter(!is.na(CLINVAR_CLASSIFICATION)))
set_other <- NROW(cps_report$content$snv_indel$variant_set$tsv %>% dplyr::filter(nchar(CPSR_CLASSIFICATION) > 0))

if(set_other == 0 | set_clinvar == 0){
  width_donut <- 6
  height_donut <- 4
}

```

```{r settings_data_sources, child='predispose/settings_data_sources_predispose.Rmd'}
```


```{r gene_selection, echo = F, fig.width = width_panel, fig.height = height_panel, eval = T}
p1 <- pcgrr::gene_selection_tiles(genes = cps_report[['metadata']][['gene_panel']][['genes']]$symbol, confidence = cps_report[['metadata']][['gene_panel']][['genes']]$confidence_level, box_w = 3.4, space = 0.4, box_h = 0.8)
p1

```
<br>

## Summary of findings

```{r donut_chart, echo = F, fig.width = width_donut, fig.height = height_donut, eval = cps_report[['content']][['snv_indel']][['eval']]}
p1 <- pcgrr::summary_donut_chart(cps_report$content$snv_indel$variant_set$tsv, plot_type = "ClinVar")
p2 <- pcgrr::summary_donut_chart(cps_report$content$snv_indel$variant_set$tsv, plot_type = "Other")

p <- NULL
if(!is.null(p1) & !is.null(p2)){
  p <- ggpubr::ggarrange(p1, p2, labels =c("",""), ncol = 2, nrow = 1)
}
if(!is.null(p1) & is.null(p2)){
  p  = p1
}
if(is.null(p1) & !is.null(p2)){
  p <- p2
}
p

```

## Germline SNVs/InDels

### Variant statistics 

Variant numbers in the selected cancer predisposition genes (n = `r NROW(cps_report[['metadata']][['gene_panel']][['genes']])`)

* Number of SNVs: __`r cps_report[['content']][['snv_indel']][['variant_statistic_cpg']][['n_snv']]`__
* Number of InDels: __`r cps_report[['content']][['snv_indel']][['variant_statistic_cpg']][['n_indel']]`__
* Number of protein-coding variants: __`r cps_report[['content']][['snv_indel']][['variant_statistic_cpg']][['n_coding']]`__

<br><br>

```{r conditional_class_1_5, child='predispose/predisposition_class1_5.Rmd', eval = cps_report[['content']][['snv_indel']][['eval']]}
```

<br><br>

```{r conditional_biomarkers, child='predispose/biomarkers_germline.Rmd', eval = cps_report[['content']][['snv_indel']][['eval']]}
```

<br><br>


```{r conditional_sf, child='predispose/sf.Rmd', eval = cps_report[['metadata']][['config']][['secondary_findings']][['show_sf']]}
```

```{r global_options2, include=FALSE}
options(scipen=0)
```


```{r conditional_gwas, child='predispose/gwas_predispose.Rmd', eval = cps_report[['metadata']][['config']][['gwas']][['gwas_hits']]}
```

```{r conditional_documentation, child='predispose/documentation_predispose.Rmd', eval = TRUE}
```

### References

