<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Personal Cancer Genome ReporteR â€¢ PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Personal Cancer Genome ReporteR">
<meta name="description" content="Functions, tools and utilities for the generation of clinical cancer genome reports with PCGR. This R package is an integrated part of the PCGR workflow (https://github.com/sigven/pcgr), it should thus not be used as a stand-alone package.">
<meta property="og:description" content="Functions, tools and utilities for the generation of clinical cancer genome reports with PCGR. This R package is an integrated part of the PCGR workflow (https://github.com/sigven/pcgr), it should thus not be used as a stand-alone package.">
<meta property="og:image" content="https://sigven.github.io/pcgr/logo.png">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-7HRLE8GQRE"></script><script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){dataLayer.push(arguments);}
   gtag('js', new Date());

   gtag('config', 'G-7HRLE8GQRE');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="articles/running.html">Running</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/input.html">Input files</a></li>
    <li><a class="dropdown-item" href="articles/output.html">Output files</a></li>
    <li><a class="dropdown-item" href="articles/variant_classification.html">Variant classification</a></li>
    <li><a class="dropdown-item" href="articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="articles/developers.html">Developer notes</a></li>
    <li><a class="dropdown-item" href="articles/primary_tumor_sites.html">Primary tumor sites</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="personal-cancer-genome-reporter-pcgr-">Personal Cancer Genome Reporter (PCGR) 
<a class="anchor" aria-label="anchor" href="#personal-cancer-genome-reporter-pcgr-"></a>
</h1>
</div>
<p>The Personal Cancer Genome Reporter (PCGR) is a stand-alone software package for functional annotation and translation of individual tumor genomes for precision cancer medicine. It interprets primarily somatic SNVs/InDels and copy number aberrations, and has additional support for interpretation of bulk RNA-seq expression data. The software classifies variants both with respect to <em>oncogenicity</em>, and <em>actionability</em>. Interactive HTML output reports allow the user to interrogate the clinical impact of the molecular findings in an individual tumor.</p>
<ul>
<li>Variant classification
<ul>
<li>according to <em>oncogenicity</em>: evaluating the oncogenic potential of somatic DNA aberrations (<a href="https://pubmed.ncbi.nlm.nih.gov/35101336/" class="external-link">ClinGen/CGC/VICC guidelines</a>)</li>
<li>according to <em>actionability</em>: mapping the therapeutic, diagnostic, and prognostic implications of somatic DNA aberrations (<a href="https://pubmed.ncbi.nlm.nih.gov/27993330/" class="external-link">AMP/ASCO/CAP guidelines</a>)</li>
</ul>
</li>
<li>Tumor mutational burden (TMB) estimation</li>
<li>Mutational signature analysis</li>
<li>Microsatellite instability (MSI) classification</li>
<li>RNA-seq support - gene expression outlier detection, sample similarity analysis, and immune contexture profiling</li>
</ul>
<p>PCGR supports both of the most recent human genome assemblies (GRCh37/GRCh38), and accepts variant calls from both tumor-control and tumor-only sequencing assays. Much of the functionality is intended for whole-exome/whole-genome sequencing assays, but you can also apply PCGR to output from targeted sequencing panels. If you are interested in the interrogation of germline variants and their relation to cancer predisposition, we recommend trying the accompanying tool <a href="https://github.com/sigven/cpsr" class="external-link">Cancer Predisposition Sequencing Reporter (CPSR)</a>.</p>
<p>Example screenshots from the <a href="https://quarto.org" class="external-link">quarto</a>-based cancer genome report by PCGR:</p>
<p><img src="img/sc2.png" alt="PCGR screenshot 1"><img src="img/sc1.png" alt="PCGR screenshot 2"><img src="img/sc3.png" alt="PCGR screenshot 3"></p>
<p>PCGR originates from the <a href="https://cancergenomics.no" class="external-link">Norwegian Cancer Genomics Consortium (NCGC)</a>, at the <a href="https://radium.no" class="external-link">Institute for Cancer Research, Oslo University Hospital, Norway</a>.</p>
<div class="section level3">
<h3 id="top-news">Top News<a class="anchor" aria-label="anchor" href="#top-news"></a>
</h3>
<ul>
<li>
<em>October 11th 2024:</em> <strong>2.1.1 release</strong>
<ul>
<li>patch to fix <a href="https://github.com/sigven/pcgr/issues/252" class="external-link">bug</a> with parsing of relative CDS start positions</li>
<li><a href="https://sigven.github.io/pcgr/articles/CHANGELOG.html">CHANGELOG</a></li>
</ul>
</li>
<li>
<em>September 29th 2024</em>: <strong>2.1.0 release</strong>
<ul>
<li>updated bundle, more oncogenic variants, CNA visualization, improved RNA-seq support, bug fixes, and more</li>
<li><a href="https://sigven.github.io/pcgr/articles/CHANGELOG.html">CHANGELOG</a></li>
</ul>
</li>
<li>
<em>August 1st 2024</em>: <strong>2.0.3 release</strong>
<ul>
<li>patch to fix purity/ploidy propagation, MAF output for tumor-only runs, and other minor issues</li>
<li><a href="https://sigven.github.io/pcgr/articles/CHANGELOG.html">CHANGELOG</a></li>
</ul>
</li>
<li>
<em>July 16th 2024</em>: <strong>2.0.2 release</strong>
<ul>
<li>patch to ensure correct reference to actionability guidelines</li>
<li><a href="https://sigven.github.io/pcgr/articles/CHANGELOG.html">CHANGELOG</a></li>
</ul>
</li>
<li>
<em>July 7th 2024</em>: <strong>2.0.1 release</strong>
<ul>
<li>patch with bug fix for mitochondrial input variants (<a href="https://github.com/sigven/pcgr/pull/245" class="external-link">pr245</a>)</li>
<li><a href="https://sigven.github.io/pcgr/articles/CHANGELOG.html">CHANGELOG</a></li>
</ul>
</li>
<li>
<em>June 2024</em>: <strong>2.0.0 release</strong>
<ul>
<li>Details in <a href="https://sigven.github.io/pcgr/articles/CHANGELOG.html">CHANGELOG</a>
</li>
<li>Massive reference data bundle upgrade, new report layout, oncogenicity classification++</li>
<li>Support for Singularity/Apptainer</li>
<li>Major data/software updates:
<ul>
<li>Ensembl VEP <code>v112</code>
</li>
<li>ClinVar (2024-06)</li>
<li>CIViC (2024-06-21)</li>
<li>GENCODE <code>v46/v19</code> (GRCh38/GRCh37)</li>
<li>CancerMine <code>v50</code> (2023-03)</li>
<li>UniProt KB <code>v2024_03</code>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="example-reports">Example reports<a class="anchor" aria-label="anchor" href="#example-reports"></a>
</h2>
<p><a href="https://doi.org/10.5281/zenodo.13855988" class="external-link"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.13855988.svg" alt="DOI"></a></p>
</div>
<div class="section level2">
<h2 id="why-use-pcgr">Why use PCGR?<a class="anchor" aria-label="anchor" href="#why-use-pcgr"></a>
</h2>
<p>The great complexity of acquired mutations in individual tumor genomes poses a severe challenge for clinical interpretation. PCGR aims to be a comprehensive reporting platform that can</p>
<ul>
<li>systematically interrogate tumor-specific variants in the context of known therapeutic, diagnostic, and prognostic biomarkers</li>
<li>highlight genomic aberrations with likely oncogenic potential</li>
<li>provide a structured and concise summary of the most relevant findings</li>
<li>present the results in a format accessible to clinical experts</li>
</ul>
<p>PCGR integrates a <a href="articles/annotation_resources.html">comprehensive set of knowledge resources</a> related to tumor biology and therapeutic biomarkers, both at the gene, and at the level of individual variants. The software generates a comprehensive molecular interpretation report that supports the translation of individual cancer genomes towards molecularly guided treatment strategies.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<ul>
<li><p><a href="articles/installation.html">Installation instructions</a></p></li>
<li><p><a href="articles/running.html#example-run">Run through an example</a></p></li>
<li>
<p>Learn more about:</p>
<ol style="list-style-type: decimal">
<li>Details regarding <a href="articles/input.html">PCGR input files</a>, and how they should be formatted</li>
<li>Configuration of <a href="articles/running.html">key settings</a>
</li>
<li>The types and contents of <a href="articles/output.html">PCGR output files</a>
</li>
<li><a href="articles/variant_classification.html">Variant classifications implemented in PCGR</a></li>
<li><a href="articles/primary_tumor_sites.html">Primary tumor sites used in PCGR</a></li>
<li>The list of <a href="articles/annotation_resources.html">gene and variant annotation resources</a> used in PCGR annotation</li>
</ol>
</li>
<li><p><a href="articles/faq.html">Frequently asked questions (FAQ)</a></p></li>
</ul>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you use PCGR or CPSR, please cite our publications:</p>
<p>Sigve Nakken, Ghislain Fournous, Daniel VodÃ¡k, Lars Birger Aaasheim, Ola Myklebost, and Eivind Hovig. <strong>Personal Cancer Genome Reporter: variant interpretation report for precision oncology</strong> (2017). <em>Bioinformatics</em>. 34(10):1778â€“1780. <a href="https://doi.org/10.1093/bioinformatics/btx817" class="external-link">doi.org/10.1093/bioinformatics/btx817</a></p>
<p>Sigve Nakken, Vladislav Saveliev, Oliver Hofmann, PÃ¥l MÃ¸ller, Ola Myklebost, and Eivind Hovig. <strong>Cancer Predisposition Sequencing Reporter (CPSR): a flexible variant report engine for high-throughput germline screening in cancer</strong> (2021). <em>Int J Cancer</em>. <a href="https://doi.org/10.1002/ijc.33749" class="external-link">doi:[10.1002/ijc.33749](doi:%5B10.1002/ijc.33749)</a></p>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>sigven AT ifi.uio.no</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/sigven/pcgr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/sigven/pcgr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><p><a href="authors.html#citation">Citing PCGR</a></p></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/sigven" class="external-link">Sigve Nakken</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8468-2050" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a> <br><small class="roles"> Author, contributor </small> <a href="https://orcid.org/0000-0002-7502-7545" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
