<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • PCGR</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://sigven.github.io/pcgr/logo.png"><script async src="https://www.googletagmanager.com/gtag/js?id=G-7HRLE8GQRE"></script><script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){dataLayer.push(arguments);}
   gtag('js', new Date());

   gtag('config', 'G-7HRLE8GQRE');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/running.html">Running</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/input.html">Input files</a></li>
    <li><a class="dropdown-item" href="../articles/output.html">Output files</a></li>
    <li><a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a></li>
    <li><a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="../articles/developers.html">Developer notes</a></li>
    <li><a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="af_distribution.html">af_distribution()</a></code>

        </dt>
        <dd>Function that plots a histogram of the the variant allelic support (tumor)</dd>
      </dl><dl><dt>

          <code><a href="append_annotation_links.html">append_annotation_links()</a></code>

        </dt>
        <dd>Function that appends multiple HTML annotation links to variant identifiers e.g. COSMIC, CLINVAR, REFSEQ etc</dd>
      </dl><dl><dt>

          <code><a href="append_cancer_association_ranks.html">append_cancer_association_ranks()</a></code>

        </dt>
        <dd>Function that appends cancer gene evidence links</dd>
      </dl><dl><dt>

          <code><a href="append_cancer_gene_evidence.html">append_cancer_gene_evidence()</a></code>

        </dt>
        <dd>Function that appends cancer gene evidence links</dd>
      </dl><dl><dt>

          <code><a href="append_dbmts_var_link.html">append_dbmts_var_link()</a></code>

        </dt>
        <dd>Function that adds miRNA target annotations (dbMTS) to genetic variant identifiers</dd>
      </dl><dl><dt>

          <code><a href="append_dbnsfp_var_link.html">append_dbnsfp_var_link()</a></code>

        </dt>
        <dd>Function that assigns HTML links to dbNSFP prediction entries</dd>
      </dl><dl><dt>

          <code><a href="append_drug_var_link.html">append_drug_var_link()</a></code>

        </dt>
        <dd>Function that adds link to targeted drugs (on and off-label) for a list of variants and associated targeted</dd>
      </dl><dl><dt>

          <code><a href="append_gwas_citation_phenotype.html">append_gwas_citation_phenotype()</a></code>

        </dt>
        <dd>Function that adds GWAS citation/phenotype to GWAS hit found through PCGR annotation</dd>
      </dl><dl><dt>

          <code><a href="append_oncogenicity_docs.html">append_oncogenicity_docs()</a></code>

        </dt>
        <dd>Function that adds oncogenicity documentation from codes</dd>
      </dl><dl><dt>

          <code><a href="append_targeted_drug_annotations.html">append_targeted_drug_annotations()</a></code>

        </dt>
        <dd>Function that adds link to targeted drugs (on and off-label) for a list of variants and associated targeted</dd>
      </dl><dl><dt>

          <code><a href="append_tcga_var_link.html">append_tcga_var_link()</a></code>

        </dt>
        <dd>Function that adds TCGA annotations (cohort, frequency etc.) to variant identifiers</dd>
      </dl><dl><dt>

          <code><a href="append_tfbs_annotation.html">append_tfbs_annotation()</a></code>

        </dt>
        <dd>Function that adds TFBS annotations (dbMTS) to genetic variant identifiers</dd>
      </dl><dl><dt>

          <code><a href="assign_amp_asco_tiers.html">assign_amp_asco_tiers()</a></code>

        </dt>
        <dd>Function that assigns tier classifications to somatic CNA segments and SNVs/InDels, based on the presence of biomarker evidence found in the variant set</dd>
      </dl><dl><dt>

          <code><a href="assign_germline_popfreq_status.html">assign_germline_popfreq_status()</a></code>

        </dt>
        <dd>Function that sets STATUS_POPFREQ_1KG_ABOVE_TOLERATED/ STATUS_POPFREQ_GNOMAD_ABOVE_TOLERATED to TRUE for variants if any population frequency exceeds max_tolerated_af</dd>
      </dl><dl><dt>

          <code><a href="assign_mutation_type.html">assign_mutation_type()</a></code>

        </dt>
        <dd>Function that assigns one of six mutation types to a list of mutations</dd>
      </dl><dl><dt>

          <code><a href="assign_somatic_classification.html">assign_somatic_classification()</a></code>

        </dt>
        <dd>Function that assigns a SOMATIC_CLASSIFICATION to variants based on evidence found in variant set, potentially limited by user-defined options</dd>
      </dl><dl><dt>

          <code><a href="assign_somatic_germline_evidence.html">assign_somatic_germline_evidence()</a></code>

        </dt>
        <dd>Function that appends several tags denoting evidence for somatic/germline status of variants</dd>
      </dl><dl><dt>

          <code><a href="biomarker_evidence.html">biomarker_evidence</a></code>

        </dt>
        <dd>Fixed data types/categories used for biomarker evidence, e.g. 'types','levels' etc.</dd>
      </dl><dl><dt>

          <code><a href="cancer_phenotypes_regex.html">cancer_phenotypes_regex</a></code>

        </dt>
        <dd>Regular expression of terms indicative of cancer-related phenotypes and syndromes</dd>
      </dl><dl><dt>

          <code><a href="check_common_colnames.html">check_common_colnames()</a></code>

        </dt>
        <dd>Function that checks whether a set of column names are present in two different data frames</dd>
      </dl><dl><dt>

          <code><a href="check_file_exists.html">check_file_exists()</a></code>

        </dt>
        <dd>Function that checks the existence of a file</dd>
      </dl><dl><dt>

          <code><a href="clinvar_germline_status.html">clinvar_germline_status()</a></code>

        </dt>
        <dd>Function that assigns a logical to STATUS_CLINVAR_GERMLINE based on whether a ClinVar entry of germline origin is found for a given variant (for entries in a data frame)</dd>
      </dl><dl><dt>

          <code><a href="color_palette.html">color_palette</a></code>

        </dt>
        <dd>Color encodings for report elements of PCGR/CPSR</dd>
      </dl><dl><dt>

          <code><a href="cosmic_sbs_signatures.html">cosmic_sbs_signatures</a></code>

        </dt>
        <dd>List of COSMIC reference mutational signatures (SBS, v3.4)</dd>
      </dl><dl><dt>

          <code><a href="cosmic_somatic_status.html">cosmic_somatic_status()</a></code>

        </dt>
        <dd>Function that assigns a logical (STATUS_COSMIC) reflecting whether a variant co-incides with an entry in COSMIC (germline)</dd>
      </dl><dl><dt>

          <code><a href="data_coltype_defs.html">data_coltype_defs</a></code>

        </dt>
        <dd>List of coltype definitions for input files to pcgrr (e.g. VCF-converted TSV, CNA TVS etc.)</dd>
      </dl><dl><dt>

          <code><a href="dbsnp_germline_status.html">dbsnp_germline_status()</a></code>

        </dt>
        <dd>Function that assigns a logical (STATUS_DBSNP) reflecting whether a variant co-incides with an entry in dbSNP (germline)</dd>
      </dl><dl><dt>

          <code><a href="deduplicate_eitems.html">deduplicate_eitems()</a></code>

        </dt>
        <dd>Function that removes redundancy in variant evidence items (i.e. if a variant is assicated with evidence at the codon level, evidence at the exon/gene level is ignored)</dd>
      </dl><dl><dt>

          <code><a href="detect_vcf_sample_name.html">detect_vcf_sample_name()</a></code>

        </dt>
        <dd>A function that detects whether the sample name in variant data frame is unique (as present in column name VCF_SAMPLE_ID), throws an error if multiple sample names are present for the CPSR workflow</dd>
      </dl><dl><dt>

          <code><a href="df_string_replace.html">df_string_replace()</a></code>

        </dt>
        <dd>Function that performs stringr::str_replace on strings of multiple string columns of a dataframe</dd>
      </dl><dl><dt>

          <code><a href="dt_display.html">dt_display</a></code>

        </dt>
        <dd>DT Display</dd>
      </dl><dl><dt>

          <code><a href="effect_prediction_algos.html">effect_prediction_algos</a></code>

        </dt>
        <dd>List of URLs for a range of variant effect prediction algorithms</dd>
      </dl><dl><dt>

          <code><a href="exclude_non_chrom_variants.html">exclude_non_chrom_variants()</a></code>

        </dt>
        <dd>Function that excludes genomic aberrations from non-nuclear chromosomes</dd>
      </dl><dl><dt>

          <code><a href="expand_biomarker_items.html">expand_biomarker_items()</a></code>

        </dt>
        <dd>Function that expands biomarker evidence items with variant annotations</dd>
      </dl><dl><dt>

          <code><a href="export_quarto_evars.html">export_quarto_evars()</a></code>

        </dt>
        <dd>Export Quarto Environment Variables</dd>
      </dl><dl><dt>

          <code><a href="filter_eitems_by_site.html">filter_eitems_by_site()</a></code>

        </dt>
        <dd>Function that filters clinical evidence items by tumor type/primary site</dd>
      </dl><dl><dt>

          <code><a href="filter_maf_file.html">filter_maf_file()</a></code>

        </dt>
        <dd>Function that takes a MAF file generated with vcf2maf and filters out variants that are presumably germline (tumor-only run)</dd>
      </dl><dl><dt>

          <code><a href="filter_read_support.html">filter_read_support()</a></code>

        </dt>
        <dd>Function that filters variant set on (depth, allelic fraction) for tumor and normal and filters according to settings</dd>
      </dl><dl><dt>

          <code><a href="generate_annotation_link.html">generate_annotation_link()</a></code>

        </dt>
        <dd>A function that generates a HTML link for selected identifiers (DBSNP, COSMIC, CLINVAR, ENTREZ)</dd>
      </dl><dl><dt>

          <code><a href="generate_report.html">generate_report()</a></code>

        </dt>
        <dd>Function that generates all contents of the cancer genome report (PCGR)</dd>
      </dl><dl><dt>

          <code><a href="generate_report_data_expression.html">generate_report_data_expression()</a></code>

        </dt>
        <dd>Function that generates expression data for PCGR report</dd>
      </dl><dl><dt>

          <code><a href="generate_report_data_kataegis.html">generate_report_data_kataegis()</a></code>

        </dt>
        <dd>Function that generates data frame with potential kataegis events</dd>
      </dl><dl><dt>

          <code><a href="generate_report_data_msi.html">generate_report_data_msi()</a></code>

        </dt>
        <dd>Function that generates MSI prediction data for PCGR report</dd>
      </dl><dl><dt>

          <code><a href="generate_report_data_rainfall.html">generate_report_data_rainfall()</a></code>

        </dt>
        <dd>Function that generates data for rainfall plot (mutation density along genome, considering SNVs only)</dd>
      </dl><dl><dt>

          <code><a href="generate_report_data_signatures.html">generate_report_data_signatures()</a></code>

        </dt>
        <dd>Function that generates mutational signatures data for PCGR report</dd>
      </dl><dl><dt>

          <code><a href="generate_report_data_tmb.html">generate_report_data_tmb()</a></code>

        </dt>
        <dd>Function that reads TSV file with TMB estimates from sample</dd>
      </dl><dl><dt>

          <code><a href="generate_report_data_trials.html">generate_report_data_trials()</a></code>

        </dt>
        <dd>Function that retrieves relevant (interventional based on molecular target) clinical trials for a given tumor type</dd>
      </dl><dl><dt>

          <code><a href="generate_tier_tsv.html">generate_tier_tsv()</a></code>

        </dt>
        <dd>Function that generates dense and tiered annotated variant datasets</dd>
      </dl><dl><dt>

          <code><a href="germline_filter_levels.html">germline_filter_levels</a></code>

        </dt>
        <dd>Data frame with germline filtering criteria</dd>
      </dl><dl><dt>

          <code><a href="get_clin_assocs_cna.html">get_clin_assocs_cna()</a></code>

        </dt>
        <dd>Function that retrieves clinical evidence items (CIVIC, CBMDB) for CNA aberrations</dd>
      </dl><dl><dt>

          <code><a href="get_dt_tables.html">get_dt_tables()</a></code>

        </dt>
        <dd>Function that gathers data tables on actionable variants for display in report (tier 1 + tier 2)</dd>
      </dl><dl><dt>

          <code><a href="get_excel_sheets.html">get_excel_sheets()</a></code>

        </dt>
        <dd>Function that produces the contents of sheets for an Excel report of PCGR output</dd>
      </dl><dl><dt>

          <code><a href="get_genome_obj.html">get_genome_obj()</a></code>

        </dt>
        <dd>Get BSgenome Object</dd>
      </dl><dl><dt>

          <code><a href="get_oncogenic_cna_events.html">get_oncogenic_cna_events()</a></code>

        </dt>
        <dd>Get oncogenic copy number events</dd>
      </dl><dl><dt>

          <code><a href="get_prevalent_site_signatures.html">get_prevalent_site_signatures()</a></code>

        </dt>
        <dd>Function that retrieves prevalent signatures for a given tumor type/primary site Data is collected from COSMIC v3.4.</dd>
      </dl><dl><dt>

          <code><a href="get_valid_chromosomes.html">get_valid_chromosomes()</a></code>

        </dt>
        <dd>Checks for valid chromosome names in data frame of variants</dd>
      </dl><dl><dt>

          <code><a href="get_variant_statistics.html">get_variant_statistics()</a></code>

        </dt>
        <dd>Function that computes various variant statistics from a data frame with variant records</dd>
      </dl><dl><dt>

          <code><a href="het_af_germline_status.html">het_af_germline_status()</a></code>

        </dt>
        <dd>Function that assigns a logical (STATUS_LIKELY_GERMLINE_HETEROZYGOUS) reflecting whether a variant is likely heterozygous (germline) - based on allelic fraction (VAF_TUMOR), presence in gnomAD and dbSNP, and no presence in TCGA and COSMIC</dd>
      </dl><dl><dt>

          <code><a href="hom_af_status.html">hom_af_status()</a></code>

        </dt>
        <dd>Function that assigns a logical (STATUS_LIKELY_GERMLINE_HOMOZYGOUS) reflecting whether a variant is likely homozygous (germline) - based on allelic fraction (VAF_TUMOR)</dd>
      </dl><dl><dt>

          <code><a href="immune_celltypes.html">immune_celltypes</a></code>

        </dt>
        <dd>Data frame with immune cell types</dd>
      </dl><dl><dt>

          <code><a href="init_cna_vstats.html">init_cna_vstats()</a></code>

        </dt>
        <dd>Function that initiates report element with CNA information</dd>
      </dl><dl><dt>

          <code><a href="init_expression_content.html">init_expression_content()</a></code>

        </dt>
        <dd>Function that initiates report element with expression information</dd>
      </dl><dl><dt>

          <code><a href="init_germline_content.html">init_germline_content()</a></code>

        </dt>
        <dd>Function that initiates report element with germline variant information (CPSR)</dd>
      </dl><dl><dt>

          <code><a href="init_kataegis_content.html">init_kataegis_content()</a></code>

        </dt>
        <dd>Function that initiates report element with kataegis information</dd>
      </dl><dl><dt>

          <code><a href="init_m_signature_content.html">init_m_signature_content()</a></code>

        </dt>
        <dd>Function that initiates report element with mutational signatures information</dd>
      </dl><dl><dt>

          <code><a href="init_msi_content.html">init_msi_content()</a></code>

        </dt>
        <dd>Function that initiates report element with MSI classification</dd>
      </dl><dl><dt>

          <code><a href="init_rainfall_content.html">init_rainfall_content()</a></code>

        </dt>
        <dd>Function that initiates report element with rainfall information</dd>
      </dl><dl><dt>

          <code><a href="init_report.html">init_report()</a></code>

        </dt>
        <dd>Function that initiates PCGR/CPSR report object</dd>
      </dl><dl><dt>

          <code><a href="init_snv_indel_vstats.html">init_snv_indel_vstats()</a></code>

        </dt>
        <dd>Function that initiates report element with SNV/InDel statistics information</dd>
      </dl><dl><dt>

          <code><a href="init_tmb_content.html">init_tmb_content()</a></code>

        </dt>
        <dd>Function that initiates report element with TMB information</dd>
      </dl><dl><dt>

          <code><a href="init_tumor_only_content.html">init_tumor_only_content()</a></code>

        </dt>
        <dd>Function that initiates report element with tumor-only information</dd>
      </dl><dl><dt>

          <code><a href="init_var_content.html">init_var_content()</a></code>

        </dt>
        <dd>Function that initiates report element with variant data</dd>
      </dl><dl><dt>

          <code><a href="kataegis_detect.html">kataegis_detect()</a></code>

        </dt>
        <dd>Function that detects kataegis events from a data frame with genomic cooordinates of mutations</dd>
      </dl><dl><dt>

          <code><a href="kataegis_input.html">kataegis_input()</a></code>

        </dt>
        <dd>Function that detects kataegis events from a data frame with genomic cooordinates of mutations</dd>
      </dl><dl><dt>

          <code><a href="load_all_eitems.html">load_all_eitems()</a></code>

        </dt>
        <dd>Function that loads all evidence items from CIViC and CGI, and combines them in a unified data.frame</dd>
      </dl><dl><dt>

          <code><a href="load_cpsr_classified_variants.html">load_cpsr_classified_variants()</a></code>

        </dt>
        <dd>Function that reads CPSR-classified variants from a TSV file</dd>
      </dl><dl><dt>

          <code><a href="load_dna_variants.html">load_dna_variants()</a></code>

        </dt>
        <dd>Function that reads and validates CNA or SNV/InDel TSV files file from PCGR/CPSR pre-report (Python) pipeline</dd>
      </dl><dl><dt>

          <code><a href="load_eitems.html">load_eitems()</a></code>

        </dt>
        <dd>Function that loads specific set of clinical variant evidence items (CIViC + CGI) based on given parameters (mutation type, variant origin, tumor type etc)</dd>
      </dl><dl><dt>

          <code><a href="load_expression_csq.html">load_expression_csq()</a></code>

        </dt>
        <dd>Load expression consequence settings</dd>
      </dl><dl><dt>

          <code><a href="load_expression_outliers.html">load_expression_outliers()</a></code>

        </dt>
        <dd>Load expression outlier results</dd>
      </dl><dl><dt>

          <code><a href="load_expression_similarity.html">load_expression_similarity()</a></code>

        </dt>
        <dd>Load expression similarity results</dd>
      </dl><dl><dt>

          <code><a href="load_reference_data.html">load_reference_data()</a></code>

        </dt>
        <dd>Function that parses and loads reference data from files in the assembly-specific PCGR bundle directory</dd>
      </dl><dl><dt>

          <code><a href="load_somatic_cna.html">load_somatic_cna()</a></code>

        </dt>
        <dd>Function that reads and validates fully annotated CNA data (segments and genes) from PCGR pre-reporting pipeline</dd>
      </dl><dl><dt>

          <code><a href="load_somatic_snv_indel.html">load_somatic_snv_indel()</a></code>

        </dt>
        <dd>Function that reads and validates an annotated somatic SNV/InDel file from PCGR pre-reporting pipeline</dd>
      </dl><dl><dt>

          <code><a href="load_yaml.html">load_yaml()</a></code>

        </dt>
        <dd>Function that loads YAML data with settings and file paths to annotated molecular profiles</dd>
      </dl><dl><dt>

          <code><a href="log4r_debug.html">log4r_debug()</a></code>

        </dt>
        <dd>Write messages to logs at a given priority level</dd>
      </dl><dl><dt>

          <code><a href="log4r_fatal.html">log4r_fatal()</a></code>

        </dt>
        <dd>Write messages to logs at a given priority level</dd>
      </dl><dl><dt>

          <code><a href="log4r_info.html">log4r_info()</a></code>

        </dt>
        <dd>Write messages to logs at a given priority level</dd>
      </dl><dl><dt>

          <code><a href="log4r_warn.html">log4r_warn()</a></code>

        </dt>
        <dd>Write messages to logs at a given priority level</dd>
      </dl><dl><dt>

          <code><a href="log_var_eitem_stats.html">log_var_eitem_stats()</a></code>

        </dt>
        <dd>Function that logs the number of evidence items found, for different levels of resolution</dd>
      </dl><dl><dt>

          <code><a href="max_af_gnomad.html">max_af_gnomad()</a></code>

        </dt>
        <dd>Function that assigns a maximum value to a variable (MAX_AF_GNOMAD) reflecting the maximum allele frequency for a given variant across gnomAD populations</dd>
      </dl><dl><dt>

          <code><a href="mkdir.html">mkdir()</a></code>

        </dt>
        <dd>Create directory</dd>
      </dl><dl><dt>

          <code><a href="msi_indel_fraction_plot.html">msi_indel_fraction_plot()</a></code>

        </dt>
        <dd>Function that plots the indel fraction for a given sample and contrasts this with the distribution for MSI-H/MSS samples from TCGA</dd>
      </dl><dl><dt>

          <code><a href="msi_indel_load_plot.html">msi_indel_load_plot()</a></code>

        </dt>
        <dd>Function that plots the indel load for a given sample and contrasts this with the distribution for MSI-H/MSS samples from TCGA</dd>
      </dl><dl><dt>

          <code><a href="oncogenicity_criteria.html">oncogenicity_criteria</a></code>

        </dt>
        <dd>Oncogenicity criteria (ClinGen/CGC/VICC)</dd>
      </dl><dl><dt>

          <code><a href="order_variants.html">order_variants()</a></code>

        </dt>
        <dd>Function that orders genomic aberrations according to order of chromosomes and chromosomal position</dd>
      </dl><dl><dt>

          <code><a href="pkg_exists.html">pkg_exists()</a></code>

        </dt>
        <dd>Does R Package Exist</dd>
      </dl><dl><dt>

          <code><a href="plot_cna_segments.html">plot_cna_segments()</a></code>

        </dt>
        <dd>Plot allele-specific copy number segments</dd>
      </dl><dl><dt>

          <code><a href="plot_signature_contributions.html">plot_signature_contributions()</a></code>

        </dt>
        <dd>Function that makes plots of mutational signature contributions in a given sample (both ggplot and plotly)</dd>
      </dl><dl><dt>

          <code><a href="plot_tmb_primary_site_tcga.html">plot_tmb_primary_site_tcga()</a></code>

        </dt>
        <dd>Function that makes a plot with TMB boxplots for reference cohorts, highlighting the TMB estimate for a given sample and the cohort/primary site of interest</dd>
      </dl><dl><dt>

          <code><a href="plot_value_boxes.html">plot_value_boxes()</a></code>

        </dt>
        <dd>Function that plots four value boxes with the most important findings in the cancer genome</dd>
      </dl><dl><dt>

          <code><a href="pon_status.html">pon_status()</a></code>

        </dt>
        <dd>Function that assigns a logical (STATUS_PON) reflecting whether a variant is co-inciding with a variant present in a panel-of-normals database (PANEL_OF_NORMALS column is TRUE)</dd>
      </dl><dl><dt>

          <code><a href="predict_msi_status.html">predict_msi_status()</a></code>

        </dt>
        <dd>Function that predicts MSI status based on fraction of indels among calls</dd>
      </dl><dl><dt>

          <code><a href="qc_var_eitems.html">qc_var_eitems()</a></code>

        </dt>
        <dd>Function that matches variants to evidence items</dd>
      </dl><dl><dt>

          <code><a href="remove_cols_from_df.html">remove_cols_from_df()</a></code>

        </dt>
        <dd>Function that removes column(s) from data frame</dd>
      </dl><dl><dt>

          <code><a href="sort_chromosomal_segments.html">sort_chromosomal_segments()</a></code>

        </dt>
        <dd>Function that sorts chromosomal segments according to chromosome and chromosomal start/end position</dd>
      </dl><dl><dt>

          <code><a href="strip_html.html">strip_html()</a></code>

        </dt>
        <dd>Strip HTML tags</dd>
      </dl><dl><dt>

          <code><a href="structure_var_eitems.html">structure_var_eitems()</a></code>

        </dt>
        <dd>Function that structures variant evidence items according to strength of evidence</dd>
      </dl><dl><dt>

          <code><a href="tcga_cohorts.html">tcga_cohorts</a></code>

        </dt>
        <dd>Data frame with all TCGA cohorts</dd>
      </dl><dl><dt>

          <code><a href="tcga_somatic_status.html">tcga_somatic_status()</a></code>

        </dt>
        <dd>Function that assigns a logical (STATUS_TCGA_SOMATIC) reflecting whether a variant co-incides with an entry in TCGA (somatic)</dd>
      </dl><dl><dt>

          <code><a href="tier_af_distribution.html">tier_af_distribution()</a></code>

        </dt>
        <dd>Function that plots a histogram of the the variant allelic support (tumor) - grouped by tiers</dd>
      </dl><dl><dt>

          <code><a href="tsv_cols.html">tsv_cols</a></code>

        </dt>
        <dd>TSV columns</dd>
      </dl><dl><dt>

          <code><a href="update_report.html">update_report()</a></code>

        </dt>
        <dd>Function that updates a PCGR/CPSR report object structure</dd>
      </dl><dl><dt>

          <code><a href="variant_db_url.html">variant_db_url</a></code>

        </dt>
        <dd>List of URLS and variant identifiers for variant/gene/protein domain databases</dd>
      </dl><dl><dt>

          <code><a href="variant_stats_report.html">variant_stats_report()</a></code>

        </dt>
        <dd>Function that generate stats for a given variant set, considering number of variants/genes affected across tiers, types of variants ()</dd>
      </dl><dl><dt>

          <code><a href="write_processed_vcf.html">write_processed_vcf()</a></code>

        </dt>
        <dd>Function that writes a VCF intended for mutational signature analysis</dd>
      </dl><dl><dt>

          <code><a href="write_report_excel.html">write_report_excel()</a></code>

        </dt>
        <dd>Function that writes key datasets of PCGR object to an Excel workbook</dd>
      </dl><dl><dt>

          <code><a href="write_report_quarto_html.html">write_report_quarto_html()</a></code>

        </dt>
        <dd>Function that writes contents of PCGR object to an HTML report (quarto-based)</dd>
      </dl><dl><dt>

          <code><a href="write_report_tsv.html">write_report_tsv()</a></code>

        </dt>
        <dd>Function that writes contents of PCGR object to various output formats (Rmarkdown/flexdashboard HTML reports, JSON, tab-separated etc)</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

