<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ PCGR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">Intro</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/running.html">Running</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/input.html">Input files</a>
    <a class="dropdown-item" href="../articles/output.html">Output files</a>
    <a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a>
    <a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a>
    <a class="dropdown-item" href="../articles/developers.html">Developer notes</a>
    <a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/faq.html">FAQ</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="append_annotation_links.html">append_annotation_links()</a></code> 
        </dt>
        <dd>Function that appends multiple HTML annotation links to variant identifiers
e.g. COSMIC, CLINVAR, REFSEQ etc</dd>
      </dl><dl><dt>
          
          <code><a href="append_dbmts_var_link.html">append_dbmts_var_link()</a></code> 
        </dt>
        <dd>Function that adds miRNA target annotations (dbMTS) to genetic variant identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="append_dbnsfp_var_link.html">append_dbnsfp_var_link()</a></code> 
        </dt>
        <dd>Function that assigns HTML links to dbNSFP prediction entries</dd>
      </dl><dl><dt>
          
          <code><a href="append_drug_var_link.html">append_drug_var_link()</a></code> 
        </dt>
        <dd>Function that adds HTML links to different genetic variant identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="append_gwas_citation_phenotype.html">append_gwas_citation_phenotype()</a></code> 
        </dt>
        <dd>Function that adds GWAS citation/phenotype to GWAS hit found through PCGR annotation</dd>
      </dl><dl><dt>
          
          <code><a href="append_otargets_pheno_link.html">append_otargets_pheno_link()</a></code> 
        </dt>
        <dd>Function that adds HTML links to different genetic variant identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="append_pfeature_descriptions.html">append_pfeature_descriptions()</a></code> 
        </dt>
        <dd>Function that adds SwissProt feature descriptions based on
key identifiers coming from PCGR pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="append_read_support.html">append_read_support()</a></code> 
        </dt>
        <dd>Function that adds read support (depth, allelic fraction) for
tumor and normal and filters according to settings</dd>
      </dl><dl><dt>
          
          <code><a href="append_tcga_var_link.html">append_tcga_var_link()</a></code> 
        </dt>
        <dd>Function that adds TCGA annotations (cohort, frequency etc.) to variant identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="append_tfbs_annotation.html">append_tfbs_annotation()</a></code> 
        </dt>
        <dd>Function that adds TFBS annotations (dbMTS) to genetic variant identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="append_ucsc_segment_link.html">append_ucsc_segment_link()</a></code> 
        </dt>
        <dd>Function that appends a link to UCSC for a genomic segment</dd>
      </dl><dl><dt>
          
          <code><a href="assign_germline_popfreq_status.html">assign_germline_popfreq_status()</a></code> 
        </dt>
        <dd>Function that sets STATUS_POPFREQ_1KG_ABOVE_TOLERATED/
STATUS_POPFREQ_GNOMAD_ABOVE_TOLERATED to TRUE for variants
if any population frequency exceeds max_tolerated_af</dd>
      </dl><dl><dt>
          
          <code><a href="assign_mutation_type.html">assign_mutation_type()</a></code> 
        </dt>
        <dd>Function that assigns one of six mutation types to a list of mutations</dd>
      </dl><dl><dt>
          
          <code><a href="assign_somatic_classification.html">assign_somatic_classification()</a></code> 
        </dt>
        <dd>Function that assigns a SOMATIC_CLASSIFICATION to variants
based on evidence found in variant set,
potentially limited by user-defined options</dd>
      </dl><dl><dt>
          
          <code><a href="assign_somatic_germline_evidence.html">assign_somatic_germline_evidence()</a></code> 
        </dt>
        <dd>Function that appends several tags denoting
evidence for somatic/germline status of variants</dd>
      </dl><dl><dt>
          
          <code><a href="assign_tier1_tier2_acmg.html">assign_tier1_tier2_acmg()</a></code> 
        </dt>
        <dd>Function that assigns evidence items for SNVs/InDels to ACMG tiers 1 and 2</dd>
      </dl><dl><dt>
          
          <code><a href="assign_tier1_tier2_acmg_cna.html">assign_tier1_tier2_acmg_cna()</a></code> 
        </dt>
        <dd>Function that assigns evidence items for SCNAs to ACMG tiers 1 and 2</dd>
      </dl><dl><dt>
          
          <code><a href="cancer_phenotypes_regex.html">cancer_phenotypes_regex</a></code> 
        </dt>
        <dd>Regular expression of terms indicative of cancer-related phenotypes and syndromes</dd>
      </dl><dl><dt>
          
          <code><a href="check_common_colnames.html">check_common_colnames()</a></code> 
        </dt>
        <dd>Function that checks whether a set of column names are present in two
different data frames</dd>
      </dl><dl><dt>
          
          <code><a href="clinvar_germline_status.html">clinvar_germline_status()</a></code> 
        </dt>
        <dd>Function that assigns a logical to STATUS_CLINVAR_GERMLINE based on
whether a ClinVar entry of germline origin is found for a given variant
(for entries in a data frame)</dd>
      </dl><dl><dt>
          
          <code><a href="color_palette.html">color_palette</a></code> 
        </dt>
        <dd>Color encodings for report elements of PCGR/CPSR</dd>
      </dl><dl><dt>
          
          <code><a href="cosmic_somatic_status.html">cosmic_somatic_status()</a></code> 
        </dt>
        <dd>Function that assigns a logical (STATUS_COSMIC) reflecting whether
a variant co-incides with an entry in COSMIC (germline)</dd>
      </dl><dl><dt>
          
          <code><a href="cpsr_acmg.html">cpsr_acmg</a></code> 
        </dt>
        <dd>Scores and documentation of ACMG evidence criteria used for variant classification
in CPSR</dd>
      </dl><dl><dt>
          
          <code><a href="custom_bed_genes.html">custom_bed_genes()</a></code> 
        </dt>
        <dd>Function that reports protein-coding geneset that overlaps BED file</dd>
      </dl><dl><dt>
          
          <code><a href="data_integrity_check.html">data_integrity_check()</a></code> 
        </dt>
        <dd>Function that performs a sanity check of query VCF for pcgrr package,
ensuring that all required tags are present in annotated VCF</dd>
      </dl><dl><dt>
          
          <code><a href="dbsnp_germline_status.html">dbsnp_germline_status()</a></code> 
        </dt>
        <dd>Function that assigns a logical (STATUS_DBSNP_GERMLINE) reflecting whether
a variant co-incides with an entry in dbSNP (germline)</dd>
      </dl><dl><dt>
          
          <code><a href="deduplicate_eitems.html">deduplicate_eitems()</a></code> 
        </dt>
        <dd>Function that removes redundancy in variant evidence items (i.e. if
a variant is assicated with evidence at the codon level, evidence
at the exon/gene level is ignored)</dd>
      </dl><dl><dt>
          
          <code><a href="detect_vcf_sample_name.html">detect_vcf_sample_name()</a></code> 
        </dt>
        <dd>A function that detects whether the sample name in
variant data frame is unique (as present in column name VCF_SAMPLE_ID), throws an error if
multiple sample names are present for the CPSR workflow</dd>
      </dl><dl><dt>
          
          <code><a href="determine_genotype.html">determine_genotype()</a></code> 
        </dt>
        <dd>Function that assigns explicit genotype (heterozygous/homozygous) from VCF GT tag</dd>
      </dl><dl><dt>
          
          <code><a href="df_string_replace.html">df_string_replace()</a></code> 
        </dt>
        <dd>Function that performs stringr::str_replace on strings of multiple
string columns of a dataframe</dd>
      </dl><dl><dt>
          
          <code><a href="effect_prediction_algos.html">effect_prediction_algos</a></code> 
        </dt>
        <dd>List of URLs for a range of variant effect prediction algorithms</dd>
      </dl><dl><dt>
          
          <code><a href="filter_eitems_by_site.html">filter_eitems_by_site()</a></code> 
        </dt>
        <dd>Function that filters clinical evidence items by tumor type/primary site</dd>
      </dl><dl><dt>
          
          <code><a href="filter_read_support.html">filter_read_support()</a></code> 
        </dt>
        <dd>Function that filters variant set on (depth, allelic fraction)
for tumor and normal and filters according to settings</dd>
      </dl><dl><dt>
          
          <code><a href="generate_annotation_link.html">generate_annotation_link()</a></code> 
        </dt>
        <dd>A function that generates a HTML link for selected
identifiers (DBSNP, COSMIC, CLINVAR, ENTREZ)</dd>
      </dl><dl><dt>
          
          <code><a href="generate_pcgr_report.html">generate_pcgr_report()</a></code> 
        </dt>
        <dd>Function that generates all contents of the cancer genome report (PCGR)</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_cna.html">generate_report_data_cna()</a></code> 
        </dt>
        <dd>Function that annotates CNV segment files</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_kataegis.html">generate_report_data_kataegis()</a></code> 
        </dt>
        <dd>Function that generates data frame with potential kataegis events</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_msi.html">generate_report_data_msi()</a></code> 
        </dt>
        <dd>Function that generates MSI prediction data for PCGR report</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_rainfall.html">generate_report_data_rainfall()</a></code> 
        </dt>
        <dd>Function that generates data for rainfall plot (mutation density
along genome, considering SNVs only)</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_signatures_mp.html">generate_report_data_signatures_mp()</a></code> 
        </dt>
        <dd>Function that generates mutational signatures data for PCGR report</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_snv_indel.html">generate_report_data_snv_indel()</a></code> 
        </dt>
        <dd>Function that generates tiered variant sets for SNVs/InDels</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_tmb.html">generate_report_data_tmb()</a></code> 
        </dt>
        <dd>Function that tiered variant sets for SNVs/InDels</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_trials.html">generate_report_data_trials()</a></code> 
        </dt>
        <dd>Function that retrieves relevant (interventional based on molecular target)
clinical trials for a given tumor type</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_tumor_only.html">generate_report_data_tumor_only()</a></code> 
        </dt>
        <dd>Function that generates germline-filtered callset and PCGR
report statistics for a given tumor-only callsets</dd>
      </dl><dl><dt>
          
          <code><a href="generate_report_data_value_box.html">generate_report_data_value_box()</a></code> 
        </dt>
        <dd>Function that generates value box data for PCGR report</dd>
      </dl><dl><dt>
          
          <code><a href="generate_tier_tsv.html">generate_tier_tsv()</a></code> 
        </dt>
        <dd>Function that generates dense and tiered annotated variant datasets</dd>
      </dl><dl><dt>
          
          <code><a href="get_calls.html">get_calls()</a></code> 
        </dt>
        <dd>Function that reads a fully annotated VCF from PCGR VEP/vcfanno pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="get_clin_assocs_cna.html">get_clin_assocs_cna()</a></code> 
        </dt>
        <dd>Function that retrieves clinical evidence items (CIVIC, CBMDB) for
CNA aberrations</dd>
      </dl><dl><dt>
          
          <code><a href="get_clin_assocs_snv_indel.html">get_clin_assocs_snv_indel()</a></code> 
        </dt>
        <dd>Function that matches clinical evidence items (CIVIC, CBMDB)
against somatic cancer variants detected in tumor</dd>
      </dl><dl><dt>
          
          <code><a href="get_cna_cytoband.html">get_cna_cytoband()</a></code> 
        </dt>
        <dd>Function that gets the chromosome bands of copy number segments</dd>
      </dl><dl><dt>
          
          <code><a href="get_cna_overlapping_transcripts.html">get_cna_overlapping_transcripts()</a></code> 
        </dt>
        <dd>Functions that finds all transcripts that overlap with CNA segments. A
new data frame with one entry per transcript-CNA overlap entry is returned</dd>
      </dl><dl><dt>
          
          <code><a href="get_genome_obj.html">get_genome_obj()</a></code> 
        </dt>
        <dd>Get BSgenome Object</dd>
      </dl><dl><dt>
          
          <code><a href="get_oncogene_tsgene_target_sets.html">get_oncogene_tsgene_target_sets()</a></code> 
        </dt>
        <dd>Function that detects and assigns oncogenes subject to copy number
amplifications, and tumor suppressor genes subject to homozygous deletions
Also detects other drug targets subject to copy number amplifications</dd>
      </dl><dl><dt>
          
          <code><a href="get_ordinary_chromosomes.html">get_ordinary_chromosomes()</a></code> 
        </dt>
        <dd>Function that excludes genomic aberrations from non-nuclear chromosomes</dd>
      </dl><dl><dt>
          
          <code><a href="get_population_tag.html">get_population_tag()</a></code> 
        </dt>
        <dd>Function that retrieves name of VCF INFO tag and
population description for gnomad/1000G population</dd>
      </dl><dl><dt>
          
          <code><a href="get_prevalent_site_signatures.html">get_prevalent_site_signatures()</a></code> 
        </dt>
        <dd>Function that retrieves prevalent signatures for a given tumor type/primary site
Data is collected from COSMIC v3.2.</dd>
      </dl><dl><dt>
          
          <code><a href="get_proper_maf_alleles.html">get_proper_maf_alleles()</a></code> 
        </dt>
        <dd>Function that transforms a tier-structured variant data frame
into a MAF-like data frame (for input to 2020plus, MutSigCV)</dd>
      </dl><dl><dt>
          
          <code><a href="get_valid_chromosome_segments.html">get_valid_chromosome_segments()</a></code> 
        </dt>
        <dd>Function that removes copy number segments that go
beyond chrosomal lengths for the given assembly</dd>
      </dl><dl><dt>
          
          <code><a href="get_valid_chromosomes.html">get_valid_chromosomes()</a></code> 
        </dt>
        <dd>Checks for valid chromosome names in data frame of variants</dd>
      </dl><dl><dt>
          
          <code><a href="het_af_germline_status.html">het_af_germline_status()</a></code> 
        </dt>
        <dd>Function that assigns a logical (STATUS_LIKELY_GERMLINE_HETEROZYGOUS) reflecting whether
a variant is likely heterozygous (germline) - based on allelic fraction (AF_TUMOR),
presence in gnomAD and dbSNP, and no presence in TCGA and COSMIC</dd>
      </dl><dl><dt>
          
          <code><a href="heterozygous_states.html">heterozygous_states</a></code> 
        </dt>
        <dd>VCF encodings of heterozygous variant states</dd>
      </dl><dl><dt>
          
          <code><a href="hom_af_status.html">hom_af_status()</a></code> 
        </dt>
        <dd>Function that assigns a logical (STATUS_LIKELY_GERMLINE_HOMOZYGOUS) reflecting whether
a variant is likely homozygous (germline) - based on allelic fraction (AF_TUMOR)</dd>
      </dl><dl><dt>
          
          <code><a href="homozygous_states.html">homozygous_states</a></code> 
        </dt>
        <dd>VCF encodings of homozygous variant states</dd>
      </dl><dl><dt>
          
          <code><a href="init_cna_content.html">init_cna_content()</a></code> 
        </dt>
        <dd>Function that initiates report element with CNA information</dd>
      </dl><dl><dt>
          
          <code><a href="init_germline_content.html">init_germline_content()</a></code> 
        </dt>
        <dd>Function that initiates report element with germline variant information (CPSR)</dd>
      </dl><dl><dt>
          
          <code><a href="init_m_signature_content.html">init_m_signature_content()</a></code> 
        </dt>
        <dd>Function that initiates report element with mutational signatures information</dd>
      </dl><dl><dt>
          
          <code><a href="init_rainfall_content.html">init_rainfall_content()</a></code> 
        </dt>
        <dd>Function that initiates report element with rainfall information</dd>
      </dl><dl><dt>
          
          <code><a href="init_report.html">init_report()</a></code> 
        </dt>
        <dd>Function that initiates PCGR/CPSR report object</dd>
      </dl><dl><dt>
          
          <code><a href="init_report_display_content.html">init_report_display_content()</a></code> 
        </dt>
        <dd>Function that initiates ranked report display information</dd>
      </dl><dl><dt>
          
          <code><a href="init_snv_indel_content.html">init_snv_indel_content()</a></code> 
        </dt>
        <dd>Function that initiates report element with SNV/InDel information</dd>
      </dl><dl><dt>
          
          <code><a href="init_tmb_content.html">init_tmb_content()</a></code> 
        </dt>
        <dd>Function that initiates report element with TMB information</dd>
      </dl><dl><dt>
          
          <code><a href="init_tumor_only_content.html">init_tumor_only_content()</a></code> 
        </dt>
        <dd>Function that initiates report element with tumor-only information</dd>
      </dl><dl><dt>
          
          <code><a href="init_valuebox_content.html">init_valuebox_content()</a></code> 
        </dt>
        <dd>Function that initiates report element with value box information</dd>
      </dl><dl><dt>
          
          <code><a href="init_var_content.html">init_var_content()</a></code> 
        </dt>
        <dd>Function that initiates report element with variant data</dd>
      </dl><dl><dt>
          
          <code><a href="kataegis_detect.html">kataegis_detect()</a></code> 
        </dt>
        <dd>Function that detects kataegis events from a data frame
with genomic cooordinates of mutations</dd>
      </dl><dl><dt>
          
          <code><a href="kataegis_input.html">kataegis_input()</a></code> 
        </dt>
        <dd>Function that detects kataegis events from a data frame
with genomic cooordinates of mutations</dd>
      </dl><dl><dt>
          
          <code><a href="load_all_eitems.html">load_all_eitems()</a></code> 
        </dt>
        <dd>Function that loads all evidence items from CIViC and CGI, and
combines them in a unified data.frame</dd>
      </dl><dl><dt>
          
          <code><a href="load_eitems.html">load_eitems()</a></code> 
        </dt>
        <dd>Function that loads specific set of clinical variant evidence items (CIViC + CGI) based
on given parameters (mutation type, variant origin, tumor type etc)</dd>
      </dl><dl><dt>
          
          <code><a href="log_var_eitem_stats.html">log_var_eitem_stats()</a></code> 
        </dt>
        <dd>Function that logs the number of evidence items found, for different
levels of resolution</dd>
      </dl><dl><dt>
          
          <code><a href="make_upset_plot_data.html">make_upset_plot_data()</a></code> 
        </dt>
        <dd>Function that makes input data for an UpSet plot
(filtering/intersection results) for the somatic-germline
classification procedure</dd>
      </dl><dl><dt>
          
          <code><a href="match_eitems_to_var.html">match_eitems_to_var()</a></code> 
        </dt>
        <dd>Function that matches variants to evidence items</dd>
      </dl><dl><dt>
          
          <code><a href="max_af_gnomad.html">max_af_gnomad()</a></code> 
        </dt>
        <dd>Function that assigns a maximum value to a variable (MAX_AF_GNOMAD) reflecting
the maximum allele frequency for a given variant across gnomAD populations</dd>
      </dl><dl><dt>
          
          <code><a href="max_af_onekg.html">max_af_onekg()</a></code> 
        </dt>
        <dd>Function that assigns a maximum value to a variable (MAX_AF_1KG) reflecting
the maximum allele frequency for a given variant across oneKG populations</dd>
      </dl><dl><dt>
          
          <code><a href="msi_indel_fraction_plot.html">msi_indel_fraction_plot()</a></code> 
        </dt>
        <dd>Function that plots the indel fraction for a given sample and
contrasts this with the distribution for MSI-H/MSS samples from TCGA</dd>
      </dl><dl><dt>
          
          <code><a href="order_variants.html">order_variants()</a></code> 
        </dt>
        <dd>Function that orders genomic aberrations according to order
of chromosomes and chromosomal position</dd>
      </dl><dl><dt>
          
          <code><a href="pkg_exists.html">pkg_exists()</a></code> 
        </dt>
        <dd>Does R Package Exist</dd>
      </dl><dl><dt>
          
          <code><a href="plot_tmb_primary_site_tcga.html">plot_tmb_primary_site_tcga()</a></code> 
        </dt>
        <dd>Function that makes a plot with TMB boxplots for TCGA cohorts, highlighting
the TMB estimate for a given sample and the cohort/primary site of interest</dd>
      </dl><dl><dt>
          
          <code><a href="plot_value_boxes.html">plot_value_boxes()</a></code> 
        </dt>
        <dd>Function that plots four value boxes with the most
important findings in the cancer genome</dd>
      </dl><dl><dt>
          
          <code><a href="pon_status.html">pon_status()</a></code> 
        </dt>
        <dd>Function that assigns a logical (STATUS_PON) reflecting whether
a variant is co-inciding with a variant present in a panel-of-normals database
(PANEL_OF_NORMALS column is TRUE)</dd>
      </dl><dl><dt>
          
          <code><a href="predict_msi_status.html">predict_msi_status()</a></code> 
        </dt>
        <dd>Function that predicts MSI status based on fraction of indels among calls</dd>
      </dl><dl><dt>
          
          <code><a href="qc_var_eitems.html">qc_var_eitems()</a></code> 
        </dt>
        <dd>Function that makes a quality control check of evidence items assigned
to variants</dd>
      </dl><dl><dt>
          
          <code><a href="recode_logical_vars.html">recode_logical_vars()</a></code> 
        </dt>
        <dd>Function that assigns proper logical values to logical entries
read from VCF file (TRUE = True (VCF), FALSE = False (VCF))</dd>
      </dl><dl><dt>
          
          <code><a href="remove_cols_from_df.html">remove_cols_from_df()</a></code> 
        </dt>
        <dd>Function that removes column(s) from data frame</dd>
      </dl><dl><dt>
          
          <code><a href="set_report_metadata.html">set_report_metadata()</a></code> 
        </dt>
        <dd>Function that set PCGR report metadata</dd>
      </dl><dl><dt>
          
          <code><a href="sort_chromosomal_segments.html">sort_chromosomal_segments()</a></code> 
        </dt>
        <dd>Function that sorts chromosomal segments according to chromosome
and chromosomal start/end position</dd>
      </dl><dl><dt>
          
          <code><a href="structure_var_eitems.html">structure_var_eitems()</a></code> 
        </dt>
        <dd>Function that structures variant evidence items according
to strength of evidence</dd>
      </dl><dl><dt>
          
          <code><a href="targeted_drugs_pr_ttype.html">targeted_drugs_pr_ttype()</a></code> 
        </dt>
        <dd>Function that retrieves targeted drugs for a given tumor type/primary site</dd>
      </dl><dl><dt>
          
          <code><a href="targeted_drugs_summarise.html">targeted_drugs_summarise()</a></code> 
        </dt>
        <dd>Function that summarises available targeted drugs for a given target</dd>
      </dl><dl><dt>
          
          <code><a href="tcga_somatic_status.html">tcga_somatic_status()</a></code> 
        </dt>
        <dd>Function that assigns a logical (STATUS_TCGA_SOMATIC) reflecting whether
a variant co-incides with an entry in TCGA (somatic)</dd>
      </dl><dl><dt>
          
          <code><a href="tier_af_distribution.html">tier_af_distribution()</a></code> 
        </dt>
        <dd>Function that plots a histogram of the the variant allelic
support (tumor) - grouped by tiers</dd>
      </dl><dl><dt>
          
          <code><a href="update_maf_allelic_support.html">update_maf_allelic_support()</a></code> 
        </dt>
        <dd>Function that updates a MAF file (produced by vcf2maf) with allelic support data</dd>
      </dl><dl><dt>
          
          <code><a href="update_report.html">update_report()</a></code> 
        </dt>
        <dd>Function that initates PCGR report object</dd>
      </dl><dl><dt>
          
          <code><a href="upset_plot_tumor_only.html">upset_plot_tumor_only()</a></code> 
        </dt>
        <dd>Function that makes an upset calls for germline-filtered variants
classification procedure</dd>
      </dl><dl><dt>
          
          <code><a href="variant_db_url.html">variant_db_url</a></code> 
        </dt>
        <dd>List of URLS and variant identifiers for variant/gene/protein domain databases</dd>
      </dl><dl><dt>
          
          <code><a href="variant_stats_report.html">variant_stats_report()</a></code> 
        </dt>
        <dd>Function that generate snv/indel + coding/noncoding stats for
a given variant set</dd>
      </dl><dl><dt>
          
          <code><a href="virtual_panel_display_html.html">virtual_panel_display_html()</a></code> 
        </dt>
        <dd>Function that makes a HTML display of virtual gene panel</dd>
      </dl><dl><dt>
          
          <code><a href="write_processed_vcf.html">write_processed_vcf()</a></code> 
        </dt>
        <dd>Function that writes a VCF intended for mutational signature analysis</dd>
      </dl><dl><dt>
          
          <code><a href="write_report_output.html">write_report_output()</a></code> 
        </dt>
        <dd>Function that writes contents of PCGR object to various output formats
(Rmarkdown/flexdashboard HTML reports, JSON, tab-separated etc)</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

