<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function that makes input data for an UpSet plot (filtering/intersection results) for the somatic-germline classification procedure — plot_filtering_stats_germline • PCGR</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function that makes input data for an UpSet plot (filtering/intersection results) for the somatic-germline classification procedure — plot_filtering_stats_germline"><meta name="description" content="param calls unfiltered calls (germline + somatic)
param config config"><meta property="og:description" content="param calls unfiltered calls (germline + somatic)
param config config"><meta property="og:image" content="https://sigven.github.io/pcgr/logo.png"><script async src="https://www.googletagmanager.com/gtag/js?id=G-7HRLE8GQRE"></script><script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){dataLayer.push(arguments);}
   gtag('js', new Date());

   gtag('config', 'G-7HRLE8GQRE');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/running.html">Running</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/input.html">Input files</a></li>
    <li><a class="dropdown-item" href="../articles/output.html">Output files</a></li>
    <li><a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a></li>
    <li><a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="../articles/developers.html">Developer notes</a></li>
    <li><a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function that makes input data for an UpSet plot (filtering/intersection results) for the somatic-germline classification procedure</h1>
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/main/R/germline.R" class="external-link"><code>R/germline.R</code></a></small>
      <div class="d-none name"><code>plot_filtering_stats_germline.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>param calls unfiltered calls (germline + somatic)
param config config</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_filtering_stats_germline</span><span class="op">(</span></span>
<span>  report <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_margin_top <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  plot_margin_bottom <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  plot_margin_left <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  plot_margin_right <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  font_family <span class="op">=</span> <span class="st">"Helvetica"</span>,</span>
<span>  font_size <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  pie_line_width <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  opacity_filtered_categories <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  hole_size_pie <span class="op">=</span> <span class="fl">0.4</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-report">report<a class="anchor" aria-label="anchor" href="#arg-report"></a></dt>
<dd><p>list object with PCGR report content</p></dd>


<dt id="arg-plot-margin-top">plot_margin_top<a class="anchor" aria-label="anchor" href="#arg-plot-margin-top"></a></dt>
<dd><p>top margin of the plot</p></dd>


<dt id="arg-plot-margin-bottom">plot_margin_bottom<a class="anchor" aria-label="anchor" href="#arg-plot-margin-bottom"></a></dt>
<dd><p>bottom margin of the plot</p></dd>


<dt id="arg-plot-margin-left">plot_margin_left<a class="anchor" aria-label="anchor" href="#arg-plot-margin-left"></a></dt>
<dd><p>left margin of the plot</p></dd>


<dt id="arg-plot-margin-right">plot_margin_right<a class="anchor" aria-label="anchor" href="#arg-plot-margin-right"></a></dt>
<dd><p>right margin of the plot</p></dd>


<dt id="arg-font-family">font_family<a class="anchor" aria-label="anchor" href="#arg-font-family"></a></dt>
<dd><p>font family for plot text</p></dd>


<dt id="arg-font-size">font_size<a class="anchor" aria-label="anchor" href="#arg-font-size"></a></dt>
<dd><p>font size for plot text</p></dd>


<dt id="arg-pie-line-width">pie_line_width<a class="anchor" aria-label="anchor" href="#arg-pie-line-width"></a></dt>
<dd><p>line width for pie chart segments</p></dd>


<dt id="arg-opacity-filtered-categories">opacity_filtered_categories<a class="anchor" aria-label="anchor" href="#arg-opacity-filtered-categories"></a></dt>
<dd><p>opacity for filtered categories</p></dd>


<dt id="arg-hole-size-pie">hole_size_pie<a class="anchor" aria-label="anchor" href="#arg-hole-size-pie"></a></dt>
<dd><p>size of the hole in the pie chart</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>filtering_stats list with data frame and plotly pie chart</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>return upset data</p>
<p>export
make_upset_plot_data &lt;- function(calls, config) columns &lt;- c()
if (config["tumor_only"]["exclude_pon"] == TRUE)
columns &lt;- c(columns, "STATUS_PON")if (config["tumor_only"]["exclude_likely_hom_germline"] == TRUE)
columns &lt;- c(columns, "STATUS_GERMLINE_HOM")if (config["tumor_only"]["exclude_likely_het_germline"] == TRUE)
columns &lt;- c(columns, "STATUS_GERMLINE_HET")if (config["tumor_only"]["exclude_dbsnp_nonsomatic"] == TRUE)
columns &lt;- c(columns, "STATUS_DBSNP")assertable::assert_colnames(
calls, c("VAR_ID",
"gnomADe_AF_ABOVE_TOLERATED",
"STATUS_CLINVAR_GERMLINE"),
only_colnames = F, quiet = T)
df &lt;- dplyr::select(calls, .data$VAR_ID,
.data$gnomADe_AF_ABOVE_TOLERATED,
.data$STATUS_CLINVAR_GERMLINE)
for (c in columns)
if (c %in% colnames(calls))
df, c &lt;- calls, cfor (v in colnames(df))
if (v != "VAR_ID")
df, v &lt;- as.integer(df, v)df &lt;- df |&gt;
dplyr::rename(
gnomAD = .data$gnomADe_AF_ABOVE_TOLERATED,
ClinVar = .data$STATUS_CLINVAR_GERMLINE)
if ("STATUS_PON" %in% colnames(df))
df &lt;- dplyr::rename(df, Panel_Of_Normals = .data$STATUS_PON)if ("STATUS_GERMLINE_HOM" %in% colnames(df))
df &lt;- dplyr::rename(df, HomAF = .data$STATUS_GERMLINE_HOM)if ("STATUS_GERMLINE_HET" %in% colnames(df))
df &lt;- dplyr::rename(df, HetAF = .data$STATUS_GERMLINE_HET)if ("STATUS_DBSNP" %in% colnames(df))
df &lt;- dplyr::rename(df, dbSNP = .data$STATUS_DBSNP)return(df)</p>
<p>#' #' Function that makes an upset calls for germline-filtered variants
#' classification procedure
#'
#' param upset_data unfiltered calls (germline + somatic)
#'
#' return p
#'
#' export
upset_plot_tumor_only &lt;- function(upset_data) isets &lt;- c()
all_negative_filters &lt;- c()
for (c in colnames(upset_data))
if (c != "VAR_ID")
if (length(unique(upset_data, c == 1)) == 1)
if (unique(upset_data, c == 1) == T)
isets &lt;- c(isets, c)
else
all_negative_filters &lt;- c(all_negative_filters, c)else
isets &lt;- c(isets, c)for (m in all_negative_filters)
upset_data, m &lt;- NULLp &lt;- UpSetR::upset(upset_data, sets = isets,
sets.bar.color = "#56B4E9",
order.by = "freq", nintersects = 20,
text.scale = 1.5, show.numbers = T,
point.size = 6, color.pal = "Blues",
empty.intersections = "on")
return(p)
Function that generates variant filtering statistics (i.e. removal of
likely non-somatic/germline events) for callsets coming from
tumor-only sequencing</p>
<p>#' param callset list object with unfiltered calls (callset$variant_unfiltered)
#' param settings list object with PCGR run configuration settings
#'
#' export
tumor_only_vfilter_stats &lt;-
function(callset = NULL,
settings = NULL) </p><div class="sourceCode"><pre><code><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="do">## raw, unfiltered set of (PASSED) variant calls that</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="do">## should be subject to filtering</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>vcalls <span class="ot">&lt;-</span> callset<span class="sc">$</span>variant_unfiltered</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>to_settings <span class="ot">&lt;-</span> settings<span class="sc">$</span>conf<span class="sc">$</span>somatic_snv<span class="sc">$</span>tumor_only<span class="do">## initiate report object with tumor-only variant filter stats</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>to_stats <span class="ot">&lt;-</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>  pcgrr<span class="sc">::</span><span class="fu">init_tumor_only_content</span>()to_stats[[<span class="st">'vfilter'</span>]][[<span class="st">'unfiltered_n'</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>  <span class="fu">NROW</span>(vcalls)<span class="do">## Assign statistics to successive filtering levels for</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="do">## different evidence criteria</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="do">## excluded germline calls found in gnomAD</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>to_stats[[<span class="st">"vfilter"</span>]][[<span class="st">"gnomad_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  <span class="fu">NROW</span>(vcalls) <span class="sc">-</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>  <span class="fu">NROW</span>(vcalls[vcalls<span class="sc">$</span>SOMATIC_CLASSIFICATION <span class="sc">==</span> <span class="st">"GERMLINE_GNOMAD"</span>, ])</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co"># pcgrr::log4r_info(paste0("Excluding coinciding germline variants in ",</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co">#                          "gnomAD populations"))</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co"># pcgrr::log4r_info(paste0("Total sample calls remaining: ",</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co">#                          to_stats$vfilter[["gnomad_n_remain"]]))## excluded germline calls found in ClinVar</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>to_stats<span class="sc">$</span>vfilter[[<span class="st">"clinvar_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>  to_stats<span class="sc">$</span>vfilter[[<span class="st">"gnomad_n_remain"</span>]] <span class="sc">-</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>  <span class="fu">NROW</span>(vcalls[vcalls<span class="sc">$</span>SOMATIC_CLASSIFICATION <span class="sc">==</span> <span class="st">"GERMLINE_CLINVAR"</span>, ])</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a><span class="co"># pcgrr::log4r_info(paste0("Excluding coinciding germline variants in ClinVar"))</span></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="co"># pcgrr::log4r_info(paste0("Total sample calls remaining: ",</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="co">#                          to_stats$vfilter[["clinvar_n_remain"]]))## excluded germline calls found in panel of normals (if provided)</span></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a>to_stats<span class="sc">$</span>vfilter[[<span class="st">"pon_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a>  to_stats<span class="sc">$</span>vfilter[[<span class="st">"clinvar_n_remain"</span>]]</span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">as.logical</span>(to_settings[[<span class="st">"exclude_pon"</span>]]) <span class="sc">==</span> <span class="cn">TRUE</span>) {</span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a>  to_stats<span class="sc">$</span>vfilter[[<span class="st">"pon_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a>    to_stats<span class="sc">$</span>vfilter[[<span class="st">"pon_n_remain"</span>]] <span class="sc">-</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a>    <span class="fu">NROW</span>(vcalls[vcalls<span class="sc">$</span>SOMATIC_CLASSIFICATION <span class="sc">==</span> <span class="st">"GERMLINE_PON"</span>, ])</span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a>  <span class="co"># pcgrr::log4r_info(</span></span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a>  <span class="co">#   paste0("Excluding putative germline variants found in calls ",</span></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a>  <span class="co">#          "from panel-of-normals (PON)"))</span></span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a>  <span class="co"># pcgrr::log4r_info(</span></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a>  <span class="co">#   paste0("Total sample calls remaining: ",</span></span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a>  <span class="co">#          to_stats$vfilter[["pon_n_remain"]]))</span></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a>}<span class="do">## excluded germline calls found with 100% allelic fraction</span></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="do">## (likely homozygous germline variants)</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a>to_stats<span class="sc">$</span>vfilter[[<span class="st">"hom_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a>  to_stats<span class="sc">$</span>vfilter[[<span class="st">"pon_n_remain"</span>]]</span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">as.logical</span>(to_settings[[<span class="st">"exclude_likely_hom_germline"</span>]]) <span class="sc">==</span> <span class="cn">TRUE</span>) {</span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a>  to_stats<span class="sc">$</span>vfilter[[<span class="st">"hom_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a>    to_stats<span class="sc">$</span>vfilter[[<span class="st">"hom_n_remain"</span>]] <span class="sc">-</span></span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a>    <span class="fu">NROW</span>(vcalls[vcalls<span class="sc">$</span>SOMATIC_CLASSIFICATION <span class="sc">==</span> <span class="st">"GERMLINE_HOMOZYGOUS"</span>, ])</span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a>  <span class="co"># pcgrr::log4r_info(</span></span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a>  <span class="co">#   paste0("Excluding likely homozygous germline variants found ",</span></span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a>  <span class="co">#          "as variants with 100% allelic fraction"))</span></span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a>  <span class="co"># pcgrr::log4r_info(paste0("Total sample calls remaining: ",</span></span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a>  <span class="co">#                          to_stats$vfilter[["hom_n_remain"]]))</span></span>
<span id="cb1-48"><a href="#cb1-48" tabindex="-1"></a>}<span class="do">## excluded germline calls found as likely heterozygous germline variants</span></span>
<span id="cb1-49"><a href="#cb1-49" tabindex="-1"></a>to_stats<span class="sc">$</span>vfilter[[<span class="st">"het_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-50"><a href="#cb1-50" tabindex="-1"></a>  to_stats<span class="sc">$</span>vfilter[[<span class="st">"hom_n_remain"</span>]]</span>
<span id="cb1-51"><a href="#cb1-51" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">as.logical</span>(to_settings[[<span class="st">"exclude_likely_het_germline"</span>]]) <span class="sc">==</span> <span class="cn">TRUE</span>) {</span>
<span id="cb1-52"><a href="#cb1-52" tabindex="-1"></a>  to_stats<span class="sc">$</span>vfilter[[<span class="st">"het_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-53"><a href="#cb1-53" tabindex="-1"></a>    to_stats<span class="sc">$</span>vfilter[[<span class="st">"het_n_remain"</span>]] <span class="sc">-</span></span>
<span id="cb1-54"><a href="#cb1-54" tabindex="-1"></a>    <span class="fu">NROW</span>(vcalls[vcalls<span class="sc">$</span>SOMATIC_CLASSIFICATION <span class="sc">==</span> <span class="st">"GERMLINE_HETEROZYGOUS"</span>, ])</span>
<span id="cb1-55"><a href="#cb1-55" tabindex="-1"></a>  <span class="co"># pcgrr::log4r_info(paste0(</span></span>
<span id="cb1-56"><a href="#cb1-56" tabindex="-1"></a>  <span class="co">#   "Excluding likely heterozygous germline variants found as variants ",</span></span>
<span id="cb1-57"><a href="#cb1-57" tabindex="-1"></a>  <span class="co">#   "with 40-60% allelic fraction and recorded in gnomAD + dbSNP"))</span></span>
<span id="cb1-58"><a href="#cb1-58" tabindex="-1"></a>  <span class="co"># pcgrr::log4r_info(paste0("Total sample calls remaining: ",</span></span>
<span id="cb1-59"><a href="#cb1-59" tabindex="-1"></a>  <span class="co">#                          to_stats$vfilter[["het_n_remain"]]))</span></span>
<span id="cb1-60"><a href="#cb1-60" tabindex="-1"></a>}<span class="do">## excluded calls with dbSNP germline status (if set in config)</span></span>
<span id="cb1-61"><a href="#cb1-61" tabindex="-1"></a>to_stats<span class="sc">$</span>vfilter[[<span class="st">"dbsnp_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-62"><a href="#cb1-62" tabindex="-1"></a>  to_stats<span class="sc">$</span>vfilter[[<span class="st">"het_n_remain"</span>]]</span>
<span id="cb1-63"><a href="#cb1-63" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">as.logical</span>(to_settings[[<span class="st">"exclude_dbsnp_nonsomatic"</span>]]) <span class="sc">==</span> <span class="cn">TRUE</span>) {  <span class="co">#pcgrr::log4r_info(</span></span>
<span id="cb1-64"><a href="#cb1-64" tabindex="-1"></a>  <span class="co">#  paste0("Excluding non-somatically associated dbSNP variants ",</span></span>
<span id="cb1-65"><a href="#cb1-65" tabindex="-1"></a>  <span class="co">#         "(dbSNP - not recorded as somatic in ClinVar",</span></span>
<span id="cb1-66"><a href="#cb1-66" tabindex="-1"></a>  <span class="co">#         "and not registered in COSMIC or found in TCGA"))  to_stats$vfilter[["dbsnp_n_remain"]] &lt;-</span></span>
<span id="cb1-67"><a href="#cb1-67" tabindex="-1"></a>    to_stats<span class="sc">$</span>vfilter[[<span class="st">"dbsnp_n_remain"</span>]] <span class="sc">-</span></span>
<span id="cb1-68"><a href="#cb1-68" tabindex="-1"></a>    <span class="fu">NROW</span>(vcalls[vcalls<span class="sc">$</span>SOMATIC_CLASSIFICATION <span class="sc">==</span> <span class="st">"GERMLINE_DBSNP"</span>, ])</span>
<span id="cb1-69"><a href="#cb1-69" tabindex="-1"></a>  <span class="co">#pcgrr::log4r_info(paste0("Total sample calls remaining: ",</span></span>
<span id="cb1-70"><a href="#cb1-70" tabindex="-1"></a>  <span class="co">#                         to_stats$vfilter[["dbsnp_n_remain"]]))</span></span>
<span id="cb1-71"><a href="#cb1-71" tabindex="-1"></a>}<span class="do">## excluded non-exonic calls (if set in config)</span></span>
<span id="cb1-72"><a href="#cb1-72" tabindex="-1"></a>to_stats<span class="sc">$</span>vfilter[[<span class="st">"nonexonic_n_remain"</span>]] <span class="ot">&lt;-</span></span>
<span id="cb1-73"><a href="#cb1-73" tabindex="-1"></a>  to_stats<span class="sc">$</span>vfilter[[<span class="st">"dbsnp_n_remain"</span>]]</span>
<span id="cb1-74"><a href="#cb1-74" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">as.logical</span>(to_settings[[<span class="st">"exclude_nonexonic"</span>]]) <span class="sc">==</span> <span class="cn">TRUE</span>) {  <span class="co">#pcgrr::log4r_info(</span></span>
<span id="cb1-75"><a href="#cb1-75" tabindex="-1"></a>  <span class="co">#  paste0("Excluding non-somatically associated dbSNP variants ",</span></span>
<span id="cb1-76"><a href="#cb1-76" tabindex="-1"></a>  <span class="co">#         "(dbSNP - not recorded as somatic in ClinVar",</span></span>
<span id="cb1-77"><a href="#cb1-77" tabindex="-1"></a>  <span class="co">#         "and not registered in COSMIC or found in TCGA"))  to_stats$vfilter[["nonexonic_n_remain"]] &lt;-</span></span>
<span id="cb1-78"><a href="#cb1-78" tabindex="-1"></a>    to_stats<span class="sc">$</span>vfilter[[<span class="st">"nonexonic_n_remain"</span>]] <span class="sc">-</span></span>
<span id="cb1-79"><a href="#cb1-79" tabindex="-1"></a>    <span class="fu">NROW</span>(vcalls[vcalls<span class="sc">$</span>EXONIC_STATUS <span class="sc">==</span> <span class="st">"nonexonic"</span>, ])</span>
<span id="cb1-80"><a href="#cb1-80" tabindex="-1"></a>  <span class="co">#pcgrr::log4r_info(paste0("Total sample calls remaining: ",</span></span>
<span id="cb1-81"><a href="#cb1-81" tabindex="-1"></a>  <span class="co">#                         to_stats$vfilter[["dbsnp_n_remain"]]))</span></span>
<span id="cb1-82"><a href="#cb1-82" tabindex="-1"></a>}to_stats[[<span class="st">"eval"</span>]] <span class="ot">&lt;-</span> <span class="fu">TRUEfor</span> (db_filter <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"gnomad"</span>, <span class="st">"dbsnp"</span>, <span class="st">"pon"</span>,</span>
<span id="cb1-83"><a href="#cb1-83" tabindex="-1"></a>                    <span class="st">"clinvar"</span>, <span class="st">"hom"</span>, <span class="st">"het"</span>, <span class="st">"nonexonic"</span>)) {</span>
<span id="cb1-84"><a href="#cb1-84" tabindex="-1"></a>  <span class="cf">if</span> (to_stats[[<span class="st">"vfilter"</span>]][[<span class="fu">paste0</span>(db_filter, <span class="st">"_n_remain"</span>)]] <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span></span>
<span id="cb1-85"><a href="#cb1-85" tabindex="-1"></a>      to_stats[[<span class="st">"vfilter"</span>]][[<span class="st">"unfiltered_n"</span>]] <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb1-86"><a href="#cb1-86" tabindex="-1"></a>    to_stats[[<span class="st">"vfilter"</span>]][[<span class="fu">paste0</span>(db_filter, <span class="st">"_frac_remain"</span>)]] <span class="ot">&lt;-</span></span>
<span id="cb1-87"><a href="#cb1-87" tabindex="-1"></a>      <span class="fu">round</span>((<span class="fu">as.numeric</span>(to_stats[[<span class="st">"vfilter"</span>]][[<span class="fu">paste0</span>(db_filter,</span>
<span id="cb1-88"><a href="#cb1-88" tabindex="-1"></a>                                                          <span class="st">"_n_remain"</span>)]]) <span class="sc">/</span></span>
<span id="cb1-89"><a href="#cb1-89" tabindex="-1"></a>               to_stats[[<span class="st">"vfilter"</span>]][[<span class="st">"unfiltered_n"</span>]]) <span class="sc">*</span> <span class="dv">100</span>, <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb1-90"><a href="#cb1-90" tabindex="-1"></a>  }</span>
<span id="cb1-91"><a href="#cb1-91" tabindex="-1"></a>}</span>
<span id="cb1-92"><a href="#cb1-92" tabindex="-1"></a><span class="fu">return</span>(to_stats)</span></code></pre></div>
Function that generates a pie chart for germline filtering statistics
for callsets coming from tumor-only sequencing
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

