
### Biomarkers {.tabset}

- Variants (__class 4/5__) in the query sample that overlap with reported clinical biomarkers from the [database for clinical interpretations of variants in cancer, CIViC](https://civic.genome.wustl.edu/#/home) are considered. Note that several variants in the query can overlap the same existing biomarker, given that biomarkers are reported at different resolutions (variant/gene level). Total number of clinical evidence items that coincide with query variants:
    * Predisposing: __`r NROW(cps_report[['content']][['snv_indel']][['clin_eitem']][['predisposing']])`__ evidence items
    * Predictive: __`r NROW(cps_report[['content']][['snv_indel']][['clin_eitem']][['predictive']])`__ evidence items
    * Prognostic: __`r NROW(cps_report[['content']][['snv_indel']][['clin_eitem']][['prognostic']])`__ evidence items
    * Diagnostic: __`r NROW(cps_report[['content']][['snv_indel']][['clin_eitem']][['diagnostic']])`__ evidence items

<br><br>


```{r prepare_germline_data, echo=F, results='asis'}

show_germline_filters <- list()
missing_germline_items <- list()
biomarker_present <- F
for(type in c('diagnostic','prognostic','predictive','predisposing')){
  show_germline_filters[[type]] <- F
  missing_germline_items[[type]] <- T
  if(NROW(cps_report[['content']][['snv_indel']][['clin_eitem']][[type]]) > 0){
    show_germline_filters[[type]] <- T
    missing_germline_items[[type]] <- F
    biomarker_present <- T
  }
}

```


```{r biomarker_note, echo=F, results = "asis", include = biomarker_present}

cat('<b>NOTE:</b> Reported biomarkers in CIViC are mapped at different resolutions (i.e. filter <b>Biomarker mapping</b>). The accuracy of a match between variants in the tumor sample and the reported biomarkers will vary accordingly (highlighted by gene symbols with different color backgrounds):\n\n')

cat('<ul><li>Biomarker match at the <mark style="background-color:black; font-weight:bold; color:white">exact variant/codon level</mark></li>')
cat(paste0('<br><li>Biomarker match at the <mark style="background-color:', cps_report[['metadata']][['color_palette']][['warning']][['values']][1],'; color:white; font-weight:bold">exon/gene level</mark></li></ul>\n'))

htmltools::br()
```




```{r active_tab_cpsr_predisposing, echo = F, results = "asis"}
if(missing_germline_items[['predisposing']] == F | 
   (missing_germline_items[['diagnostic']] == T & 
    missing_germline_items[['prognostic']] == T &
    missing_germline_items[['predictive']] == T)){
  cat("")
  cat("#### Predisposing {.active}")
  cat("")
}else{
  cat("")
  cat("#### Predisposing")
  cat("")
}
```


<br>

```{r germline_predisposing, echo=F, results = 'asis', eval = show_germline_filters[['predisposing']]}

variants_germline_predisposing_shared <- crosstalk::SharedData$new(cps_report[['content']][['snv_indel']][['clin_eitem']][['predisposing']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_germline_predisposing_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical significance", variants_germline_predisposing_shared, ~CLINICAL_SIGNIFICANCE),
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_germline_predisposing_shared, ~EVIDENCE_LEVEL)
  ),
  list(
    crosstalk::filter_select("SYMBOL", "Gene", variants_germline_predisposing_shared, ~SYMBOL),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_germline_predisposing_shared, ~BIOMARKER_MAPPING)
  )
)

cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_germline_predisposing_shared, escape=F,extensions=c("Buttons","Responsive"), 
              options = 
                list(buttons = c('csv','excel'),
                     pageLength = 6,
                     dom = 'Bfrtip')) %>%
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated', 
                                                     'A: FDA/NCCN/ELN guidelines', 
                                                     'B: Clinical evidence', 
                                                     'B1: Clinical evidence: late trials', 
                                                     'B2: Clinical evidence: early trials', 
                                                     'C: Case study', 
                                                     'D: Preclinical evidence', 
                                                     'E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", 
                                                     "#56B4E9","#56B4E9","#0072B2",
                                                     "#E69F00", "#F0E442"))) %>%
  DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',cps_report[['metadata']][['color_palette']][['warning']][['values']][1], cps_report[['metadata']][['color_palette']][['warning']][['values']][1])))



```

```{r germline_predisposing_missing, echo=F, results = 'asis', eval = missing_germline_items[['predisposing']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br><br>


```{r active_tab_cpsr_predictive, echo = F, results = "asis"}
if(missing_germline_items[['predictive']] == F & 
   missing_germline_items[['predisposing']] == T){
  cat("")
  cat("#### Predictive {.active}")
  cat("")
}else{
  cat("")
  cat("#### Predictive")
  cat("")
}
```


<br>

```{r germline_predictive, echo=F, results = 'asis', eval = show_germline_filters[['predictive']]}

variants_germline_predictive_shared <- crosstalk::SharedData$new(cps_report[['content']][['snv_indel']][['clin_eitem']][['predictive']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_germline_predictive_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical significance", variants_germline_predictive_shared, ~CLINICAL_SIGNIFICANCE),
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_germline_predictive_shared, ~EVIDENCE_LEVEL)
  ),
  list(
    crosstalk::filter_select("SYMBOL", "Gene", variants_germline_predictive_shared, ~SYMBOL),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_germline_predictive_shared, ~BIOMARKER_MAPPING),
    crosstalk::filter_select("THERAPEUTIC_CONTEXT", "Therapeutic context", variants_germline_predictive_shared, ~THERAPEUTIC_CONTEXT)
  )
)

cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_germline_predictive_shared, escape=F,extensions=c("Buttons","Responsive"), options=list(buttons = c('csv','excel'),
                                                                                                               pageLength = 6, dom = 'Bfrtip')) %>%
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated',
                                                     'A: FDA/NCCN/ELN guidelines',
                                                     'B: Clinical evidence',
                                                     'B1: Clinical evidence: late trials',
                                                     'B2: Clinical evidence: early trials',
                                                     'C: Case study',
                                                     'D: Preclinical evidence',
                                                     'E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", 
                                                     "#56B4E9","#56B4E9","#0072B2",
                                                     "#E69F00", "#F0E442"))) %>%
  DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", 
                  fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',cps_report[['metadata']][['color_palette']][['warning']][['values']][1], cps_report[['metadata']][['color_palette']][['warning']][['values']][1])))



```

```{r germline_predictive_missing, echo=F, results = 'asis', eval = missing_germline_items[['predictive']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br><br>


```{r active_tab_cpsr_prognostic, echo = F, results = "asis"}
if(missing_germline_items[['prognostic']] == F & 
   missing_germline_items[['predisposing']] == T & 
   missing_germline_items[['predictive']] == T){
  cat("")
  cat("#### Prognostic {.active}")
  cat("")
}else{
  cat("")
  cat("#### Prognostic")
  cat("")
}
```


<br>

```{r germline_prognostic, echo=F, results='asis', eval = show_germline_filters[['prognostic']]}

variants_germline_prognostic_shared <- crosstalk::SharedData$new(cps_report[['content']][['snv_indel']][['clin_eitem']][['prognostic']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("SYMBOL", "Gene", variants_germline_prognostic_shared, ~SYMBOL),
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_germline_prognostic_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical significance", variants_germline_prognostic_shared, ~CLINICAL_SIGNIFICANCE)
  ),
  list(
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_germline_prognostic_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_germline_prognostic_shared, ~BIOMARKER_MAPPING)
  )
)
cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_germline_prognostic_shared, escape=F,extensions=c("Buttons","Responsive"), options=list(buttons = c('csv','excel'), 
                                                                                                               pageLength = 6, dom = 'Bfrtip')) %>%
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated',
                                                     'A: FDA/NCCN/ELN guidelines',
                                                     'B: Clinical evidence',
                                                     'B1: Clinical evidence: late trials',
                                                     'B2: Clinical evidence: early trials',
                                                     'C: Case study',
                                                     'D: Preclinical evidence',
                                                     'E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", 
                                                     "#56B4E9","#56B4E9","#0072B2",
                                                     "#E69F00", "#F0E442"))) %>%
  DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", 
                  fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'),
                                         c('#000','#000',cps_report[['metadata']][['color_palette']][['warning']][['values']][1], cps_report[['metadata']][['color_palette']][['warning']][['values']][1])))


```

```{r germline_prognostic_missing, echo=F, results = 'asis', eval = missing_germline_items[['prognostic']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br><br>

```{r active_tab_cpsr_diagnostic, echo = F, results = "asis"}
if(missing_germline_items[['diagnostic']] == F &
  missing_germline_items[['prognostic']] == T & 
   missing_germline_items[['predisposing']] == T & 
   missing_germline_items[['predictive']] == T){
  cat("")
  cat("#### Diagnostic {.active}")
  cat("")
}else{
  cat("")
  cat("#### Diagnostic")
  cat("")
}
```


<br>

```{r germline_diagnostic, echo=F, results='asis', eval = show_germline_filters[['diagnostic']]}

variants_germline_diagnostic_shared <- crosstalk::SharedData$new(cps_report[['content']][['snv_indel']][['clin_eitem']][['diagnostic']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("SYMBOL", "Gene", variants_germline_diagnostic_shared, ~SYMBOL),
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_germline_diagnostic_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical association", variants_germline_diagnostic_shared, ~CLINICAL_SIGNIFICANCE)

  ),
  list(
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_germline_diagnostic_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_germline_diagnostic_shared, ~BIOMARKER_MAPPING)
  )
)
cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_germline_diagnostic_shared, escape=F,extensions=c("Buttons","Responsive"), 
              options=list(buttons = c('csv','excel'), 
                           pageLength = 6, dom = 'Bfrtip')) %>%
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated',
                                                     'A: FDA/NCCN/ELN guidelines',
                                                     'B: Clinical evidence',
                                                     'B1: Clinical evidence: late trials',
                                                     'B2: Clinical evidence: early trials',
                                                     'C: Case study',
                                                     'D: Preclinical evidence',
                                                     'E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", 
                                                     "#56B4E9","#56B4E9","#0072B2",
                                                     "#E69F00", "#F0E442"))) %>%
  DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", 
                  fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'),
                                         c('#000','#000',cps_report[['metadata']][['color_palette']][['warning']][['values']][1], cps_report[['metadata']][['color_palette']][['warning']][['values']][1])))


```

```{r germline_diagnostic_missing, echo=F, results = 'asis', eval = missing_germline_items[['diagnostic']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br><br>


