

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>PCGR annotation resources &mdash; Personal Cancer Genome Reporter 0.9.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Input" href="input.html" />
    <link rel="prev" title="FAQ" href="FAQ.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Personal Cancer Genome Reporter
          

          
          </a>

          
            
            
              <div class="version">
                0.9.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">FAQ</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">PCGR annotation resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basic-variant-consequence-annotation">Basic variant consequence annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#insilico-predictions-of-effect-of-coding-variants"><em>Insilico</em> predictions of effect of coding variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="#variant-frequency-databases">Variant frequency databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="#variant-databases-of-clinical-utility">Variant databases of clinical utility</a></li>
<li class="toctree-l2"><a class="reference internal" href="#protein-domains-functional-features">Protein domains/functional features</a></li>
<li class="toctree-l2"><a class="reference internal" href="#knowledge-resources-on-gene-and-protein-targets">Knowledge resources on gene and protein targets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pathway-databases">Pathway databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="#notes-on-variant-annotation-datasets">Notes on variant annotation datasets</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome-mapping">Genome mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="#data-quality">Data quality</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="tier_systems.html">Tier models</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">CHANGELOG</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Personal Cancer Genome Reporter</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>PCGR annotation resources</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/annotation_resources.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="pcgr-annotation-resources">
<h1>PCGR annotation resources<a class="headerlink" href="#pcgr-annotation-resources" title="Permalink to this headline">¶</a></h1>
<div class="section" id="basic-variant-consequence-annotation">
<h2>Basic variant consequence annotation<a class="headerlink" href="#basic-variant-consequence-annotation" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="http://www.ensembl.org/info/docs/tools/vep/index.html">VEP</a> -
Variant Effect Predictor release 101 (<a class="reference external" href="https://www.gencodegenes.org/human/">GENCODE
v35</a> as gene reference
database (v19 for grch37))</p></li>
</ul>
</div>
<div class="section" id="insilico-predictions-of-effect-of-coding-variants">
<h2><em>Insilico</em> predictions of effect of coding variants<a class="headerlink" href="#insilico-predictions-of-effect-of-coding-variants" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="https://sites.google.com/site/jpopgen/dbNSFP">dBNSFP</a> - database
of non-synonymous functional predictions (v4.1, June 2020)</p></li>
</ul>
</div>
<div class="section" id="variant-frequency-databases">
<h2>Variant frequency databases<a class="headerlink" href="#variant-frequency-databases" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="http://exac.broadinstitute.org/">gnomAD</a> - germline variant
frequencies exome-wide (r2.1, October 2018)</p></li>
<li><p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/SNP/">dbSNP</a> - database of short
genetic variants (build 153)</p></li>
<li><p><a class="reference external" href="http://cancerhotspots.org">Cancer Hotspots</a> - a resource for
statistically significant mutations in cancer (v2, 2017)</p></li>
<li><p><a class="reference external" href="https://portal.gdc.cancer.gov/">TCGA</a> - somatic mutations
discovered across 33 tumor type cohorts (release 27.0, October 2020)</p></li>
<li><p><a class="reference external" href="http://docs.icgc.org/pcawg/">ICGC-PCAWG</a> - ICGC Pancancer
Analysis of Whole Genomes - (release 28, March 17th, 2019)</p></li>
</ul>
</div>
<div class="section" id="variant-databases-of-clinical-utility">
<h2>Variant databases of clinical utility<a class="headerlink" href="#variant-databases-of-clinical-utility" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/clinvar/">ClinVar</a> - database of
clinically related variants (November 2020)</p></li>
<li><p><a class="reference external" href="http://docm.genome.wustl.edu">DoCM</a> - database of curated
mutations (v3.2, April 2016)</p></li>
<li><p><a class="reference external" href="http://civic.genome.wustl.edu">CIViC</a> - clinical interpretations
of variants in cancer (November 18th 2020)</p></li>
<li><p><a class="reference external" href="http://www.cancergenomeinterpreter.org/biomarkers">CGI</a> - Cancer
Genome Interpreter Cancer Biomarkers Database (CGI) (January 17th
2018)</p></li>
<li><p><a class="reference external" href="https://www.ebi.ac.uk/chembl/">ChEMBL</a> - database of drugs,
drug-like small molecules and their targets (ChEMBL_27, May 2020)</p></li>
</ul>
</div>
<div class="section" id="protein-domains-functional-features">
<h2>Protein domains/functional features<a class="headerlink" href="#protein-domains-functional-features" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="http://www.uniprot.org">UniProt/SwissProt KnowledgeBase</a> -
resource on protein sequence and functional information (2020_05,
October 2020)</p></li>
<li><p><a class="reference external" href="http://pfam.xfam.org">Pfam</a> - database of protein families and
domains (v33, May 2020)</p></li>
</ul>
</div>
<div class="section" id="knowledge-resources-on-gene-and-protein-targets">
<h2>Knowledge resources on gene and protein targets<a class="headerlink" href="#knowledge-resources-on-gene-and-protein-targets" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="https://zenodo.org/record/3525385#.XcHblUVKiL4">CancerMine</a> -
Literature-mined database of tumor suppressor genes/proto-oncogenes
(v30, November 2020)</p></li>
<li><p><a class="reference external" href="https://www.targetvalidation.org/">Open Targets Platform</a> -
Database on disease-target associations, molecularly targeted drugs
and tractability aggregated from multiple sources (literature,
pathways, mutations) (2020_09)</p></li>
<li><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/29625053">TCGA driver genes</a>
- predicted cancer driver genes based on application of multiple
driver gene prediction tools on TCGA pan-cancer cohort</p></li>
</ul>
</div>
<div class="section" id="pathway-databases">
<h2>Pathway databases<a class="headerlink" href="#pathway-databases" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="http://www.genome.jp/kegg/pathway.htm">KEGG PATHWAY Database</a> -
November 20th 2020</p></li>
<li><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/29625050">Oncogenic Signaling Pathways -
TCGA</a> - Sanchez-Vega
et al., <em>Cell</em>, 2018</p></li>
</ul>
</div>
<div class="section" id="notes-on-variant-annotation-datasets">
<h2>Notes on variant annotation datasets<a class="headerlink" href="#notes-on-variant-annotation-datasets" title="Permalink to this headline">¶</a></h2>
<div class="section" id="genome-mapping">
<h3>Genome mapping<a class="headerlink" href="#genome-mapping" title="Permalink to this headline">¶</a></h3>
<p>A requirement for PCGR variant annotation datasets is that variants have
been mapped unambiguously to the reference human genome. For most
datasets this requirement is not an issue (i.e. dbSNP, ClinVar etc.). A
fraction of variants in the annotation datasets related to clinical
interpretation, CIViC and CBMDB, has however not been mapped to the
genome. Whenever possible, we have utilized
<a class="reference external" href="http://bioinformatics.mdanderson.org/transvarweb/">TransVar</a> to
identify the actual genomic variants (e.g. <em>g.chr7:140453136A&gt;T</em>) that
correspond to variants reported at the amino acid level or with other
HGVS nomenclature (e.g. <em>p.V600E</em>).</p>
<p>For variants that have been mapped to a specific build (GRCh37/GRCh38),
we have utilized the <a class="reference external" href="http://crossmap.sourceforge.net/">crossmap</a>
package to lift the datasets to the other build.</p>
</div>
<div class="section" id="data-quality">
<h3>Data quality<a class="headerlink" href="#data-quality" title="Permalink to this headline">¶</a></h3>
<p><strong>Clinical biomarkers</strong></p>
<p>Clinical biomarkers included in PCGR are limited to the following:</p>
<ul class="simple">
<li><p>Evidence items for specific markers in CIViC must be <em>accepted</em>
(<em>submitted</em> evidence items are not considered)</p></li>
<li><p>Markers reported at the variant level (e.g. <strong>BRAF p.V600E</strong>)</p></li>
<li><p>Markers reported at the codon level (e.g. <strong>KRAS p.G12</strong>)</p></li>
<li><p>Markers reported at the exon level (e.g. <strong>KIT exon 11 mutation</strong>)</p></li>
<li><p>Within the <a class="reference external" href="https://www.cancergenomeinterpreter.org/biomarkers">Cancer bioMarkers database
(CGI)</a>, only
markers collected from FDA/NCCN guidelines, scientific literature,
and clinical trials are included (markers collected from conference
abstracts etc. are not included)</p></li>
<li><p>Copy number gains/losses</p></li>
</ul>
<p>See also comment on a <a class="reference external" href="https://github.com/sigven/pcgr/issues/37#issuecomment-391966286">closed GitHib
issue</a></p>
<p><strong>IMPORTANT NOTE</strong>: The variant consequence reported by CIViC may
deviate from what is reported by PCGR. PCGR picks the variant
consequence according to VEP’s <em>pick</em> option (depending on a ranked list
of criteria that can be configured by the user), and this particular
transcript consequence may differ from what has been reported in the
literature.</p>
<p><strong>Molecularly targeted drugs</strong></p>
<ul class="simple">
<li><p>For targeted drugs extracted from <a class="reference external" href="https://www.targetvalidation.org">Open Targets
Platform</a>, we distinguish
between drugs in late clinical development (phase 3-4), versus those
in early clinical development (phase 1-2).</p></li>
</ul>
<p><strong>Gene-disease associations</strong></p>
<ul class="simple">
<li><p>Cancer phenotype associations retrieved from the <a class="reference external" href="https://www.targetvalidation.org/">Open Targets
Platform</a> are largely based on
the <a class="reference external" href="https://docs.targetvalidation.org/getting-started/scoring">association
score</a>
developed by the Open Targets Platform, with a couple of extra
post-processing steps:</p>
<ul>
<li><p>Phenotype associations in Open Targets Platform are assembled from
<a class="reference external" href="https://docs.targetvalidation.org/data-sources/data-sources">a variety of different data
sources</a>.
Target-disease associations included in PCGR must be supported by
<strong>at least two distinct sources</strong></p></li>
<li><p>The weakest associations, here defined as those with an
association score &lt; 0.4 (scale from 0 to 1), are ommitted</p></li>
<li><p>As is done within the Open Targets Platform, association scores
(for genes) are represented with varying shades of blue: the
darker the blue, the stronger the association. Variant hits in
tier 3/4 and the noncoding section are arranged according to this
association score. If several disease subtypes are associated with
a gene, the maximum association score is chosen.</p></li>
</ul>
</li>
</ul>
<p><strong>Tumor suppressor genes/proto-oncogenes</strong></p>
<ul class="simple">
<li><p>Status as oncogenes and/or tumor suppressors genes are done according
to the following scheme in PCGR:</p>
<ul>
<li><p>Five or more publications in the biomedical literature that
suggests an oncogenic/tumor suppressor role for a given gene (as
collected from the <a class="reference external" href="http://bionlp.bcgsc.ca/cancermine/">CancerMine text-mining
resource</a>), <strong>OR</strong></p></li>
<li><p>At least two publications from CancerMine that suggests an
oncogenic/tumor suppressor role for a given gene <strong>AND</strong> an
existing record for the same gene as a tumor suppressor/oncogene
in the <a class="reference external" href="http://ncg.kcl.ac.uk/">Network of Cancer Genes (NCG)</a></p></li>
<li><p>Status as oncogene is ignored if a given gene has three times as
much (literature evidence) support for a role as a tumor
suppressor gene (and vice versa)</p></li>
<li><p>Oncogenes/tumor suppressor candidates from CancerMine/NCG that are
found in the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/30096302">curated list of false positive cancer drivers
compiled by Bailey et al. (Cell,
2018)</a> have been
excluded</p></li>
</ul>
</li>
</ul>
<p><strong>TCGA somatic calls</strong></p>
<ul class="simple">
<li><p>TCGA employs four different variant callers for detection of somatic
variants (SNVs/InDels): <em>mutect2, varscan2, somaticsniper and muse</em>.
In the TCGA dataset bundled with PCGR, somatic SNVs are restricted to
those that are detected by at least two independent callers
(i.e. calls found by a single algorithm are considered low-confident
and disregarded)</p></li>
</ul>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="input.html" class="btn btn-neutral float-right" title="Input" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="FAQ.html" class="btn btn-neutral float-left" title="FAQ" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017-2020, Sigve Nakken

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>