<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pcgrr">
<title>Annotation resources â€¢ PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotation resources">
<meta property="og:description" content="pcgrr">
<meta property="og:image" content="https://sigven.github.io/pcgr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Intro</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/running.html">Running</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/input.html">Input files</a>
    <a class="dropdown-item" href="../articles/output.html">Output files</a>
    <a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a>
    <a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a>
    <a class="dropdown-item" href="../articles/developers.html">Developer notes</a>
    <a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/faq.html">FAQ</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Annotation resources</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/HEAD/../vignettes/annotation_resources.Rmd" class="external-link"><code>../vignettes/annotation_resources.Rmd</code></a></small>
      <div class="d-none name"><code>annotation_resources.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="basic-variant-consequence-annotation">Basic variant consequence annotation<a class="anchor" aria-label="anchor" href="#basic-variant-consequence-annotation"></a>
</h3>
<ul>
<li>
<a href="http://www.ensembl.org/info/docs/tools/vep/index.html" class="external-link">VEP</a> -
Variant Effect Predictor release 105 (<a href="https://www.gencodegenes.org/human/" class="external-link">GENCODE v39</a> as gene
reference database (v19 for grch37))</li>
</ul>
</div>
<div class="section level3">
<h3 id="insilico-predictions-of-effect-of-coding-variants">
<em>Insilico</em> predictions of effect of coding variants<a class="anchor" aria-label="anchor" href="#insilico-predictions-of-effect-of-coding-variants"></a>
</h3>
<ul>
<li>
<a href="https://sites.google.com/site/jpopgen/dbNSFP" class="external-link">dBNSFP</a> -
database of non-synonymous functional predictions (v4.2, March
2021)</li>
</ul>
</div>
<div class="section level3">
<h3 id="variant-frequency-databases">Variant frequency databases<a class="anchor" aria-label="anchor" href="#variant-frequency-databases"></a>
</h3>
<ul>
<li>
<a href="http://exac.broadinstitute.org/" class="external-link">gnomAD</a> - germline
variant frequencies exome-wide (r2.1, October 2018)</li>
<li>
<a href="http://www.ncbi.nlm.nih.gov/SNP/" class="external-link">dbSNP</a> - database of
short genetic variants (build 154)</li>
<li>
<a href="http://cancerhotspots.org" class="external-link">Cancer Hotspots</a> - a resource
for statistically significant mutations in cancer (v2, 2017)</li>
<li>
<a href="https://portal.gdc.cancer.gov/" class="external-link">TCGA</a> - somatic
mutations discovered across 33 tumor type cohorts (release 31.0, October
2021)</li>
<li>
<a href="http://docs.icgc.org/pcawg/" class="external-link">ICGC-PCAWG</a> - ICGC
Pancancer Analysis of Whole Genomes - (release 28, March 17th,
2019)</li>
</ul>
</div>
<div class="section level3">
<h3 id="variant-databases-of-clinical-utility">Variant databases of clinical utility<a class="anchor" aria-label="anchor" href="#variant-databases-of-clinical-utility"></a>
</h3>
<ul>
<li>
<a href="http://www.ncbi.nlm.nih.gov/clinvar/" class="external-link">ClinVar</a> -
database of clinically related variants (February 2022)</li>
<li>
<a href="http://docm.genome.wustl.edu" class="external-link">DoCM</a> - database of
curated mutations (v3.2, April 2016)</li>
<li>
<a href="http://civic.genome.wustl.edu" class="external-link">CIViC</a> - clinical
interpretations of variants in cancer (February 1st 2022)</li>
<li>
<a href="http://www.cancergenomeinterpreter.org/biomarkers" class="external-link">CGI</a>
- Cancer Genome Interpreter Cancer Biomarkers Database (CGI) (January
17th 2018)</li>
<li>
<a href="https://www.ebi.ac.uk/chembl/" class="external-link">ChEMBL</a> - database of
drugs, drug-like small molecules and their targets (ChEMBL_28, February
2021)</li>
</ul>
</div>
<div class="section level3">
<h3 id="protein-domainsfunctional-features">Protein domains/functional features<a class="anchor" aria-label="anchor" href="#protein-domainsfunctional-features"></a>
</h3>
<ul>
<li>
<a href="http://www.uniprot.org" class="external-link">UniProt/SwissProt KnowledgeBase</a>
- resource on protein sequence and functional information (2021_04,
November 2021)</li>
<li>
<a href="http://pfam.xfam.org" class="external-link">Pfam</a> - database of protein
families and domains (v35.0, November 2021)</li>
</ul>
</div>
<div class="section level3">
<h3 id="knowledge-resources-on-gene-and-protein-targets">Knowledge resources on gene and protein targets<a class="anchor" aria-label="anchor" href="#knowledge-resources-on-gene-and-protein-targets"></a>
</h3>
<ul>
<li>
<a href="https://zenodo.org/record/4925789#.YNT8K5MzY7Q" class="external-link">CancerMine</a> -
Literature-mined database of tumor suppressor genes/proto-oncogenes
(v42, December 2021)</li>
<li>
<a href="https://www.targetvalidation.org/" class="external-link">Open Targets
Platform</a> - Database on disease-target associations, molecularly
targeted drugs and tractability aggregated from multiple sources
(literature, pathways, mutations) (2021.11)</li>
<li>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/29625053" class="external-link">TCGA driver
genes</a> - predicted cancer driver genes based on application of
multiple driver gene prediction tools on TCGA pan-cancer cohort</li>
</ul>
</div>
<div class="section level3">
<h3 id="pathway-databases">Pathway databases<a class="anchor" aria-label="anchor" href="#pathway-databases"></a>
</h3>
<ul>
<li>
<a href="http://www.genome.jp/kegg/pathway.htm" class="external-link">KEGG PATHWAY
Database</a> - December 21st 2021</li>
<li>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/29625050" class="external-link">Oncogenic
Signaling Pathways - TCGA</a> - Sanchez-Vega et al., <em>Cell</em>,
2018</li>
</ul>
</div>
<div class="section level3">
<h3 id="notes-on-variant-annotation-datasets">Notes on variant annotation datasets<a class="anchor" aria-label="anchor" href="#notes-on-variant-annotation-datasets"></a>
</h3>
<div class="section level4">
<h4 id="genome-mapping">Genome mapping<a class="anchor" aria-label="anchor" href="#genome-mapping"></a>
</h4>
<p>A requirement for PCGR variant annotation datasets is that variants
have been mapped unambiguously to the reference human genome. For most
datasets this requirement is not an issue (i.e.Â dbSNP, ClinVar etc.). A
fraction of variants in the annotation datasets related to clinical
interpretation, CIViC and CBMDB, has however not been mapped to the
genome. Whenever possible, we have utilized <a href="http://bioinformatics.mdanderson.org/transvarweb/" class="external-link">TransVar</a> to
identify the actual genomic variants
(e.g.Â <em>g.chr7:140453136A&gt;T</em>) that correspond to variants
reported at the amino acid level or with other HGVS nomenclature
(e.g.Â <em>p.V600E</em>).</p>
<p>For variants that have been mapped to a specific build
(GRCh37/GRCh38), we have utilized the <a href="http://crossmap.sourceforge.net/" class="external-link">crossmap</a> package to lift the
datasets to the other build.</p>
</div>
<div class="section level4">
<h4 id="data-quality">Data quality<a class="anchor" aria-label="anchor" href="#data-quality"></a>
</h4>
<p><strong>Clinical biomarkers</strong></p>
<p>Clinical biomarkers included in PCGR are limited to the
following:</p>
<ul>
<li>Evidence items for specific markers in CIViC must be
<em>accepted</em> (<em>submitted</em> evidence items are not
considered)</li>
<li>Markers reported at the variant level (e.g.Â <strong>BRAF
p.V600E</strong>)</li>
<li>Markers reported at the codon level (e.g.Â <strong>KRAS
p.G12</strong>)</li>
<li>Markers reported at the exon level (e.g.Â <strong>KIT exon 11
mutation</strong>)</li>
<li>Within the <a href="https://www.cancergenomeinterpreter.org/biomarkers" class="external-link">Cancer
bioMarkers database (CGI)</a>, only markers collected from FDA/NCCN
guidelines, scientific literature, and clinical trials are included
(markers collected from conference abstracts etc. are not included)</li>
<li>Copy number gains/losses</li>
</ul>
<p>See also comment on a <a href="https://github.com/sigven/pcgr/issues/37#issuecomment-391966286" class="external-link">closed
GitHib issue</a></p>
<p><strong>IMPORTANT NOTE</strong>: The variant consequence reported by
CIViC may deviate from what is reported by PCGR. PCGR picks the variant
consequence according to VEPâ€™s <em>pick</em> option (depending on a
ranked list of criteria that can be configured by the user), and this
particular transcript consequence may differ from what has been reported
in the literature.</p>
<p><strong>Molecularly targeted drugs</strong></p>
<ul>
<li>For targeted drugs extracted from <a href="https://www.targetvalidation.org" class="external-link">Open Targets Platform</a>, we
distinguish between drugs in late clinical development (phase 3-4),
versus those in early clinical development (phase 1-2).</li>
</ul>
<p><strong>Gene-disease associations</strong></p>
<ul>
<li>Cancer phenotype associations retrieved from the <a href="https://www.targetvalidation.org/" class="external-link">Open Targets Platform</a> are
largely based on the <a href="https://docs.targetvalidation.org/getting-started/scoring" class="external-link">association
score</a> developed by the Open Targets Platform, with a couple of extra
post-processing steps:
<ul>
<li>Phenotype associations in Open Targets Platform are assembled from
<a href="https://docs.targetvalidation.org/data-sources/data-sources" class="external-link">a
variety of different data sources</a>. Target-disease associations
included in PCGR must be supported by <strong>at least two distinct
sources</strong>
</li>
<li>The weakest associations, here defined as those with an association
score &lt; 0.4 (scale from 0 to 1), are ommitted</li>
<li>As is done within the Open Targets Platform, association scores (for
genes) are represented with varying shades of blue: the darker the blue,
the stronger the association. Variant hits in tier 3/4 and the noncoding
section are arranged according to this association score. If several
disease subtypes are associated with a gene, the maximum association
score is chosen.</li>
</ul>
</li>
</ul>
<p><strong>Tumor suppressor genes/proto-oncogenes</strong></p>
<ul>
<li>Status as oncogenes and/or tumor suppressors genes are done
according to the following scheme in PCGR:
<ul>
<li>Five or more publications in the biomedical literature that suggests
an oncogenic/tumor suppressor role for a given gene (as collected from
the <a href="http://bionlp.bcgsc.ca/cancermine/" class="external-link">CancerMine text-mining
resource</a>), <strong>OR</strong>
</li>
<li>At least two publications from CancerMine that suggests an
oncogenic/tumor suppressor role for a given gene <strong>AND</strong> an
existing record for the same gene as a tumor suppressor/oncogene in the
<a href="http://ncg.kcl.ac.uk/" class="external-link">Network of Cancer Genes (NCG)</a>
</li>
<li>Status as oncogene is ignored if a given gene has three times as
much (literature evidence) support for a role as a tumor suppressor gene
(and vice versa)</li>
<li>Oncogenes/tumor suppressor candidates from CancerMine/NCG that are
found in the <a href="https://www.ncbi.nlm.nih.gov/pubmed/30096302" class="external-link">curated list of
false positive cancer drivers compiled by Bailey et al.Â (Cell, 2018)</a>
have been excluded</li>
</ul>
</li>
</ul>
<p><strong>TCGA somatic calls</strong></p>
<ul>
<li>TCGA employs four different variant callers for detection of somatic
variants (SNVs/InDels): <em>mutect2, varscan2, somaticsniper and
muse</em>. In the TCGA dataset bundled with PCGR, somatic SNVs are
restricted to those that are detected by at least two independent
callers (i.e.Â calls found by a single algorithm are considered
low-confident and disregarded)</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
