<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Changelog • PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-7HRLE8GQRE"></script><script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){dataLayer.push(arguments);}
   gtag('js', new Date());

   gtag('config', 'G-7HRLE8GQRE');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/running.html">Running</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/input.html">Input files</a></li>
    <li><a class="dropdown-item" href="../articles/output.html">Output files</a></li>
    <li><a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a></li>
    <li><a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="../articles/developers.html">Developer notes</a></li>
    <li><a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/main/vignettes/CHANGELOG.Rmd" class="external-link"><code>vignettes/CHANGELOG.Rmd</code></a></small>
      <div class="d-none name"><code>CHANGELOG.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="v2-2-4">v2.2.4<a class="anchor" aria-label="anchor" href="#v2-2-4"></a>
</h2>
<ul>
<li>Date: <strong>2025-09-08</strong>
<ul>
<li>fixed bug in site to TCGA cohort matching
(<a href="https://github.com/sigven/pcgr/issues/275" class="external-link">issue275</a>)</li>
<li>fixed bug in copy number segment annotations with no protein-coding
overlap
(<a href="https://github.com/sigven/pcgr/issues/278" class="external-link">issue278</a>)</li>
<li>added gene <code>loss</code> biomarkers (loss-of-function) as
records to match for CNA homozygous deletions
(<a href="https://github.com/sigven/pcgr/issues/277" class="external-link">issue277</a>)</li>
<li>added new option <code>--sex</code> to enable sex-adjusted
annotation of CNA segments (i.e. hemizygous deletions for males on
chrX)</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-2-3">v2.2.3<a class="anchor" aria-label="anchor" href="#v2-2-3"></a>
</h2>
<ul>
<li>Date: <strong>2025-07-18</strong>
<ul>
<li>update Ensembl VEP <code>v113.4</code> (see <a href="https://github.com/bioconda/bioconda-recipes/pull/54330" class="external-link">bioconda
issue</a>)</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-2-2">v2.2.2<a class="anchor" aria-label="anchor" href="#v2-2-2"></a>
</h2>
<ul>
<li>Date: <strong>2025-07-15</strong>
<ul>
<li>fix bug in Grantham distance score evaluation
(<a href="https://github.com/sigven/pcgr/issues/271" class="external-link">issue271</a>)</li>
<li>add support for heterozygous deletions in CNA interpretation -
variant classes are now <em>homdel</em>, <em>hetdel</em>, and
<em>gain</em>
</li>
<li>fix missing pagination of some data tables with actionable
markers</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-2-1">v2.2.1<a class="anchor" aria-label="anchor" href="#v2-2-1"></a>
</h2>
<ul>
<li>Date: <strong>2025-03-23</strong>
<ul>
<li>fix bug in CPSR for ClinVar variants with non-standard significance
levels</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-2-0">v2.2.0<a class="anchor" aria-label="anchor" href="#v2-2-0"></a>
</h2>
<ul>
<li>Date: <strong>2025-03-22</strong>
</li>
<li>Major data updates
<ul>
<li>ClinVar (2025-03)</li>
<li>dbNSFP (v5.0)</li>
<li>CIViC (2025-03-13)</li>
<li>GENCODE v47 (VEP v113)</li>
<li>PanelApp (2025-02)</li>
<li>UniProt KB (2025-01)</li>
<li>Cancer Gene Census (v101)</li>
</ul>
</li>
<li>Cleaned and improved oncogenicity classification of variants
<ul>
<li>Added Grantham distance for amino acid alterations - taken into
account</li>
<li>only consider LP variant classification when oncogenicity score = 4
and if at least three oncogenic criteria matches <em>or</em>
<code>ONCG_OS1</code> is matching for oncogenicity classification</li>
</ul>
</li>
<li>Multiple cosmetic changes to HTML report - e.g. collapsed
call-outs</li>
<li>Added oncogenicity documentation in variant tables of HTML report
(indicates which criteria that was matched for a given variant)</li>
<li>Removed <code>VEP_ALL_CSQ</code> from variant tables in HTML report
to reduce file size - still available in TSV output</li>
<li>Added MutSpliceDB splice site effects (used e.g. for
loss-of-function annotation)</li>
<li>Added MaxEntScan plugin in VEP for splice site disruption prediction
(used e.g. for loss-of-function annotation)</li>
<li>Added more prediction algorithms from dbNSFP</li>
<li>Fixed bug that caused crash for missing values in DP/AF values of
input VCF</li>
<li>Fixed erroneous re-formatting of MAF integer columns to floats</li>
<li>Fixed cases for which similar Entrez gene identifiers mapped to
different gene names</li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-1-2">v2.1.2<a class="anchor" aria-label="anchor" href="#v2-1-2"></a>
</h2>
<ul>
<li><p>Date: <strong>2024-10-21</strong></p></li>
<li><p>Highlight also VUS variants (ClinVar-classified) in the variant
oncogenicity section of the HTML report</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-1-1">v2.1.1<a class="anchor" aria-label="anchor" href="#v2-1-1"></a>
</h2>
<ul>
<li><p>Date: <strong>2024-10-11</strong></p></li>
<li><p>Fix bug in parsing of relative CDS position
(<a href="https://github.com/sigven/pcgr/issues/252" class="external-link">issue252</a>)</p></li>
<li><p>Move some code in the qmd templates to <code>pcgrr</code>
functions</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-1-0">v2.1.0<a class="anchor" aria-label="anchor" href="#v2-1-0"></a>
</h2>
<ul>
<li>Date: <strong>2024-09-27</strong>
</li>
<li>Major data updates
<ul>
<li>ClinVar (2024-09)</li>
<li>dbNSFP (v4.8)</li>
<li>NCI Thesaurus <code>v24.07e</code>
</li>
<li>CIViC (2024-09-18)</li>
</ul>
</li>
<li>Reduction (~15%) in overall data bundle size - removed unused data
files (e.g. expression counts)</li>
<li>Fixed bug in MAF output for tumor-only runs
<a href="https://github.com/sigven/pcgr/issues/250" class="external-link">issue250</a>, also
ensure that non-exonic variants are excluded if setting
<code>--exclude_nonexonic</code> is used)</li>
<li>Fixed bug in annotation of splice site mutation hotspots (e.g. MET
exon 14 skipping)</li>
<li>Highlighted variants with known pathogenic/likely pathogenic
clinical significance in ClinVar (regardless of phenotype and variant
origin) in the variant oncogenicity section of the HTML report</li>
<li>Created interactive visualization support for allele-specific copy
number data (HTML report)</li>
<li>Slight change to the default transcript consequence pick order in
VEP based on observations of prioritized transcripts (<em>mane_select
&gt; mane_plus_clinical &gt; canonical &gt; biotype &gt; ccds &gt; rank
&gt; tsl &gt; appris &gt;length</em>)</li>
<li>Pulled in known oncogenic variants from ClinVar (assessed through
ClinGen/CGC/VICC SOP, oncogenic/likely oncogenic) into the variant
oncogenicity assessment algorithm</li>
<li>Added option <code>--no_html</code> to disable HTML report
generation</li>
<li>Added option <code>--input_cpsr</code> - re-offering the possibility
to integrate CPSR-classified germline variants in the PCGR HTML
report</li>
<li>Added <code>HGVSc_RefSeq</code> as output column in TSV/HTML - using
MANE Select RefSeq transcript identifiers (works primarily for
grch38)</li>
<li>Pulled in coding sequence start annotation for protein-coding
transcripts from GENCODE, enabling a more useful annotation of promoter
variants (e.g. TERT)</li>
<li>Created new column <code>ALTERATION</code> in variant tables of HTML
report, a combination of <code>HGVSp</code>, <code>HGVSc</code> (if
<code>HGVSp</code> not available)</li>
<li>New output file for tumor-only runs, the complete set of calls,
filtered and unfiltered, in a TSV file</li>
<li>Re-processed all RNA-seq reference cohorts (TCGA, DepMap,
TreeHouse), ensuring that all cohorts are using the same unit
(<code>log2(TPM+0.001)</code>)</li>
<li>Separated outlier gene expression results into separate tabs in the
HTML report, added them to Excel workbook output</li>
<li>Added section on kataegis events in the HTML report</li>
<li>Fixed bug in plotting of reference TMB distributions for different
TMB algorithms (<code>--tmb_display</code> option)</li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-0-3">v2.0.3<a class="anchor" aria-label="anchor" href="#v2-0-3"></a>
</h2>
<ul>
<li>Date: <strong>2024-08-01</strong>
</li>
<li>Ensure correct propagation of purity/ploidy in output report</li>
<li>Ensure that MAF output is properly filtered for tumor-only runs</li>
<li>Ensure properly copying of quarto templates (abandon
<code>file.copy</code>), both for PCGR and CPSR</li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-0-2">v2.0.2<a class="anchor" aria-label="anchor" href="#v2-0-2"></a>
</h2>
<ul>
<li>Date: <strong>2024-07-16</strong>
</li>
<li>Ensure correct reference to variant actionability guidelines -
AMP/ASCO/CAP (<em>not</em> ACMG/AMP), both in code and in docs (thanks
to <a href="https://github.com/HomoPolyethylen" class="external-link">HomoPolyethylen</a> for
pointing this out)</li>
<li>fix bug in missing assignment of tier 3 variants (AMP/ASCO/CAP)</li>
<li>ensure non-exonic biomarker variants (e.g. TERT) are written to
Excel sheet</li>
<li>specify (value boxes, plots) that MSI classification is based on
coding variants</li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-0-1">v2.0.1<a class="anchor" aria-label="anchor" href="#v2-0-1"></a>
</h2>
<ul>
<li>Date: <strong>2024-07-07</strong>
</li>
<li>Fixed bug for chrM variants in input - not properly annotated by
VEP, and not correctly processed in <code>pcgrr</code>. Any
mitochondrial variants found in input VCF are now removed during VCF
pre-processing.</li>
</ul>
</div>
<div class="section level2">
<h2 id="v2-0-0">v2.0.0<a class="anchor" aria-label="anchor" href="#v2-0-0"></a>
</h2>
<ul>
<li>Date: <strong>2024-06-26</strong>
</li>
<li>Major data updates
<ul>
<li>ClinVar (2024-06)</li>
<li>NCI Thesaurus <code>v24.05d</code>
</li>
<li>Open Targets Platform <code>v2024.06</code>
</li>
<li>CIViC (2024-06-21)</li>
<li>CGI Cancer Biomarkers database (2022/10/17)</li>
<li>GENCODE <code>v46/v19</code> (GRCh38/GRCh37)</li>
<li>Cancer Gene Census</li>
<li>CancerMine <code>v50</code> (2023-03)</li>
<li>Pfam <code>v35.0</code> (November 2021)</li>
<li>Disease Ontology/EFO</li>
<li>UniProt KB <code>v2024_03</code>
</li>
</ul>
</li>
<li>Major software updates
<ul>
<li>Ensembl VEP <code>v112</code>
</li>
</ul>
</li>
<li><a href="https://github.com/sigven/pcgr/compare/v1.4.1...v2.0.0" class="external-link">Diff between v2.0.0 and v1.4.1</a></li>
</ul>
<div class="section level3">
<h3 id="addedchanged">Added/changed<a class="anchor" aria-label="anchor" href="#addedchanged"></a>
</h3>
<ul>
<li>New report generation framework - <a href="https://quarto.org" class="external-link">quarto</a>
<ul>
<li>multiple options related to RMarkdown output are now deprecated</li>
</ul>
</li>
<li>Re-organized data bundle structure
<ul>
<li>Users need to download an assembly-specific VEP cache separately
from the Ensembl VEP website, and provide its path to the new required
argument <code>--vep_dir</code> in the <code>pcgr</code> command</li>
</ul>
</li>
<li>Re-engineered data bundle generation pipeline</li>
<li>Improved data bundle documentation
<ul>
<li>An HTML report with an overview of the contents of the data bundle
is shipped with the reference data itself, also available <a href="https://rpubs.com/sigven/pcgr_refdata" class="external-link">here (grch38)</a>.</li>
</ul>
</li>
<li>Singularity/Apptainer support</li>
<li>Moved more of the code base to initial Python workflow steps
(biomarker matching, CNA segment annotation, RNA expression analysis,
oncogenicity classification)</li>
<li>Variants are now classified with respect to both oncogenicity and
actionability, and the previous global tier classification (tier 1-5) is
thus deprecated</li>
<li>New copy number input format - allele-specific (chrom, start, end,
n_major, n_minor)
<ul>
<li>New argument <code>n_copy_gain</code> - Minimum number of total copy
number for segments considered as gains/amplifications (default: 6)</li>
</ul>
</li>
<li>RNA-bulk expression input permitted in the <code>pcgr</code> command
<ul>
<li>
<code>--input_rna_expression</code> - accepts a TSV file with gene
expression values</li>
<li>
<code>--expression_sim</code> - boolean flag to enable expression
similarity analysis</li>
<li>
<code>--expression_sim_db</code> - Comma-separated string of
databases for used in RNA expression similarity analysis, default:
tcga,depmap,treehouse</li>
</ul>
</li>
<li>TMB calculations can be adjusted using several parameters:
<ul>
<li>
<code>--tmb_display</code> - Type of TMB measure to show in report
(coding_and_silent, coding_non_silent, missense_only)</li>
<li>
<code>--tmb_dp_min</code> - Minimum depth for a position to be
considered for TMB calculation (default: 0) - requires allelic support
information from VCF</li>
<li>
<code>--tmb_af_min</code> - Minimum allele frequency for a position
to be considered for TMB calculation (default: 0) - requires allelic
support information from VCF</li>
</ul>
</li>
<li>A multi-sheet Excel workbook output with analysis output is
provided, suitable e.g. for aggregation of results across samples</li>
<li>argument name changes to <code>pcgr</code>:
<ul>
<li>
<code>--pcgr_dir</code> renamed to <code>--refdata_dir</code>
</li>
<li>
<code>--clinvar_ignore_noncancer</code> renamed to
<code>--clinvar_report_noncancer</code>, meaning that variants found in
ClinVar, yet attributed to <em>non-cancer related</em> phenotypes, are
now excluded from reporting by default</li>
<li>
<code>--vep_gencode_all</code> renamed to
<code>--vep_gencode_basic</code>, meaning that the gene variant
annotation is now using <em>all</em> GENCODE transcripts by default, not
only the <em>basic</em> set</li>
<li>
<code>--preserved_info_tags</code> renamed to
<code>--retained_info_tags</code>
</li>
<li>
<code>--basic</code> renamed to <code>--no_reporting</code>
</li>
<li>
<code>--target_size_mb</code> renamed to
<code>--effective_target_size_mb</code>
</li>
</ul>
</li>
<li>LOFTEE plugin in VEP removed as loss-of-function variant classifier
(due to low level of maintenance, and outdated dependency requirements).
For now, a simplified LoF-annotation is used as a replacement, looking
primarily at CSQ types (<code>stop_gained</code>,
<code>frameshift_variant</code>, <code>splice_acceptor_variant</code>,
<code>splice_donor_variant</code>). Furthermore, frameshift/stop-gain
variants that are found within the last 5% of the coding sequence length
are deemed non-LOF, as are splice donor variants not disrupting the
canonical site (GC&gt;GT). An even more advanced LoF-annotation is
planned for a future release.</li>
<li>Biomarkers are matched much more comprehensively than in previous
versions, matching at the genomic level, codon, exon, amino acid and
gene level (both principal and non-principal transcript matches)</li>
</ul>
</div>
<div class="section level3">
<h3 id="removed">Removed<a class="anchor" aria-label="anchor" href="#removed"></a>
</h3>
<ul>
<li>Options for configuring RMarkdown output,
i.e. <code>--report_theme</code>,
<code>report_nonfloating_toc</code>
</li>
<li>
<code>--cpsr_report</code> and <code>--include_trials</code>, which
can provide the report with associated pathogenic germline variants
(from CPSR) and potential clinical trial oppertunities is currenly on
hold for a forthcoming release</li>
<li>
<code>--no_vcf_validate</code> - VCF validation is simplified, not
relying on <em>vcf-validator</em> anymore</li>
<li>Options to filter tumor-only calls using 1000 Genomes Project
database, i.e. <code>--maf_onekg_eur</code>,
<code>--maf_onekg_amr</code>, <code>--maf_onekg_eas</code>,
<code>--maf_onekg_afr</code>, <code>--maf_onekg_sas</code>,
<code>--maf_onekg_global</code>
</li>
<li><code>--cell_line</code></li>
<li>
<code>--logr_gain</code>, and <code>--logr_homdel</code>
</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="v1-4-1">v1.4.1<a class="anchor" aria-label="anchor" href="#v1-4-1"></a>
</h2>
<ul>
<li>Date: <strong>2023-03-14</strong>
</li>
<li><a href="https://github.com/sigven/pcgr/compare/v1.4.0...v1.4.1" class="external-link">Diff between v1.4.1 and v1.4.0</a></li>
</ul>
<div class="section level3">
<h3 id="changes">Changes<a class="anchor" aria-label="anchor" href="#changes"></a>
</h3>
<ul>
<li>Skip favicon for Rmarkdown by
<a href="https://github.com/sigven" class="external-link">@sigven</a> in
<a href="https://github.com/sigven/pcgr/pull/210" class="external-link">pr210</a>
</li>
<li>Use biocondarised vcf2tsvpy by
<a href="https://github.com/pdiakumis" class="external-link">@pdiakumis</a> in
<a href="https://github.com/sigven/pcgr/pull/211" class="external-link">pr211</a>
</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="v1-4-0">v1.4.0<a class="anchor" aria-label="anchor" href="#v1-4-0"></a>
</h2>
<ul>
<li>Date: <strong>2023-03-08</strong>
</li>
<li><a href="https://github.com/sigven/pcgr/compare/v1.3.0...v1.4.0" class="external-link">Diff between v1.4.0 and v1.3.0</a></li>
</ul>
<div class="section level3">
<h3 id="changes-1">Changes<a class="anchor" aria-label="anchor" href="#changes-1"></a>
</h3>
<ul>
<li>Pick trans consequence patch by
<a href="https://github.com/sigven" class="external-link">@sigven</a> in
<a href="https://github.com/sigven/pcgr/pull/206" class="external-link">pr206</a>
</li>
<li>Use VEP vcf as input to vcf2maf.pl by
<a href="https://github.com/sigven" class="external-link">@sigven</a> in
<a href="https://github.com/sigven/pcgr/pull/207" class="external-link">pr207</a>
</li>
<li>Better handling of pcgr/pcgrr conda env custom naming by
<a href="https://github.com/pdiakumis" class="external-link">@pdiakumis</a> in
<a href="https://github.com/sigven/pcgr/pull/208" class="external-link">pr208</a>
</li>
<li>Update report style by
<a href="https://github.com/sigven" class="external-link">@sigven</a> in
<a href="https://github.com/sigven/pcgr/pull/209" class="external-link">pr209</a>
</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="v1-3-0">v1.3.0<a class="anchor" aria-label="anchor" href="#v1-3-0"></a>
</h2>
<ul>
<li>Date: <strong>2023-02-28</strong>
</li>
<li><a href="https://github.com/sigven/pcgr/compare/v1.2.0...v1.3.0" class="external-link">Diff between v1.3.0 and v1.2.0</a></li>
</ul>
<div class="section level3">
<h3 id="changes-2">Changes<a class="anchor" aria-label="anchor" href="#changes-2"></a>
</h3>
<ul>
<li>
<code>pcgr_summarise.py</code>: proritize protein-coding BIOTYPE csq
(<a href="https://github.com/sigven/pcgr/pull/201" class="external-link">pr201</a>)</li>
<li>
<code>cpsr.py</code>: expose <code>--pcgrr_conda</code> option to
flexibly activate pcgrr env by a non-default pcgrr name</li>
<li>docs: update <code>input.Rmd</code>, <code>running.Rmd</code>
</li>
<li>
<code>cpsr_validate_input.py</code>: refactor for efficient custom
gene egrep</li>
<li>code reformat via autopep8 for <code>annoutils.py</code>,
<code>pcgr_vcfanno.py</code>
</li>
<li>GitHub Actions:
<ul>
<li>bump docker actions <code>setup-buildx-action</code> (v1–v2),
<code>build-push-action</code> (v2–v4)</li>
<li>use <code>miniforge-variant</code> instead of
<code>mamba-version: "*"</code>
</li>
<li>replace <code>::set-output</code> since deprecated</li>
</ul>
</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="v1-2-0">v1.2.0<a class="anchor" aria-label="anchor" href="#v1-2-0"></a>
</h2>
<ul>
<li>Date: <strong>2022-11-11</strong>
</li>
<li><a href="https://github.com/sigven/pcgr/compare/v1.1.0...v1.2.0" class="external-link">Diff between v1.2.0 and v1.1.0</a></li>
</ul>
<div class="section level3">
<h3 id="changes-3">Changes<a class="anchor" aria-label="anchor" href="#changes-3"></a>
</h3>
<ul>
<li>Keep only autosomal, X, Y, M/MT chromosomes</li>
<li>Import bcftools as dependency</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="v1-1-0">v1.1.0<a class="anchor" aria-label="anchor" href="#v1-1-0"></a>
</h2>
<ul>
<li>Date: <strong>2022-10-28</strong>
</li>
<li><a href="https://github.com/sigven/pcgr/compare/v1.0.3...v1.1.0" class="external-link">Diff between v1.1.0 and v1.0.3</a></li>
</ul>
<div class="section level3">
<h3 id="changes-4">Changes<a class="anchor" aria-label="anchor" href="#changes-4"></a>
</h3>
<ul>
<li>Remove Docker command wrappers (note: this does not remove the
Docker <em>functionality</em> from PCGR; instead it removes the legacy
wrappers that were created in the original PCGR version). This along
with a lot of other general changes are summarised in
<a href="https://github.com/sigven/pcgr/pull/193" class="external-link">pr193</a>. Of note:
<ul>
<li>
<code>--no_docker</code> and <code>--docker_uid</code> CLI arguments
are now obsolete.</li>
<li>
<code>--version</code> CLI argument added for
<code>pcgr/cpsr.py</code>
</li>
<li>declutter repetitive log messages</li>
<li>refactor <code>pcgr/cpsr.py</code> script</li>
</ul>
</li>
<li>Update documentation and declutter logging; refactor dict creation
(<a href="https://github.com/sigven/pcgr/pull/192" class="external-link">pr192</a>).</li>
<li>Minor refactor
(<a href="https://github.com/sigven/pcgr/pull/194" class="external-link">pr194</a>):
<ul>
<li>switch to using Python’s native <code>os.remove</code> and
<code>os.rename</code> for glob cleanup</li>
<li>keep decompressed VCF only if <code>--vcf2maf</code> option is
specified. The <a href="https://github.com/mskcc/vcf2maf" class="external-link">vcf2maf</a>
tool does not support compressed VCFs - see <a href="https://github.com/mskcc/vcf2maf/issues/235" class="external-link">issue235</a>.</li>
</ul>
</li>
<li>Fix for CLI argument <code>--cna_overlap_pct</code>
<a href="https://github.com/sigven/pcgr/pull/196" class="external-link">pr196</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="new-contributors">New Contributors<a class="anchor" aria-label="anchor" href="#new-contributors"></a>
</h3>
<ul>
<li>
<a href="https://github.com/niklasmueboe" class="external-link">@niklasmueboe</a>
(<a href="https://github.com/sigven/pcgr/pull/196" class="external-link">pr196</a>).</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="v1-0-3">v1.0.3<a class="anchor" aria-label="anchor" href="#v1-0-3"></a>
</h2>
<ul>
<li>Date: <strong>2022-05-24</strong>
</li>
</ul>
<div class="section level5">
<h5 id="fixed">Fixed<a class="anchor" aria-label="anchor" href="#fixed"></a>
</h5>
<ul>
<li>Bug in clinical trials sorting,
<a href="https://github.com/sigven/pcgr/pull/191" class="external-link">pr191</a>
</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="v1-0-2">v1.0.2<a class="anchor" aria-label="anchor" href="#v1-0-2"></a>
</h2>
<ul>
<li>Date: <strong>2022-03-30</strong>
</li>
</ul>
<div class="section level5">
<h5 id="fixed-1">Fixed<a class="anchor" aria-label="anchor" href="#fixed-1"></a>
</h5>
<ul>
<li>JSON output for CPSR,
<a href="https://github.com/sigven/pcgr/issues/44" class="external-link">issue44</a>
</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="v1-0-1">v1.0.1<a class="anchor" aria-label="anchor" href="#v1-0-1"></a>
</h2>
<ul>
<li>Date: <strong>2022-03-09</strong>
</li>
</ul>
<div class="section level5">
<h5 id="fixed-2">Fixed<a class="anchor" aria-label="anchor" href="#fixed-2"></a>
</h5>
<ul>
<li>Writing to JSON crashes when size of input VCF is huge (variants in
the order of millions). If raw input set (VCF) contains &gt; 500,000
variants, this set will, prior to reporting, be reduced by
<ul>
<li><ol style="list-style-type: upper-alpha">
<li>exclusion of intergenic and intronic variants, and</li>
</ol></li>
<li><ol start="2" style="list-style-type: upper-alpha">
<li>exclusion of upstream_gene/downstream_gene variants (if variant set
is still above 500,000 after step A)</li>
</ol></li>
</ul>
</li>
<li>Bug in signature analysis
(<a href="https://github.com/sigven/pcgr/issues/187" class="external-link">issue187</a>) for
cases where the input variant set fits to &gt; 18 different
aetiologies.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v1-0-0">v1.0.0<a class="anchor" aria-label="anchor" href="#v1-0-0"></a>
</h2>
<ul>
<li><p>Date: <strong>2022-02-25</strong></p></li>
<li><p>Data updates: ClinVar, GWAS catalog, GENCODE, CIViC, CancerMine,
KEGG, ChEMBL, Open Targets Platform, Disease Ontology, Experimental
Factor Ontology</p></li>
</ul>
<div class="section level5">
<h5 id="added">Added<a class="anchor" aria-label="anchor" href="#added"></a>
</h5>
<ul>
<li>Command-line options
<ul>
<li>VEP options
<ul>
<li>
<code>--vep_gencode_all</code> - use all GENCODE transcripts during
VEP annotation (not only the basic GENCODE set)</li>
<li>
<code>--prevalence_reference_signatures</code> - set minimum
prevalence (percent) for selection of reference signatures included in
refitting procedure for a given tumor type</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="changed">Changed<a class="anchor" aria-label="anchor" href="#changed"></a>
</h5>
<ul>
<li>Complete restructure of Python and R components.Installation now
relies on two separate <a href="https://docs.conda.io/en/latest/" class="external-link">conda</a> packages,
<code>pcgr</code> (Python component) and <code>pcgrr</code> (R
component). Direct Docker support remains, with the Dockerfile
simplified to rely exclusively on the installation of the above Conda
packages.</li>
</ul>
</div>
<div class="section level5">
<h5 id="removed-1">Removed<a class="anchor" aria-label="anchor" href="#removed-1"></a>
</h5>
<ul>
<li>VCF validation step. Feedback from users suggested that Ensembl’s
<code>vcf-validator</code> was often too stringent so its use has been
deprecated. The <code>--no_vcf_validate</code> option remains for
backwards compatibility.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-9-2">v0.9.2<a class="anchor" aria-label="anchor" href="#v0-9-2"></a>
</h2>
<ul>
<li><p>Date: <strong>2021-06-30</strong></p></li>
<li><p>Data updates: ClinVar, GWAS catalog, CIViC, CancerMine, dbNSFP,
KEGG, ChEMBL, Disease Ontology/EFO, Open Targets Platform, UniProt KB,
GENCODE</p></li>
<li><p>Software upgrades: R v4.1, Bioconductor v3.13, VEP (104)
++</p></li>
</ul>
<div class="section level5">
<h5 id="changed-1">Changed<a class="anchor" aria-label="anchor" href="#changed-1"></a>
</h5>
<ul>
<li>TOML-based configuration for PCGR is abandoned, all options to PCGR
are now configured through command-line parameters
<ul>
<li>
<strong>NOTE</strong>: We recommend to turn on
<code>--show_noncoding</code> and <code>--vcf2maf</code> (prevously
turned on by default in TOML). For tumor-only runs, we recommend to
include <code>--exclude_dbsnp_nonsomatic</code> and
<code>exclude_nonexonic</code>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="added-1">Added<a class="anchor" aria-label="anchor" href="#added-1"></a>
</h5>
<ul>
<li>Command-line options
<ul>
<li>Previously set in TOML file)
<ul>
<li>Allelic support
<ul>
<li><code>--tumor_dp_tag</code></li>
<li><code>--tumor_af_tag</code></li>
<li><code>--control_dp_tag</code></li>
<li><code>--control_af_tag</code></li>
<li><code>--call_conf_tag</code></li>
</ul>
</li>
<li>Tumor-only options
<ul>
<li><code>--maf_onekg_eur</code></li>
<li><code>--maf_onekg_amr</code></li>
<li><code>--maf_onekg_afr</code></li>
<li><code>--maf_onekg_eas</code></li>
<li><code>--maf_onekg_sas</code></li>
<li><code>--maf_onekg_global</code></li>
<li><code>--maf_gnomad_nfe</code></li>
<li><code>--maf_gnomad_asj</code></li>
<li><code>--maf_gnomad_fin</code></li>
<li><code>--maf_gnomad_oth</code></li>
<li><code>--maf_gnomad_amr</code></li>
<li><code>--maf_gnomad_afr</code></li>
<li><code>--maf_gnomad_eas</code></li>
<li><code>--maf_gnomad_sas</code></li>
<li><code>--maf_gnomad_global</code></li>
<li><code>--exclude_pon</code></li>
<li><code>--exclude_likely_het_germline</code></li>
<li><code>--exclude_likely_hom_germline</code></li>
<li><code>--exclude_dbsnp_nonsomatic</code></li>
<li><code>--exclude_nonexonic</code></li>
</ul>
</li>
<li><code>--report_theme</code></li>
<li>
<code>--preserved_info_tags</code> (previously
<code>custom_tags (TOML)</code>)</li>
<li>
<code>--show_noncoding</code> (previously
<code>list_noncoding (TOML)</code>)</li>
<li>
<code>--vcfanno_n_proc</code> (previously
<code>n_vcfanno_proc (TOML)</code>)</li>
<li>
<code>--vep_n_forks</code> (previously
<code>n_vep_forks (TOML)</code>)</li>
<li><code>--vep_pick_order</code></li>
<li>
<code>--vep_no_intergenic</code> (previously
<code>vep_skip_intergenic (TOML)</code>)</li>
<li><code>--vcf2maf</code></li>
</ul>
</li>
<li>New options
<ul>
<li>
<code>--report_nonfloating_toc</code> (<strong>NEW</strong>) - add
the TOC at the top of the HTML report, not floating at the left of the
document</li>
<li>
<code>--cpsr_report</code> (<strong>NEW</strong>) - add a dedicated
section in PCGR with main germline findings from CPSR analysis - (use
the gzipped JSON output from CPSR as input)</li>
<li>
<code>--vep_regulatory</code> (<strong>NEW</strong>) - append
regulatory annotations to variants (TF binding sites etc.)</li>
<li>
<code>--include_artefact_signatures</code> (<strong>NEW</strong>) -
include sequencing artefacts in the reference collection of mutational
signatures (COSMIC v3.2)</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="fixed-3">Fixed<a class="anchor" aria-label="anchor" href="#fixed-3"></a>
</h5>
<ul>
<li>Bug in writing (large) report contents to JSON (issue <a href="https://github.com/sigven/pcgr/issues/118" class="external-link">#118</a>)</li>
<li>Bug (typo) in merge of clinical evidence items from different
sources (CIVIC + CGI) (issue <a href="https://github.com/sigven/pcgr/issues/126" class="external-link">#126</a>)</li>
<li>Bug in value box for number of (high-confident) kataegis events -
rmarkdown (issue <a href="https://github.com/sigven/pcgr/issues/122" class="external-link">#122</a>)</li>
<li>Bug in value box for tumor purity/ploidy -rmarkdown (issue <a href="https://github.com/sigven/pcgr/issues/129" class="external-link">#129</a>)</li>
</ul>
</div>
<div class="section level5">
<h5 id="removed-2">Removed<a class="anchor" aria-label="anchor" href="#removed-2"></a>
</h5>
<ul>
<li>Command-line options
<ul>
<li>
<code>--conf</code> - TOML-based configuration file</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-9-1">v0.9.1<a class="anchor" aria-label="anchor" href="#v0-9-1"></a>
</h2>
<ul>
<li><p>Date: <strong>2020-11-30</strong></p></li>
<li>
<p>Data updates:</p>
<ul>
<li>ClinVar,</li>
<li>GWAS catalog</li>
<li>CIViC</li>
<li>CancerMine</li>
<li>dbNSFP</li>
<li>KEGG</li>
<li>ChEMBL/DGIdb</li>
<li>Disease Ontology, Experimental Factor Ontology</li>
</ul>
</li>
</ul>
<div class="section level5">
<h5 id="added-2">Added<a class="anchor" aria-label="anchor" href="#added-2"></a>
</h5>
<ul>
<li>added possibility to configure algorithm for TMB calculation,
optional argument <code>tmb_algorithm</code> - all coding variants
(<strong>all_coding</strong>) or non-synonymous variants only
(<strong>nonsyn</strong>)</li>
<li>R code subject to static analysis with <a href="https://github.com/jimhester/lintr" class="external-link">lintr</a>
</li>
<li>Improved Conda recipe (i.e. <code>meta.yaml</code>) with version
pinning of all package dependencies</li>
</ul>
</div>
<div class="section level5">
<h5 id="changed-2">Changed<a class="anchor" aria-label="anchor" href="#changed-2"></a>
</h5>
<ul>
<li>Removed DisGeNET annotations from output (associations from Open
Targets Platform serve same purpose)</li>
<li>Version pinning of software dependencies in Dockerfile:
<ul>
<li>All R packages necessary for PCGR is installed using the <a href="https://rstudio.github.io/renv/index.html" class="external-link">renv framework</a>,
ensuring improved versioning and reproducibility</li>
<li>Other tools/utilities and Python libraries that have been version
pinned:
<ul>
<li>bedtools, samtools, numpy, cython, scipy, cyvcf2, toml, pandas</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-9-0rc">v0.9.0rc<a class="anchor" aria-label="anchor" href="#v0-9-0rc"></a>
</h2>
<ul>
<li><p>Date: <strong>2020-09-24</strong></p></li>
<li><p>Data updates: ClinVar, GWAS catalog, GENCODE, CIViC, CancerMine,
UniProt KB, dbNSFP, Pfam, KEGG, Open Targets Platform</p></li>
<li><p>Software updates: VEP 101</p></li>
</ul>
<div class="section level5">
<h5 id="fixed-4">Fixed<a class="anchor" aria-label="anchor" href="#fixed-4"></a>
</h5>
<ul>
<li>An extra comma was mistakenly present in the template for tier 2
variants, issue <a href="https://github.com/sigven/pcgr/issues/96" class="external-link">#96</a>
</li>
<li>Missing protein domain annotations for grch38, issue <a href="https://github.com/sigven/pcgr/issues/96" class="external-link">#116</a>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="changed-3">Changed<a class="anchor" aria-label="anchor" href="#changed-3"></a>
</h5>
<ul>
<li>All arguments to <code>pcgr.py</code> is now non-positional</li>
<li>Arguments to <code>pcgr.py</code> are divided into two groups:
<em>required</em> and <em>optional</em>
</li>
<li>Options <strong><em>allelic_support:tumor_dp_min</em></strong>,
<strong><em>allelic_support:tumor_af_min</em></strong>,
<strong><em>allelic_support:control_dp_min</em></strong>,
<strong><em>allelic_support:control_af_max</em></strong> in PCGR
configuration file are now optional arguments
<code>--tumor_dp_min</code>, <code>--tumor_af_min</code>,
<code>--control_dp_min</code>,
–control_af_max<code>in</code>cpsr.py`</li>
<li>Option <strong><em>mutational_burden:mutational_burden</em></strong>
in PCGR configuration file is now optional argument
<code>--estimate_tmb</code> in <code>pcgr.py</code>
</li>
<li>Option <strong><em>msi:msi</em></strong> in PCGR configuration file
is now optional argument <code>--estimate_msi_status</code> in
<code>pcgr.py</code>
</li>
<li>Option <strong>mutational_signatures:mutational_signatures</strong>
in PCGR configuration file is now optional argument
<code>--estimate_signatures</code> in <code>pcgr.py</code>
</li>
<li>Options
<strong>mutational_signatures:mutsignatures_signature_limit</strong>,
<strong>mutational_signatures:mutsignatures_normalization</strong>,
<strong>mutational_signatures:mutsignatures_mutation_limit</strong>,
<strong>mutational_signatures:mutsignatures_cutoff</strong> are removed
(used for deconstructSigs analysis, which is no longer in use)</li>
<li>Optional argument <code>--cna_overlap_pct</code> in
<code>pcgr.py</code> replaces <strong>cna:cna_overlap_pct</strong> in
PCGR configuration file</li>
<li>Optional argument <code>--logr_gain</code> in <code>pcgr.py</code>
replaces <strong>cna:logr_gain</strong> in PCGR configuration file</li>
<li>Optional argument <code>--logr_homdel</code> in <code>pcgr.py</code>
replaces <strong>cna:logr_homdel</strong> in PCGR configuration
file</li>
<li>Removed <strong>mutational_burden:tmb_low_limit</strong> and
<strong>mutational_burden:tmb_intermediate_limit</strong> - TMB is no
longer interpreted in the context of thresholds</li>
<li>Classifications of genes as tumor suppressors/oncogenes are now
based on a combination of CancerMine citation count and presence in
Network of Cancer Genes</li>
<li>
<strong>Settings</strong> section of report is now divived into
three:
<ul>
<li>Metadata - sample and sequencing assay</li>
<li>Report configuration</li>
</ul>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="added-3">Added<a class="anchor" aria-label="anchor" href="#added-3"></a>
</h5>
<ul>
<li>Optional argument <code>--include_trials</code> in
<code>pcgr.py</code> - includes a section with annotated clinical trials
for the tumor type in question</li>
<li>Optional argument <code>--assay</code> in <code>pcgr.py</code> -
designates type of sequencing assay</li>
<li>Optional argument <code>--cell_line</code> in <code>pcgr.py</code> -
designates runs of tumor cell lines (only for display, not used to
configure any analysis)</li>
<li>Optional argument <code>--min_mutations_signatures</code> in
<code>pcgr.py</code> - minimum number of required mutations for
mutational signature analysis with MutationalPatterns</li>
<li>Optional argument <code>--all_reference_signatures</code> in
<code>pcgr.py</code> - considers all reference signatures during fitting
of mutational profile to known signatures</li>
<li>Optional argument <code>--estimate_signatures</code> now also
includes detection of potential kataegis events (WGS/WES assays only),
and rainfall plot in the flexdashboard output</li>
<li>The user can now distinguish (through color codes) whether a
biomarker has been mapped exactly (nucleotide change) or at a regional
level (codon/exon)</li>
<li>All variant-associated biomarkers (regardless of assignment to TIER
1/2) are now found in a new section (SNVs/InDels)</li>
<li>For copy number amplifications, other putative drug targets in
cancer are listed in a new section</li>
<li>Detailed documentation of report contents are added to the
Documentation section</li>
<li>References are updated and all provided with DOI</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-8-4">v0.8.4<a class="anchor" aria-label="anchor" href="#v0-8-4"></a>
</h2>
<ul>
<li><p>Date: <strong>2019-11-18</strong></p></li>
<li><p>Data updates: ClinVar, CIViC, CancerMine, UniProt KB</p></li>
<li><p>Software updates: VEP 98.3</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="v0-8-3">v0.8.3<a class="anchor" aria-label="anchor" href="#v0-8-3"></a>
</h2>
<ul>
<li><p>Date: <strong>2019-10-14</strong></p></li>
<li><p>Data updates: ClinVar, GWAS catalog, GENCODE, CIViC,
CancerMine</p></li>
<li><p>Software updates: VEP 98.2, vcf2tsv</p></li>
</ul>
<div class="section level5">
<h5 id="fixed-5">Fixed<a class="anchor" aria-label="anchor" href="#fixed-5"></a>
</h5>
<ul>
<li>More improved mapping between Ensembl transcripts and UniProt
accessions (using also RefSeq accessions where available)</li>
</ul>
</div>
<div class="section level5">
<h5 id="added-4">Added<a class="anchor" aria-label="anchor" href="#added-4"></a>
</h5>
<ul>
<li>Possibility to filter evidence items by RATING in interactive data
tables</li>
</ul>
</div>
<div class="section level5">
<h5 id="changed-4">Changed<a class="anchor" aria-label="anchor" href="#changed-4"></a>
</h5>
<ul>
<li>Option <em>target_size_mb</em> in pcgr.py replaces
<em>target_size_mb</em> in configuration file, more convenient in terms
of configuring runs</li>
<li>Option <em>tumor_type</em> in pcgr.py replaces <em>tumor_type</em>
in configuration file</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-8-2">v0.8.2<a class="anchor" aria-label="anchor" href="#v0-8-2"></a>
</h2>
<ul>
<li><p>Date: <strong>2019-09-29</strong></p></li>
<li><p>Data updates: ClinVar, GWAS catalog, GENCODE, DiseaseOntology,
CIViC, CancerMine, UniProt KB</p></li>
<li><p>Software updates: VEP 97.3, vcfanno <a href="https://github.com/brentp/vcfanno/releases/tag/v0.3.2" class="external-link">0.3.2</a>,
LOFTEE (VEP plugin) <a href="https://github.com/konradjk/loftee/releases" class="external-link">1.0.3</a></p></li>
</ul>
<div class="section level5">
<h5 id="fixed-6">Fixed<a class="anchor" aria-label="anchor" href="#fixed-6"></a>
</h5>
<ul>
<li>Bug in concatenation of clinical evidence items from different
sources (CIVIC + CBMDB) (issues <a href="https://github.com/sigven/pcgr/issues/83" class="external-link">#83</a>,<a href="https://github.com/sigven/pcgr/issues/87" class="external-link">#87</a>)</li>
<li>Silent variants that coincide with biomarkers reported at codon
level are ignored</li>
<li>Distinction between clinical evidence items of different origins
(somatic + germline)</li>
<li>Improved mapping between Ensembl transcripts and UniProt accessions
(using also RefSeq accessions where available)</li>
<li>Bug in UpSetPlot for cases where filtering produce less than two
intersecting sets</li>
</ul>
</div>
<div class="section level5">
<h5 id="added-5">Added<a class="anchor" aria-label="anchor" href="#added-5"></a>
</h5>
<ul>
<li>New field <a href="https://www.ensembl.org/info/genome/genebuild/mane.html" class="external-link">‘mane’</a>
as criteria for pick order in configuration file (VEP)</li>
<li>Sample identifier to copy number annotation output (convenient for
concatenation of output from multiple samples)</li>
<li>Capturing allelic depth (t_depth, t_ref_count etc.) in vcf2maf
output (enhancement <a href="https://github.com/sigven/pcgr/issues/52" class="external-link">#52</a>)</li>
<li>Option <em>tumor_only</em> in pcgr.py, replaces
<em>vcf_tumor_only</em> in configuration file, more convenient in terms
of configuration</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-8-1">v0.8.1<a class="anchor" aria-label="anchor" href="#v0-8-1"></a>
</h2>
<ul>
<li>Date: <strong>2019-05-22</strong>
</li>
</ul>
<div class="section level5">
<h5 id="added-6">Added<a class="anchor" aria-label="anchor" href="#added-6"></a>
</h5>
<ul>
<li>
<em>Cancer_NOS.toml</em> as configuration file for unspecified tumor
types</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-8-0">v0.8.0<a class="anchor" aria-label="anchor" href="#v0-8-0"></a>
</h2>
<ul>
<li>Date: <strong>2019-05-20</strong>
</li>
</ul>
<div class="section level5">
<h5 id="fixed-7">Fixed<a class="anchor" aria-label="anchor" href="#fixed-7"></a>
</h5>
<ul>
<li>Bug in value box for Tier 2 variants (new line carriage) <a href="https://github.com/sigven/pcgr/issues/73" class="external-link">Issue #73</a>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="added-7">Added<a class="anchor" aria-label="anchor" href="#added-7"></a>
</h5>
<ul>
<li>Upgraded VEP to v96
<ul>
<li>Skipping the <em>–regulatory</em> VEP option to avoid forking issues
and to improve speed (See <a href="https://github.com/Ensembl/ensembl-vep/issues/384" class="external-link">this
issue</a>)</li>
<li>Added option to configure <em>pick-order</em> for choice of primary
transcript in configuration file</li>
</ul>
</li>
<li>Pre-made configuration files for each tumor type in <em>conf</em>
folder</li>
<li>Possibility to append a CNA plot file (.png format) to the section
of the report with <em>Somatic CNAs</em> <a href="https://github.com/sigven/pcgr/issues/58" class="external-link">previous feature
request</a>
</li>
<li>Added possibility to input estimates of <strong>tumor
purity</strong> and <strong>ploidy</strong>
<ul>
<li>shown as value boxes in <em>Main results</em>
</li>
</ul>
</li>
<li>Tumor mutational burden is now compared with the distribution of TMB
observed for TCGA’s cohorts (organized by primary site)
<ul>
<li>Default target size is now 34Mb (approx. estimate from exome-wide
calculation of protein-coding parts of GENCODE)</li>
</ul>
</li>
<li>Added flexibility for variant filtering in tumor-only input callsets
<ul>
<li>Added additional options to exclude likely germline variants (both
requires the tumor VAF tag to be correctly specified in the input VCF)
<ul>
<li>
<strong>exclude_likely_hom_germline</strong> - removes any variant
with an allelic fraction of 1 (100%) - very unlikely somatic event</li>
</ul>
</li>
<li>
<strong>exclude_likely_het_germline</strong> - removes any variant
with
<ul>
<li>an allelic fraction between 0.4 and 0.6, and</li>
<li>presence in dbSNP + gnomAD, and</li>
<li>no presence as somatic event in COSMIC/TCGA</li>
</ul>
</li>
<li>Added possibility to input <em>PANEL-OF-NORMALS</em> VCF - this to
support the many labs that have sequenced a database/pool of healthy
controls. This set of variants are utilized in PCGR to improve the
variant filtering when running in tumor-only mode. The
<em>PANEL-OF-NORMALS</em> annotation work as follows:
<ul>
<li>all variants in the tumor that coincide with any variant listed in
the <em>PANEL-OF-NORMALS</em> VCF is appended with a
<strong>PANEL_OF_NORMALS</strong> flag in the query VCF with tumor
variants.</li>
</ul>
</li>
<li>If configuration parameter <strong>exclude_pon</strong> is set to
True in <strong>tumor_only</strong> runs, all variants with a
<strong>PANEL_OF_NORMALS</strong> flag are filtered/excluded</li>
</ul>
</li>
<li>For tumor-only runs, added an <a href="https://github.com/hms-dbmi/UpSetR#Demo" class="external-link">UpSet plot</a> showing
how different filtering sources (gnomAD, 1KG Project, panel-of-normals
etc) contribute in the germline filtering procedure</li>
<li>Variants in <em>Tier 3 / Tier 4 / Noncoding</em> are now sorted (and
color-coded) according to the target (gene) association score to the
cancer phenotype, as provided by the <a href="https://docs.targetvalidation.org/getting-started/scoring" class="external-link">OpenTargets
Platform</a>
</li>
<li>Added annotation of TCGA’s ten oncogenic signaling pathways</li>
<li>Added <em>EXONIC_STATUS</em> annotation tag (VCF and TSV)
<ul>
<li>
<em>exonic</em> denotes all protein-altering AND cannonical
splicesite altering AND synonymous variants, <em>nonexonic</em> denotes
the complement</li>
</ul>
</li>
<li>Added <em>CODING_STATUS</em> annotation tag (VCF and TSV)
<ul>
<li>
<em>coding</em> denotes all protein-altering AND cannonical
splicesite altering, <em>noncoding</em> denotes the complement</li>
</ul>
</li>
<li>Added <em>SYMBOL_ENTREZ</em> annotation tag (VCF)
<ul>
<li>Official gene symbol from NCBI EntreZ (SYMBOL provided by VEP can
sometimes be non-official/alias (i.e. for GENCODE v19/grch37))</li>
</ul>
</li>
<li>Added <em>SIMPLEREPEATS_HIT</em> annotation tag (VCF and TSV)
<ul>
<li>Variant overlaps UCSC <em>simpleRepeat</em> sequence repeat track -
used for MSI prediction</li>
</ul>
</li>
<li>Added <em>WINMASKER_HIT</em> annotation tag (VCF and TSV)
<ul>
<li>Variant overlaps UCSC <em>windowmaskerSdust</em> sequence repeat
track - used for MSI prediction</li>
</ul>
</li>
<li>Added <em>PUTATIVE_DRIVER_MUTATION</em> annotation tag (VCF and TSV)
<ul>
<li>Putative cancer driver mutation discovered by multiple approaches
from 9,423 tumor exomes in TCGA. Format:
symbol:hgvsp:ensembl_transcript_id:discovery_approaches</li>
</ul>
</li>
<li>Added <em>OPENTARGETS_DISEASE_ASSOCS</em> annotation tag (VCF and
TSV)
<ul>
<li>Associations between protein targets and disease based on multiple
lines of evidence (mutations,affected pathways,GWAS, literature etc).
Format: CUI:EFO_ID:IS_DIRECT:OVERALL_SCORE</li>
</ul>
</li>
<li>Added <em>OPENTARGETS_TRACTABILITY_COMPOUND</em> annotation tag (VCF
and TSV)
<ul>
<li>Confidence for the existence of a modulator (small molecule) that
interacts with the target (protein) to elicit a desired biological
effect</li>
</ul>
</li>
<li>Added <em>OPENTARGTES_TRACTABILITY_ANTIBODY</em> annotation tag (VCF
and TSV)
<ul>
<li>Confidence for the existence of a modulator (antibody) that
interacts with the target (protein) to elicit a desired biological
effect</li>
</ul>
</li>
<li>Added <em>CLINVAR_REVIEW_STATUS_STARS</em> annotation tag
<ul>
<li>Rating of the ClinVar variant (0-4 stars) with respect to level of
review</li>
</ul>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="changed-5">Changed<a class="anchor" aria-label="anchor" href="#changed-5"></a>
</h5>
<ul>
<li>Moved from <a href="https://www.intogen.org/" class="external-link">IntoGen’s driver
mutation resource</a> to <a href="https://doi.org/10.1016/j.cell.2018.02.060" class="external-link">TCGA’s putative driver
mutation list</a> in display of driver mutation status</li>
<li>Moved option for vcf_validation from configuration file to run
script (<code>--no_vcf_validate</code>)</li>
</ul>
</div>
<div class="section level5">
<h5 id="removed-3">Removed<a class="anchor" aria-label="anchor" href="#removed-3"></a>
</h5>
<ul>
<li>Original tier model ‘pcgr’</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-7-0">v0.7.0<a class="anchor" aria-label="anchor" href="#v0-7-0"></a>
</h2>
<ul>
<li>Date: <strong>2018-11-27</strong>
</li>
</ul>
<div class="section level5">
<h5 id="fixed-8">Fixed<a class="anchor" aria-label="anchor" href="#fixed-8"></a>
</h5>
<ul>
<li>Bug in assignment of variants to tier1/tier2 <a href="https://github.com/sigven/pcgr/issues/61" class="external-link">Issue #61</a>
</li>
<li>Missing config option for <em>maf_gnomad_asj</em> in TOML file (also
setting operator to <code>&lt;=</code>) <a href="https://github.com/sigven/pcgr/issues/60" class="external-link">Issue #60</a>
</li>
<li>Bug in new CancerMine oncogene/tumor suppressor annotation <a href="https://github.com/sigven/pcgr/issues/53" class="external-link">Issue #53</a>
</li>
<li>vcfanno fix for empty Description (upgrade to vcfanno v0.3.1 <a href="https://github.com/sigven/pcgr/issues/49" class="external-link">Issue #49</a>)</li>
<li>Bug in message showing too few variants for MSI prediction, <a href="https://github.com/sigven/pcgr/issues/55" class="external-link">Issue #55</a>
</li>
<li>Bug in appending of custom VCF tags
<ul>
<li>Still unsolved: how to disambiguate identical FORMAT and INFO tags
in vcf2tsv</li>
</ul>
</li>
<li>Bug in SCNA value box display for multiple copy number hits (<a href="https://github.com/sigven/pcgr/issues/47" class="external-link">Issue #47</a>)</li>
<li>Bug in vcf2tsv (handling INFO tags encoded with ‘Type = String’, <a href="https://github.com/sigven/pcgr/issues/39" class="external-link">Issue #39</a>)</li>
<li>Bug in search of UniProt functional features (BED feature regions
spanning exons are now handled)</li>
<li>Stripped off HTML elements (TCGA_FREQUENCY, DBSNP) in TSV
output</li>
<li>Some effect predictions from dbNSFP were not properly parsed
(e.g. multiple prediction entries from multiple transcript isoforms),
these should now be retrieved correctly</li>
<li>Removed ‘COSM’ prefix in COSMIC mutation links</li>
<li>Bug in retrieval of splice site predictions from dbscSNV</li>
</ul>
</div>
<div class="section level5">
<h5 id="added-8">Added<a class="anchor" aria-label="anchor" href="#added-8"></a>
</h5>
<ul>
<li>Possibility to run PCGR in a non-Docker environment (e.g. using the
<em>–no-docker</em> option). Thanks to an excellent contribution by <a href="https://github.com/vladsaveliev" class="external-link">Vlad Saveliev</a>, <a href="https://github.com/sigven/pcgr/issues/35" class="external-link">Issue #35</a>
<ul>
<li>Added possibility to add docker user-id</li>
</ul>
</li>
<li>Possibility for MAF file output (converted with vcf2maf), must be
configured by the user in the TOML file (i.e. <em>vcf2maf = true</em>,
<a href="https://github.com/sigven/pcgr/issues/17" class="external-link">Issue #17</a>)</li>
<li>Possibility for adding custom VCF INFO tags to PCGR output files
(JSON/TSV), must be configured by the user in the TOML file
(i.e. <em>custom_tags</em>)</li>
<li>Added MUTATION_HOTSPOT_CANCERTYPE in data tables (i.e. listing tumor
types in which hotspot mutations have been found)</li>
<li>Included the ‘rs’ prefix for dbSNP identifiers (HTML and TSV
output)</li>
<li>Individual entries/columns for variant effect predictions:
<ul>
<li>Individual algorithms: SIFT_DBNSFP, M_CAP_DBNSFP, MUTPRED_DBNSFP,
MUTATIONTASTER_DBNSFP, MUTATIONASSESSOR_DBNSFP, FATHMM_DBNSFP,
FATHMM_MKL_DBNSFP, PROVEAN_DBNSFP</li>
<li>Ensemble predictions (META_LR_DBNSFP), dbscSNV splice site
predictions (SPLICE_SITE_RF_DBNSFP, SPLICE_SITE_ADA_DBNSFP)</li>
</ul>
</li>
<li>Upgraded samtools to v1.9 (makes vcf2maf work properly)</li>
<li>Added Ensembl gene/transcript id and corresponding RefSeq mRNA id to
TSV/JSON</li>
<li>Added for future implementation:
<ul>
<li>SeqKat + karyoploteR for exploration of
<em>kataegis/hypermutation</em>
</li>
<li>CELLector - genomics-guided selection of cancer cell lines</li>
</ul>
</li>
<li>Upgraded VEP to v94</li>
</ul>
</div>
<div class="section level5">
<h5 id="changed-6">Changed<a class="anchor" aria-label="anchor" href="#changed-6"></a>
</h5>
<ul>
<li>Changed CANCER_MUTATION_HOTSPOT to MUTATION_HOTSPOT</li>
<li>Moved from <a href="https://bioinfo.uth.edu/TSGene/" class="external-link">TSGene 2.0</a>
to <a href="https://zenodo.org/record/1336650#.W9QMdRMzaL4" class="external-link">CancerMine</a> for
annotation of tumor suppressor genes and proto-oncogenes
<ul>
<li>A minimum of n=3 citations were required to include
literatured-mined tumor suppressor genes and proto-oncogenes from
CancerMine</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-6-2-1">v0.6.2.1<a class="anchor" aria-label="anchor" href="#v0-6-2-1"></a>
</h2>
<ul>
<li>Date: <strong>2018-05-14</strong>
</li>
</ul>
<div class="section level5">
<h5 id="fixed-9">Fixed<a class="anchor" aria-label="anchor" href="#fixed-9"></a>
</h5>
<ul>
<li>Bug in copy number annotation (broad/focal)</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-6-2">v0.6.2<a class="anchor" aria-label="anchor" href="#v0-6-2"></a>
</h2>
<ul>
<li>Date: <strong>2018-05-09</strong>
</li>
</ul>
<div class="section level5">
<h5 id="fixed-10">Fixed<a class="anchor" aria-label="anchor" href="#fixed-10"></a>
</h5>
<ul>
<li>Bug in copy number segment display (missing variable initalization,
<a href="https://github.com/sigven/pcgr/issues/34" class="external-link">Issue #34</a>))</li>
<li>Typo in gnomAD filter statistic (fraction, <a href="https://github.com/sigven/pcgr/issues/31" class="external-link">Issue #31</a>)</li>
<li>Bug in mutational signature analysis for grch38 (forgot to pass
BSgenome object, <a href="https://github.com/sigven/pcgr/issues/27" class="external-link">Issue #27</a>)</li>
<li>Missing proper ASCII-encoding in vcf2tsv conversion, <a href="https://github.com/sigven/pcgr/issues/35" class="external-link">Issue #</a>
</li>
<li>Removed ‘Noncoding mutations’ section when no input VCF is
present</li>
<li>Bug in annotation of copy number event type (focal/broad)</li>
<li>Bug in copy number annotation (missing protein-coding
transcripts)</li>
<li>Updated MSI prediction (variable importance, performance
measures)</li>
</ul>
</div>
<div class="section level5">
<h5 id="added-9">Added<a class="anchor" aria-label="anchor" href="#added-9"></a>
</h5>
<ul>
<li>Genome assembly is appended to every output file</li>
<li>Issue warning for copy number segment that goes beyond chromosomal
lengths of specified assembly (segments will be skipped)</li>
<li>Added missing subtypes for ‘Skin_Cancer_NOS’ in the cancer phenotype
dataset</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-6-1">v0.6.1<a class="anchor" aria-label="anchor" href="#v0-6-1"></a>
</h2>
<ul>
<li>Date: <strong>2018-05-02</strong>
</li>
</ul>
<div class="section level5">
<h5 id="fixed-11">Fixed<a class="anchor" aria-label="anchor" href="#fixed-11"></a>
</h5>
<ul>
<li>Bug in tier assignment ‘pcgr_acmg’ (case for no variants in
tier1,2,3)</li>
<li>Bug in tier assignment ‘pcgr_acmg’ (no tumor type specified,
evidence items with weak support detected)</li>
<li>Bug: duplicated variants in ‘Tier 3’ resulting from genes encoded
with dual roles as tumor suppressor genes/oncogenes</li>
<li>Bug: duplicated variants in ‘Tier 1/Noncoding variants’ resulting
from rare cases of noncoding variants occurring in Tier 1 (synonymous
variants with biomarker role)</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="v0-6-0">v0.6.0<a class="anchor" aria-label="anchor" href="#v0-6-0"></a>
</h2>
<ul>
<li>Date: <strong>2018-04-25</strong>
</li>
</ul>
<div class="section level5">
<h5 id="added-10">Added<a class="anchor" aria-label="anchor" href="#added-10"></a>
</h5>
<ul>
<li>New argument in pcgr.py
<ul>
<li>
<em>assembly</em> (grch37/grch38)</li>
</ul>
</li>
<li>New option in pcgr.py
<ul>
<li>
<em>–basic</em> - run comprehensive VCF annotation only, skip report
generation and additional analyses</li>
</ul>
</li>
<li>New sections in HTML report
<ul>
<li>
<em>Settings and annotation sources</em> - now also listing key PCGR
configuration settings</li>
<li>
<em>Main findings</em> - Six value boxes indicating the main
findings of clinical relevance</li>
</ul>
</li>
<li>New configuration options
<ul>
<li>
<strong>tier_model</strong>(string) - choice between
<em>pcgr_acmg</em> and <em>pcgr</em>
</li>
<li>
<strong>mutational_burden</strong> - set TMB tertile limits
<ul>
<li><em>tmb_low_limit (float)</em></li>
<li><em>tmb_intermediate_limit (float)</em></li>
</ul>
</li>
<li>
<strong>tumor_type</strong> - choose between 34 tumor types/classes:
<ul>
<li><em>Adrenal_Gland_Cancer_NOS (logical)</em></li>
<li><em>Ampullary_Carcinoma_NOS (logical)</em></li>
<li><em>Biliary_Tract_Cancer_NOS (logical)</em></li>
<li><em>Bladder_Urinary_Tract_Cancer_NOS (logical)</em></li>
<li><em>Blood_Cancer_NOS (logical)</em></li>
<li><em>Bone_Cancer_NOS (logical)</em></li>
<li><em>Breast_Cancer_NOS (logical)</em></li>
<li><em>CNS_Brain_Cancer_NOS (logical)</em></li>
<li><em>Colorectal_Cancer_NOS (logical)</em></li>
<li><em>Cervical_Cancer_NOS (logical)</em></li>
<li><em>Esophageal_Stomach_Cancer_NOS (logical)</em></li>
<li><em>Head_And_Neck_Cancer_NOS (logical)</em></li>
<li><em>Hereditary_Cancer_NOS (logical)</em></li>
<li><em>Kidney_Cancer_NOS (logical)</em></li>
<li><em>Leukemia_NOS (logical)</em></li>
<li><em>Liver_Cancer_NOS (logical)</em></li>
<li><em>Lung_Cancer_NOS (logical)</em></li>
<li><em>Lymphoma_Hodgkin_NOS (logical)</em></li>
<li><em>Lymphoma_Non_Hodgkin_NOS (logical)</em></li>
<li><em>Ovarian_Fallopian_Tube_Cancer_NOS (logical)</em></li>
<li><em>Pancreatic_Cancer_NOS (logical)</em></li>
<li><em>Penile_Cancer_NOS (logical)</em></li>
<li><em>Peripheral_Nervous_System_Cancer_NOS (logical)</em></li>
<li><em>Peritoneal_Cancer_NOS (logical)</em></li>
<li><em>Pleural_Cancer_NOS (logical)</em></li>
<li><em>Prostate_Cancer_NOS (logical)</em></li>
<li><em>Skin_Cancer_NOS (logical)</em></li>
<li><em>Soft_Tissue_Cancer_NOS (logical)</em></li>
<li><em>Stomach_Cancer_NOS (logical)</em></li>
<li><em>Testicular_Cancer_NOS (logical)</em></li>
<li><em>Thymic_Cancer_NOS (logical)</em></li>
<li><em>Thyroid_Cancer_NOS (logical)</em></li>
<li><em>Uterine_Cancer_NOS (logical)</em></li>
<li><em>Vulvar_Vaginal_Cancer_NOS (logical)</em></li>
</ul>
</li>
<li>
<strong>mutational_signatures</strong>
<ul>
<li>
<em>mutsignatures_cutoff (float)</em> - discard any signature
contributions with a weight less than the cutoff</li>
</ul>
</li>
<li>
<strong>cna</strong>
<ul>
<li>
<em>transcript_cna_overlap (float)</em> - minimum percent overlap
between copy number segment and transcripts (average) for tumor
suppressor gene/proto-oncogene to be reported</li>
</ul>
</li>
<li>
<strong>allelic_support</strong>
<ul>
<li>If input VCF has correctly formatted depth/allelic fraction as INFO
tags, users can add thresholds on depth/support that are applied prior
to report generation
<ul>
<li>
<em>tumor_dp_min (integer)</em> - minimum sequencing depth for
variant in tumor sample</li>
<li>
<em>tumor_af_min (float)</em> - minimum allelic fraction for variant
in tumor sample</li>
<li>
<em>normal_dp_min (integer)</em> - minimum sequencing depth for
variant in normal sample</li>
<li>
<em>normal_af_max (float)</em> - maximum allelic fraction for
variant in normal sample</li>
</ul>
</li>
</ul>
</li>
<li>
<strong>visual</strong>
<ul>
<li>
<em>report_theme (string)</em> - visual theme of report
(Bootstrap)</li>
</ul>
</li>
<li>
<strong>other</strong>
<ul>
<li>
<em>vcf_validation (logical)</em> - keep/skip VCF validation by <a href="https://github.com/EBIvariation/vcf-validator" class="external-link">vcf-validator</a>
</li>
</ul>
</li>
</ul>
</li>
<li>New output file - JSON output of HTML report content</li>
<li>New INFO tags of PCGR-annotated VCF
<ul>
<li><em>CANCER_PREDISPOSITION</em></li>
<li><em>PFAM_DOMAIN</em></li>
<li><em>TCGA_FREQUENCY</em></li>
<li><em>TCGA_PANCANCER_COUNT</em></li>
<li><em>ICGC_PCAWG_OCCURRENCE</em></li>
<li><em>ICGC_PCAWG_AFFECTED_DONORS</em></li>
<li><em>CLINVAR_MEDGEN_CUI</em></li>
</ul>
</li>
<li>New column entries in annotated SNV/InDel TSV file:
<ul>
<li><em>CANCER_PREDISPOSITION</em></li>
<li><em>ICGC_PCAWG_OCCURRENCE</em></li>
<li><em>TCGA_FREQUENCY</em></li>
</ul>
</li>
<li>New column in CNA output
<ul>
<li>
<em>TRANSCRIPTS</em> - aberration-overlapping transcripts (Ensembl
transcript IDs)</li>
<li>
<em>MEAN_TRANSCRIPT_CNA_OVERLAP</em> - Mean overlap (%) betweeen
gene transcripts and aberration segment</li>
</ul>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="removed-4">Removed<a class="anchor" aria-label="anchor" href="#removed-4"></a>
</h5>
<ul>
<li>Elements of databundle (now annotated directly through VEP):
<ul>
<li>dbsnp</li>
<li>gnomad/exac</li>
<li>1000G project</li>
</ul>
</li>
<li>INFO tags of PCGR-annotated VCF
<ul>
<li><em>DBSNPBUILDID</em></li>
<li><em>DBSNP_VALIDATION</em></li>
<li><em>DBSNP_SUBMISSIONS</em></li>
<li><em>DBSNP_MAPPINGSTATUS</em></li>
<li><em>GWAS_CATALOG_PMID</em></li>
<li><em>GWAS_CATALOG_TRAIT_URI</em></li>
<li><em>DOCM_DISEASE</em></li>
</ul>
</li>
<li>Output files
<ul>
<li>TSV files with mutational signature results and biomarkers
(i.e. <em>sample_id.pcgr.snvs_indels.biomarkers.tsv</em> and
<em>sample_id.pcgr.mutational_signatures.tsv</em>)
<ul>
<li>Data can still be retrieved - now from the JSON dump</li>
</ul>
</li>
<li>MAF file
<ul>
<li>The previous MAF output was generated in a custom fashion, a more
accurate MAF output based on <a href="https://github.com/mskcc/vcf2maf" class="external-link uri">https://github.com/mskcc/vcf2maf</a> will be incorporated in
the next release</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="changed-7">Changed<a class="anchor" aria-label="anchor" href="#changed-7"></a>
</h5>
<ul>
<li>HTML report sections
<ul>
<li>
<em>Tier statistics</em> and <em>Variant statistics</em> are now
grouped into the section <em>Tier and variant statistics</em>
</li>
<li>
<em>Tier 5</em> is now <em>Noncoding mutations</em> (i.e. not
considered a tier per se)</li>
<li>Sliders for allelic fraction in the <em>Global variant browser</em>
are now fixed from 0 to 1 (0.05 intervals)</li>
</ul>
</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
