<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>FAQ • PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FAQ">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-7HRLE8GQRE"></script><script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){dataLayer.push(arguments);}
   gtag('js', new Date());

   gtag('config', 'G-7HRLE8GQRE');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/running.html">Running</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/input.html">Input files</a></li>
    <li><a class="dropdown-item" href="../articles/output.html">Output files</a></li>
    <li><a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a></li>
    <li><a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="../articles/developers.html">Developer notes</a></li>
    <li><a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a></li>
  </ul>
</li>
<li class="active nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>FAQ</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/main/vignettes/faq.Rmd" class="external-link"><code>vignettes/faq.Rmd</code></a></small>
      <div class="d-none name"><code>faq.Rmd</code></div>
    </div>

    
    
<p>Frequently asked questions regarding PCGR usage and
functionality:</p>
<p><strong>1. I do not see any data related to allelic depth/support in
my report. I thought that PCGR can grab this information automatically
from my VCF?</strong></p>
<p><em>Answer: VCF variant genotype data (i.e. AD/DP) is something that
you as a user need to specify explicitly when running PCGR. In our
experience, there is currently no uniform way that variant callers
format these types of data (allelic fraction/depth, tumor/normal) in the
VCF, and this makes it very challenging for PCGR to automatically grab
this information from any VCF. Please take a careful look at the example
VCF files (<code>examples</code> folder) that comes with PCGR for how
PCGR expects this information to be formatted, and make sure your VCF is
formatted accordingly. There is also an in-depth explanation on the
matter <a href="input.html#formatting-of-allelic-depthsupport-dpad">described
here</a></em></p>
<p><strong>2. Is it possible to utilize PCGR for analysis of multiple
samples?</strong></p>
<p><em>Answer: As the name of the tool implies, PCGR was developed for
the detailed analysis of individual tumor samples. However, if you take
advantage of the different outputs from PCGR, it can also be utilized
for analysis of multiple samples. First, make sure your input files are
organized per sample (i.e. one VCF file per sample, one CNA file per
sample), so that they can be fed directly to PCGR. Now, once all samples
have been processed with PCGR, note that all the tab-separated output
files (i.e. annotated SNVs, gene copy numbers) contain the sample
identifier, which enable them to be aggregated and suitable for a
downstream multi-sample analysis. Also note the multi-sheet Excel
workbook, which contains numerous outputs from PCGR, and can be
processed to aggregate findings across samples.</em></p>
<p><strong>3. I do not see the expected transcript-specific consequence
for a particular variant. In what way is the primary variant consequence
established?</strong></p>
<p><em>Answer: PCGR relies upon</em> <a href="https://www.ensembl.org/info/docs/tools/vep/index.html" class="external-link">VEP</a>
<em>for consequence prioritization, in which a specific
transcript-specific consequence is chosen as the primary variant
consequence. In the PCGR configuration file, you may customise how this
is chosen by changing the order of criteria applied when choosing a
primary consequence block - parameter</em> <a href="https://www.ensembl.org/info/docs/tools/vep/script/vep_other.html#pick_options" class="external-link">vep_pick_order</a></p>
<p><strong>4. Is it possible to use RefSeq as the underlying gene
transcript model in PCGR?</strong></p>
<p><em>Answer: PCGR uses GENCODE as the primary gene transcript model,
but we provide cross-references to corresponding RefSeq transcripts when
this is available.</em></p>
<p><strong>5. I have a VCF with structural variants detected in my tumor
sample, can PCGR process those as well?</strong></p>
<p><em>Answer: This is currently not supported as input for PCGR, but is
something we want to incorporate in the future.</em></p>
<p><strong>6. Is it possible to see all the invididual cancer subtypes
that belong to each of the 30 different tumor sites?</strong></p>
<p><em>Answer: Yes, see</em> <a href="primary_tumor_sites.html">an
overview of phenotypes associated with primary tumor sites</a>. See also
the related GitHub repository <a href="https://github.com/sigven/phenOncoX" class="external-link">phenOncoX</a></p>
<p><strong>7. Are there any plans to incorporate genomic biomarker
evidence from</strong> <a href="https://www.oncokb.org" class="external-link">OncoKB</a>
<strong>in PCGR?</strong></p>
<p><em>Answer: No. PCGR relies upon publicly available, open-source
resources, and further that the PCGR reference bundle can be distributed
freely to the user community. It is our understanding that</em> <a href="https://www.oncokb.org/terms" class="external-link">OncoKB’s terms of use</a> <em>do not
fit well with this strategy.</em></p>
<p><strong>8. I have RNA fusion data that I want to analyse and include
in the report. Is this possible with PCGR?</strong></p>
<p><em>Answer: This is currently not supported as input for PCGR, but is
something we are actively working on. The focus will be on whether
detected RNA fusion events are previously known (i.e. seen in other
tumor samples, e.g. from the Mitelman database), and whether some of
them are presently in use as biomarkers for diagnosis or
treatment.</em></p>
<p><strong>9. Is it possible for the users to update the data bundle to
get the most recent versions of all underlying data
sources?</strong></p>
<p><em>Answer: As of now, the data bundle is updated only with each
release of PCGR. The data harmonization pipeline of knowledge databases
in PCGR contain numerous and complex procedures, with several cleaning,
quality control, and re-formatting steps, and is semi-automated in its
present form. The versions of all databases and key software elements
are outlined in each PCGR report.</em></p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
