<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pcgrr">
<title>FAQ • PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FAQ">
<meta property="og:description" content="pcgrr">
<meta property="og:image" content="https://sigven.github.io/pcgr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Intro</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/running.html">Running</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/input.html">Input files</a>
    <a class="dropdown-item" href="../articles/output.html">Output files</a>
    <a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a>
    <a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a>
    <a class="dropdown-item" href="../articles/developers.html">Developer notes</a>
    <a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/faq.html">FAQ</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>FAQ</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/HEAD/../vignettes/faq.Rmd" class="external-link"><code>../vignettes/faq.Rmd</code></a></small>
      <div class="d-none name"><code>faq.Rmd</code></div>
    </div>

    
    
<p>Frequently asked questions regarding PCGR usage and
functionality:</p>
<p><strong>1. I do not see any data related to allelic depth/support in
my report. I thought that PCGR can grab this information automatically
from my VCF?</strong></p>
<p><em>Answer: VCF variant genotype data (i.e. AD/DP) is something that
you as a user need to specify explicitly when running PCGR. In our
experience, there is currently no uniform way that variant callers
format these types of data (allelic fraction/depth, tumor/normal) in the
VCF, and this makes it very challenging for PCGR to automatically grab
this information from any VCF. Please take a careful look at the example
VCF files (<code>examples</code> folder) that comes with PCGR for how
PCGR expects this information to be formatted, and make sure your VCF is
formatted accordingly. There is also an in-depth explanation on the
matter <a href="input.html#formatting-of-allelic-depthsupport-dpad">described
here</a></em></p>
<p><strong>2. Is it possible to utilize PCGR for analysis of multiple
samples?</strong></p>
<p><em>Answer: As the name of the tool implies, PCGR was developed for
the detailed analysis of individual tumor samples. However, if you take
advantage of the different outputs from PCGR, it can also be utilized
for analysis of multiple samples. First, make sure your input files are
organized per sample (i.e. one VCF file per sample, one CNA file per
sample), so that they can be fed directly to PCGR. Now, once all samples
have been processed with PCGR, note that all the tab-separated output
files (i.e. tiers, mutational signatures, cna segments) contain the
sample identifier, which enable them to be aggregated and suitable for a
downstream multi-sample analysis.</em></p>
<p><em>Also note that the compressed JSON output pr. sample run contains
<strong>ALL</strong> information presented in the report. Explore the
JSON contents e.g. with the</em> <a href="https://github.com/jeroen/jsonlite" class="external-link">jsonlite package</a> <em>in
R:</em></p>
<p><code>report_data &lt;- jsonlite::fromJSON('&lt;sample_id&gt;.pcgr_acmg.grch37.json.gz')</code></p>
<p><em>E.g. tiered SNV/InDel output</em>:</p>
<p><code>head(report_data$content$snv_indel$variant_set$tsv)</code></p>
<p><em>Or TMB estimate</em>:</p>
<p><code>report_data$content$tmb$variant_statistic$tmb_estimate</code></p>
<p><em>Expanding further upon the example commands shown above, we are
planning to offer a cohort analysis module in the pcgrr R package in the
near future, which can assist users more easily with the aggregation of
findings across individual samples.</em></p>
<p><strong>3. I do not see the expected transcript-specific consequence
for a particular variant. In what way is the primary variant consequence
established?</strong></p>
<p><em>Answer: PCGR relies upon</em> <a href="https://www.ensembl.org/info/docs/tools/vep/index.html" class="external-link">VEP</a>
<em>for consequence prioritization, in which a specific
transcript-specific consequence is chosen as the primary variant
consequence. In the PCGR configuration file, you may customise how this
is chosen by changing the order of criteria applied when choosing a
primary consequence block - parameter</em> <a href="https://www.ensembl.org/info/docs/tools/vep/script/vep_other.html#pick_options" class="external-link">vep_pick_order</a></p>
<p><strong>4. Is it possible to use RefSeq as the underlying gene
transcript model in PCGR?</strong></p>
<p><em>Answer: PCGR uses GENCODE as the primary gene transcript model,
but we provide cross-references to corresponding RefSeq transcripts when
this is available.</em></p>
<p><strong>5. I have a VCF with structural variants detected in my tumor
sample, can PCGR process those as well?</strong></p>
<p><em>Answer: This is currently not supported as input for PCGR, but is
something we want to incorporate in the future.</em></p>
<p><strong>6. I am surprised to see a particular gene being located in
TIER 3 for my sample, since I know that this gene is of potential
clinical significance in the tumor type I am investigating?</strong></p>
<p><em>Answer: PCGR classifies variants into tiers of significance
through an implementation of</em> <a href="variant_classification.html">published guidelines by ACMG/AMP</a>.
<em>No manual efforts for individual tumor types are conducted beyond
this rule-based scheme. The users need to keep this in mind when
interpreting the tier contents of the report.</em></p>
<p><strong>7. Is it possible to see all the invididual cancer subtypes
that belong to each of the 30 different tumor sites?</strong></p>
<p><em>Answer: Yes, see</em> <a href="primary_tumor_sites.html">an
overview of phenotypes associated with primary tumor sites</a>. See also
the related GitHub repository <a href="https://github.com/sigven/oncoPhenoMap" class="external-link">oncoPhenoMap</a></p>
<p><strong>8. Is there any plans to incorporate data from</strong> <a href="https://www.oncokb.org" class="external-link">OncoKB</a> <strong>in PCGR?</strong></p>
<p><em>Answer: No. PCGR relies upon publicly available open-source
resources, and further that the PCGR data bundle can be distributed
freely to the user community. It is our understanding that</em> <a href="https://www.oncokb.org/terms" class="external-link">OncoKB’s terms of use</a> <em>do not
fit well with this strategy.</em></p>
<p><strong>9. Is it possible for the users to update the data bundle to
get the most recent versions of all underlying data
sources?</strong></p>
<p><em>Answer: As of now, the data bundle is updated only with each
release of PCGR. The data harmonization pipeline of knowledge databases
in PCGR contain numerous and complex procedures, with several quality
control and re-formatting steps, and and cannot be fully automated in
its present form. The version of all databases and key software elements
are outlined in each PCGR report.</em></p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
