<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Input files • PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Input files">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-7HRLE8GQRE"></script><script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){dataLayer.push(arguments);}
   gtag('js', new Date());

   gtag('config', 'G-7HRLE8GQRE');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/running.html">Running</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/input.html">Input files</a></li>
    <li><a class="dropdown-item" href="../articles/output.html">Output files</a></li>
    <li><a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a></li>
    <li><a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="../articles/developers.html">Developer notes</a></li>
    <li><a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Input files</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/main/vignettes/input.Rmd" class="external-link"><code>vignettes/input.Rmd</code></a></small>
      <div class="d-none name"><code>input.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h2>
<p>The PCGR workflow accepts three main input files:</p>
<ul>
<li>An unannotated, single-sample <a href="https://github.com/samtools/hts-specs#variant-calling-data-files" class="external-link">VCF
file</a> (&gt;= v4.2) with called somatic variants (SNVs/InDels)</li>
<li>A file with allele-specific copy number segments (tab-separated
values - TSV)</li>
<li>A file with transcript/gene expression levels (tab-separated values
- TSV)</li>
</ul>
<p>The input VCF is a required input file, while the somatic copy number
file and gene expression file are optional. The following arguments to
the <code>pcgr</code> command are used for input files:</p>
<ul>
<li>
<code>--input_vcf</code> (required argument to
<code>pcgr</code>)</li>
<li><code>--input_cna</code></li>
<li><code>--input_rna_expression</code></li>
</ul>
<p>In addition to these three main input files, the user can also opt to
provide a <a href="#panel-of-normals-pon-vcf">panel-of-normals VCF</a>
file (<code>--pon_vcf</code>) for tumor-only variant filtering, as well
as an input file with <a href="#germline-variants">CPSR-classified
germline variants (TSV)</a> (<code>--input_cpsr</code>).</p>
<div class="section level3">
<h3 id="vcf">VCF<a class="anchor" aria-label="anchor" href="#vcf"></a>
</h3>
<ul>
<li>We <strong>strongly</strong> recommend that the input VCF is
compressed and indexed using <a href="http://www.htslib.org/doc/bgzip.html" class="external-link">bgzip</a> and <a href="http://www.htslib.org/doc/tabix.html" class="external-link">tabix</a>
</li>
<li>If the input VCF contains <a href="https://glow.readthedocs.io/en/latest/etl/variant-splitter.html" class="external-link">multi-allelic
sites</a>, these will be subject to <a href="http://genome.sph.umich.edu/wiki/Vt#Decompose" class="external-link">decomposition</a>.
Either way, we encourage that users prepare the input VCF <em>without
the presence of multi-allelic sites</em>.</li>
<li>Variants used for reporting should be designated as
<code>PASS</code> in the VCF FILTER column. Variants denoted with
e.g. <code>Reject</code> as a FILTER value will not be subject to
analysis in PCGR. For records with undefined values in the FILTER column
(<code>'.'</code>), these will be considered as <code>PASS</code>
variants.</li>
</ul>
<div class="section level4">
<h4 id="formatting-of-variant-sequencing-depthallelic-support-dpaf">Formatting of variant sequencing depth/allelic support (DP/AF)<a class="anchor" aria-label="anchor" href="#formatting-of-variant-sequencing-depthallelic-support-dpaf"></a>
</h4>
<p>The representation of variant genotype data (allelic depth and
support in tumor vs. control sample) is usually formatted in the
genotype fields of a VCF file, on a per-sample basis. However,
considering the VCF output for the <a href="https://www.biostars.org/p/19104/" class="external-link">numerous somatic SNV/InDel
callers</a> that are in use, we have experienced a general lack of
uniformity for how this information is encoded in the genotype fields.
In order for PCGR to recognize this type of information robustly, we
currently require that you as a user encode this unambiguously in the
INFO field of your VCF file.</p>
<p>Shown below is how the VCF header for these entries should look like
in your input VCF, and how the corresponding variant data is encoded per
record:</p>
<p><strong>VCF header</strong></p>
<pre><code><span><span class="co">##INFO=&lt;ID=TVAF,Number=.,Type=Float,Description="Allelic fraction of alternative allele in tumor"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=TDP,Number=.,Type=Integer,Description="Read depth across variant site in tumor"&gt;</span></span></code></pre>
<p><strong>VCF records</strong></p>
<pre><code>#CHROM  POS ID  REF ALT QUAL    FILTER  INFO
1       23519660        .       G       C       .       PASS    TVAF=0.0980;TDP=51</code></pre>
<p>When you run PCGR, make sure to feed the name of the INFO tags to the
tool, i.e. using the <code>tumor_dp_tag</code>and
<code>tumor_af_tag</code> options:</p>
<pre><code>pcgr --tumor_dp_tag TDP --tumor_af_tag TVAF</code></pre>
<p>As an effort to support the users with this procedure, we are hoping
to establish a <em>VCF pre-processing script</em>, that accepts any VCF
(from any caller), and that will ensure that the file adheres to the
formatting requirements of PCGR. See also <a href="https://github.com/sigven/pcgr/issues/136" class="external-link">this issue</a>.</p>
</div>
<div class="section level4">
<h4 id="other-notes-regarding-input-vcf">Other notes regarding input VCF<a class="anchor" aria-label="anchor" href="#other-notes-regarding-input-vcf"></a>
</h4>
<ul>
<li>PCGR generates a number of VCF INFO annotation tags that are
appended to the query VCF. We therefore encourage the users to submit
query VCF files that <em>have not</em> been subject to annotations by
other means, but rather a VCF file that comes directly from variant
calling. If not, there are likely to be INFO tags in the query VCF file
that coincide with those produced by PCGR.</li>
<li>Note that you can preserve particular tags in the INFO field towards
the TSV output of PCGR. Sometimes, it can be convenient to investigate
particular properties of the variants (encoded in the VCF) against
functional annotations (as provided by PCGR). To achieve this, use the
option <code>--retained_info_tags &lt;TAG1&gt;,&lt;TAG2&gt;</code>
etc.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="panel-of-normals-vcf">Panel-of-normals VCF<a class="anchor" aria-label="anchor" href="#panel-of-normals-vcf"></a>
</h3>
<p>The user can submit a panel-of-normals (PoN) VCF file to PCGR,
indicating variants that should be excluded/filtered from the input VCF
in a <strong>tumor-only setting</strong> (i.e. option
<code>--tumor_only</code>).</p>
<ul>
<li><code>--pon_vcf &lt;VCF_FILE&gt;</code></li>
</ul>
<p>The PoN VCF file needs to contain the following.</p>
<ol style="list-style-type: decimal">
<li>The following description needs to be present in the VCF
header:</li>
</ol>
<pre><code><span><span class="co">##INFO=&lt;ID=PANEL_OF_NORMALS,Number=0,Type=Flag,Description="Overlap with germline call among panel of normals"&gt;</span></span></code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Each record in the PoN VCF file needs to contain the
<code>PANEL_OF_NORMALS</code> flag in the INFO field</li>
</ol>
</div>
<div class="section level3">
<h3 id="copy-number-segments---allele-specific">Copy number segments - allele-specific<a class="anchor" aria-label="anchor" href="#copy-number-segments---allele-specific"></a>
</h3>
<p>A tab-separated values file with allele-specific copy number
aberrations <strong>MUST</strong> contain the following four
columns:</p>
<ul>
<li><code>Chromosome</code></li>
<li><code>Start</code></li>
<li><code>End</code></li>
<li><code>nMajor</code></li>
<li><code>nMinor</code></li>
</ul>
<p>Here, <em>Chromosome</em>, <em>Start</em>, and <em>End</em> denote
the chromosomal segment, while <em>nMajor</em> and <em>nMinor</em>
denote major and minor copy number for particular segment, a standard
output of somatic copy number alteration callers. Note that coordinates
must be <strong>one-based</strong> (i.e. chromosomes start at 1, not 0),
and that copy numbers should be formatted as whole integers, not
floating values. Below shows the initial part of a copy number segment
file that is formatted correctly according to PCGR’s requirements:</p>
<pre><code>  Chromosome    Start   End nMajor nMinor
  1 3218329 3550598 1 0
  1 3552451 4593614 1 1
  1 4593663 6433129 2 1</code></pre>
<p>Importantly, you can configure predefined thresholds for segments
that are considered to be high-level copy-number amplifications through
the following option:</p>
<ul>
<li>
<code>--n_copy_gain</code>: minimum (total) copy number for a
segment to be considered an amplification, <em>default 6</em>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="germline-variants">Germline variants<a class="anchor" aria-label="anchor" href="#germline-variants"></a>
</h3>
<p>The user can submit a file with germline variants processed and
classified with <a href="https://sigven.github.io/cpsr" class="external-link">CPSR</a>, which
caters for an integration of germline and somatic findings in the output
report. This file corresponds to output file
<code>&lt;sample_id&gt;.cpsr.&lt;genome_assembly&gt;.classification.tsv.gz</code>
from the CPSR pipeline. Make sure the genome assembly is the same as the
one used for the somatic variant input files.</p>
</div>
<div class="section level3">
<h3 id="gene-expression">Gene expression<a class="anchor" aria-label="anchor" href="#gene-expression"></a>
</h3>
<p>The user can submit a file with bulk gene/transcript expression data
to PCGR, indicating the relative expression levels of genes in the query
sample. PCGR may conduct an expression outlier analysis (compared
against other cohorts of samples), and also perform an RNA expression
similarity analysis, i.e. correlation of expression profile with other
tumor samples.</p>
<p>The tab-separated values file with gene expression estimates
<strong>MUST</strong> contain the following two columns:</p>
<ul>
<li><code>TargetID</code></li>
<li><code>TPM</code></li>
</ul>
<p>The <code>TPM</code> column indicates the expression level of a given
target, and must be provided as <em>Transcripts per million</em> (not
log-transformed), a relatively common output measure from bulk RNA-seq
pipelines. The <code>TargetID</code> should list transcript or gene
identifiers, preferably at the transcript level. Permitted identifier
types including Ensembl/RefSeq identifiers for transcript-level
expression estimates, and gene symbols and Ensembl gene identifiers for
gene-level expression estimates.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
