<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pcgrr">
<title>Input files • PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Input files">
<meta property="og:description" content="pcgrr">
<meta property="og:image" content="https://sigven.github.io/pcgr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Intro</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/running.html">Running</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/input.html">Input files</a>
    <a class="dropdown-item" href="../articles/output.html">Output files</a>
    <a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a>
    <a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a>
    <a class="dropdown-item" href="../articles/developers.html">Developer notes</a>
    <a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/faq.html">FAQ</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Input files</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/HEAD/../vignettes/input.Rmd" class="external-link"><code>../vignettes/input.Rmd</code></a></small>
      <div class="d-none name"><code>input.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h2>
<p>The PCGR workflow accepts two main input files:</p>
<ul>
<li>An unannotated, single-sample <a href="https://github.com/samtools/hts-specs#variant-calling-data-files" class="external-link">VCF
file</a> (&gt;= v4.2) with called somatic variants (SNVs/InDels)</li>
<li>A copy number segment file</li>
</ul>
<p>The input VCF is a required input file, while the somatic copy number
file is optional. The following arguments to the <code>pcgr</code>
command are used for input files:</p>
<ul>
<li>
<code>--input_vcf</code> (required argument to
<code>pcgr</code>)</li>
<li><code>--input_cna</code></li>
</ul>
<p>In addition to these two main input files, the user can also opt to
provide a <a href="#panel-of-normals-pon-vcf">panel-of-normals VCF</a>
file, as well as a <a href="#cpsr-report">CPSR report (JSON)</a>, as
input.</p>
<div class="section level3">
<h3 id="vcf">VCF<a class="anchor" aria-label="anchor" href="#vcf"></a>
</h3>
<ul>
<li>We <strong>strongly</strong> recommend that the input VCF is
compressed and indexed using <a href="http://www.htslib.org/doc/bgzip.html" class="external-link">bgzip</a> and <a href="http://www.htslib.org/doc/tabix.html" class="external-link">tabix</a>
</li>
<li>If the input VCF contains <a href="https://glow.readthedocs.io/en/latest/etl/variant-splitter.html" class="external-link">multi-allelic
sites</a>, these will be subject to <a href="http://genome.sph.umich.edu/wiki/Vt#Decompose" class="external-link">decomposition</a>.
Either way, we encourage that users prepare the input VCF <em>without
the presence of multi-allelic sites</em>.</li>
<li>Variants used for reporting should be designated as
<code>PASS</code> in the VCF FILTER column. Variants denoted with
e.g. <code>Reject</code> as a FILTER value will not be subject to
analysis in PCGR. For records with undefined values in the FILTER column
(<code>'.'</code>), these will be considered as <code>PASS</code>
variants.</li>
</ul>
<div class="section level4">
<h4 id="formatting-of-allelic-depthsupport-dpad">Formatting of allelic depth/support (DP/AD)<a class="anchor" aria-label="anchor" href="#formatting-of-allelic-depthsupport-dpad"></a>
</h4>
<p>The representation of variant genotype data (allelic depth and
support in tumor vs. control sample) is usually formatted in the
genotype fields of a VCF file, on a per-sample basis. However,
considering the VCF output for the <a href="https://www.biostars.org/p/19104/" class="external-link">numerous somatic SNV/InDel
callers</a> that are in use, we have experienced a general lack of
uniformity for how this information is encoded in the genotype fields.
In order for PCGR to recognize this type of information robustly, we
currently require that you encode this unambiguously in the INFO field
of your VCF file.</p>
<p>Shown below is how the VCF header for these entries should look like
in your input VCF, and how the corresponding variant data is encoded per
record:</p>
<p><strong>VCF header</strong></p>
<pre><code><span><span class="co">##INFO=&lt;ID=TVAF,Number=.,Type=Float,Description="Allelic fraction of alternative allele in tumor"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=TDP,Number=.,Type=Integer,Description="Read depth across variant site in tumor"&gt;</span></span></code></pre>
<p><strong>VCF records</strong></p>
<pre><code>#CHROM  POS ID  REF ALT QUAL    FILTER  INFO
1       23519660        .       G       C       .       PASS    TVAF=0.0980;TDP=51</code></pre>
<p>When you run PCGR, make sure to feed the name of the INFO tags to the
tool, i.e. using the <code>tumor_dp_tag</code>and
<code>tumor_af_tag</code> options:</p>
<pre><code>pcgr --tumor_dp_tag TDP --tumor_af_tag TVAF</code></pre>
<p>As an effort to support the users with this procedure, we are
currently trying to establish a <em>VCF pre-processing script</em>, that
accepts any VCF (from any caller), and that will ensure that the file
adheres to the formatting requirements of PCGR. See also <a href="https://github.com/sigven/pcgr/issues/136" class="external-link">this issue</a>.</p>
</div>
<div class="section level4">
<h4 id="other-notes-regarding-input-vcf">Other notes regarding input VCF<a class="anchor" aria-label="anchor" href="#other-notes-regarding-input-vcf"></a>
</h4>
<ul>
<li>PCGR generates a number of VCF INFO annotation tags that are
appended to the query VCF. We therefore encourage the users to submit
query VCF files that have not been subject to annotations by other
means, but rather a VCF file that comes directly from variant calling.
If not, there are likely to be INFO tags in the query VCF file that
coincide with those produced by PCGR.</li>
<li>Note that you can preserve particular tags in the INFO field towards
the TSV output of PCGR. Sometimes, it can be convenient to investigate
particular properties of the variants (encoded in the VCF) against
functional annotations (as provided by PCGR). To achieve this, use the
option
<code>--preserved_info_tags &lt;TAG1&gt;,&lt;TAG2&gt; etc</code>.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="panel-of-normals-vcf">Panel-of-normals VCF<a class="anchor" aria-label="anchor" href="#panel-of-normals-vcf"></a>
</h3>
<p>The user can submit a panel-of-normals (PoN) VCF file to PCGR,
indicating variants that should be excluded/filtered from the input VCF
in a <strong>tumor-only setting</strong> (i.e. option
<code>--tumor_only</code>).</p>
<ul>
<li><code>--pon_vcf &lt;VCF_FILE&gt;</code></li>
</ul>
<p>The PoN VCF file needs to contain the following.</p>
<ol style="list-style-type: decimal">
<li>The following description needs to be present in the VCF
header:</li>
</ol>
<pre><code><span><span class="co">##INFO=&lt;ID=PANEL_OF_NORMALS,Number=0,Type=Flag,Description="Overlap with germline call among panel of normals"&gt;</span></span></code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Each record in the PoN VCF file needs to contain the
<code>PANEL_OF_NORMALS</code> flag in the INFO field</li>
</ol>
</div>
<div class="section level3">
<h3 id="copy-number-segments">Copy number segments<a class="anchor" aria-label="anchor" href="#copy-number-segments"></a>
</h3>
<p>The tab-separated values file with copy number aberrations
<strong>MUST</strong> contain the following four columns:</p>
<ul>
<li><code>Chromosome</code></li>
<li><code>Start</code></li>
<li><code>End</code></li>
<li><code>Segment_Mean</code></li>
</ul>
<p>Here, <em>Chromosome</em>, <em>Start</em>, and <em>End</em> denote
the chromosomal segment, and <em>Segment_Mean</em> denotes the log(2)
ratio for a particular segment, which is a common output of somatic copy
number alteration callers. Note that coordinates must be
<strong>one-based</strong> (i.e. chromosomes start at 1, not 0). Below
shows the initial part of a copy number segment file that is formatted
correctly according to PCGR’s requirements:</p>
<pre><code>  Chromosome    Start   End Segment_Mean
  1 3218329 3550598 0.0024
  1 3552451 4593614 0.1995
  1 4593663 6433129 -1.0277</code></pre>
<p>Importantly, you can configure predefined log(2) thresholds for
segments that are considered to be high-level copy-number amplifications
and homozygous deletions, through the following options:</p>
<ul>
<li>
<code>--logr_gain</code>: minimum log(2) value for a segment to be
considered an amplification, <em>default 0.8</em>
</li>
<li>
<code>--logr_homdel</code>: maximum log(2) value for a segment to be
considered a homozygous deletion, <em>default -0.8</em>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="cpsr-report">CPSR report<a class="anchor" aria-label="anchor" href="#cpsr-report"></a>
</h3>
<p>One may feed the compressed JSON file that is output from a cancer
predisposition analysis with CPSR (<a href="https://github.com/sigven/cpsr" class="external-link uri">https://github.com/sigven/cpsr</a>) into PCGR through the
<code>--cpsr_report</code> option. This will result in a dedicated
germline section of the PCGR report.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
