<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pcgrr">
<title>Output files â€¢ PCGR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Output files">
<meta property="og:description" content="pcgrr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Intro</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/running.html">Running</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/input.html">Input files</a>
    <a class="dropdown-item" href="../articles/output.html">Output files</a>
    <a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a>
    <a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a>
    <a class="dropdown-item" href="../articles/developers.html">Developer notes</a>
    <a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/faq.html">FAQ</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Output files</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/HEAD/../vignettes/output.Rmd" class="external-link"><code>../vignettes/output.Rmd</code></a></small>
      <div class="d-none name"><code>output.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="output-files">Output files<a class="anchor" aria-label="anchor" href="#output-files"></a>
</h2>
<p>PCGR generates multiple output files, including interactive HTML
reports and pure variant annotation files (tab-separated values
(TSV)).</p>
<div class="section level3">
<h3 id="html-report---dashboard">HTML report - dashboard<a class="anchor" aria-label="anchor" href="#html-report---dashboard"></a>
</h3>
<p>An interactive and tier-structured HTML report (dashboard-like) that
shows the most relevant findings in the query cancer genome has the
following naming convention:</p>
<ul>
<li>
<code>&lt;sample_id&gt;.pcgr_acmg.&lt;genome_assembly&gt;.flexdb.html</code>
<ul>
<li>The <strong>sample_id</strong> is provided as input by the user, and
reflects a unique identifier of the tumor-normal sample pair to be
analyzed.</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="html-report---rmarkdown">HTML report - rmarkdown<a class="anchor" aria-label="anchor" href="#html-report---rmarkdown"></a>
</h3>
<p>An interactive and tier-structured HTML report that shows the most
relevant findings in the query cancer genome has the following naming
convention:</p>
<ul>
<li>
<code>&lt;sample_id&gt;.pcgr_acmg.&lt;genome_assembly&gt;.html</code>
<ul>
<li>The <strong>sample_id</strong> is provided as input by the user, and
reflects a unique identifier of the tumor-normal sample pair to be
analyzed.</li>
</ul>
</li>
</ul>
<p>The report is structured in seven main sections, described in more
detail below:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Report and query settings</strong>
<ul>
<li>Lists key configurations provided by user</li>
</ul>
</li>
<li>
<strong>Main results</strong>
<ul>
<li>Nine value boxes that highlight the main properties of the tumor
sample:
<ol style="list-style-type: decimal">
<li>Tumor purity - as provided by the user</li>
<li>Tumor ploidy - as provided by the user</li>
<li>Mutational signatures - two most prevalent signatures (other than
aging)</li>
<li>Tier 1 variants (top four)</li>
<li>Tier 2 variants (top four)</li>
<li>Tumor mutational burden (TMB)</li>
<li>Microsatellite instability (MSI) status</li>
<li>Somatic copy number aberrations of clinical significance</li>
<li>Kataegis events</li>
</ol>
</li>
</ul>
</li>
<li>
<strong>Somatic SNVs/InDels</strong>
<ul>
<li>
<em>Tumor mutational burden (TMB)</em>
<ul>
<li>given a coding target region size specified by the user (ideally the
<strong>callable target size</strong>), an estimate of the mutational
burden is provided</li>
<li>The estimated TMB is shown in the context of TMB distributions from
different primary sites in TCGA</li>
</ul>
</li>
<li>
<em>Tier &amp; variant statistics</em>
<ul>
<li>indicate total variant numbers across variant types, coding types
and tiers</li>
</ul>
</li>
<li>
<em>Global distribution - allelic support</em>
<ul>
<li>distribution (histogram) of variant allelic support for somatic
variants (will only be present in the report if specific fields in input
VCF is defined and specified by the user)</li>
</ul>
</li>
<li>
<em>Global variant browser</em>
<ul>
<li>permits exploration of the whole SNV/InDel dataset by filtering
along several dimensions (call confidence, variant sequencing
depth/support, variant consequence etc.)</li>
</ul>
</li>
<li>
<em>Tier tables</em>
<ul>
<li>Variants are organized into five (tier 1-4 + noncoding) interactive
datatables) according to clinical utility</li>
<li>Contents of the tier tables are outlined below</li>
</ul>
</li>
</ul>
</li>
<li>
<strong>Somatic CNAs</strong>
<ul>
<li>
<em>Segments - amplifications and homozygous deletions</em>
<ul>
<li>Based on user-defined/default log-ratio thresholds of gains/losses,
the whole CNA dataset can be navigated further through filters:</li>
<li>cytoband</li>
<li>type of CNA event - <em>focal</em> (less than 25% of chromosome arm
affected) or <em>broad</em>
</li>
<li>log ratio</li>
</ul>
</li>
<li>
<em>Proto-oncogenes subject to copy number amplifications</em>
<ul>
<li>Datatable listing known proto-oncogenes covered by
user-defined/default amplifications and potential targeted
therapies</li>
</ul>
</li>
<li>
<em>Tumor suppressor genes subject to homozygous deletions</em>
<ul>
<li>Datatable listing known tumor suppressor genes covered by
user-defined/default losses and potential targeted therapies</li>
</ul>
</li>
<li><em>Other drug targets subject to copy number
aplifications</em></li>
<li>
<em>Copy number aberrations as biomarkers for prognosis, diagnosis,
and drug response (both of <b>strong</b> significance and
<b>potential</b> significance)</em>
<ul>
<li>Interactive data table where the user can navigate aberrations
acting as biomarkers across therapeutic contexts, tumor subtypes,
evidence levels etc</li>
</ul>
</li>
</ul>
</li>
<li>
<strong>MSI status</strong>
<ul>
<li>Indicates predicted microsatellite stability from the somatic
mutation profile and supporting evidence (details of the underlying MSI
statistical classifier can be found <a href="http://rpubs.com/sigven/msi_classification_v3" class="external-link">here</a>)</li>
<li>The MSI classifier was trained on TCGA exome samples.</li>
</ul>
</li>
<li>
<strong>Mutational signatures</strong>
<ul>
<li>Estimation of relative contribution of <a href="http://cancer.sanger.ac.uk/cosmic/signatures" class="external-link">67 known mutational
signatures</a> in tumor sample (using <a href="https://github.com/raerose01/deconstructSigs" class="external-link">MutationalPatterns</a>
as the underlying framework)</li>
<li>Datatable with signatures and proposed underlying etiologies</li>
</ul>
</li>
<li>
<strong>Kataegis events</strong>
<ul>
<li>Detection of potential kataegis events using the algorithm outlined
in the <a href="https://github.com/MeichunCai/KataegisPortal" class="external-link">Kataegis
Portal</a>
</li>
</ul>
</li>
<li>
<strong>Germline findings</strong>
<ul>
<li>Main findings from CPSR report</li>
</ul>
</li>
<li>
<strong>Clinial trials</strong>
<ul>
<li>Interactive table with annotated, molecularly targeted clinical
trials from <a href="https//clinicaltrials.gov">clinicaltrials.gov</a>
in the relevant tumor type</li>
</ul>
</li>
<li>
<strong>Documentation</strong>
<ul>
<li>Annotation resources - software, databases and tools with version
information</li>
<li>Report content</li>
<li>References - supporting scientific literature (key report
elements)</li>
</ul>
</li>
</ol>
<div class="section level4">
<h4 id="interactive-datatables">Interactive datatables<a class="anchor" aria-label="anchor" href="#interactive-datatables"></a>
</h4>
<p>The interactive datatables contain a number of hyperlinked
annotations similar to those defined for the annotated VCF file,
including the following:</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="84%">
</colgroup>
<thead><tr class="header">
<th>Annotation</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>SYMBOL</code></td>
<td>Gene symbol (Entrez/NCBI)</td>
</tr>
<tr class="even">
<td><code>PROTEIN_CHANGE</code></td>
<td>Amino acid change (VEP)</td>
</tr>
<tr class="odd">
<td><code>CANCER_TYPE</code></td>
<td>Biomarker (tier 1/2): associated cancer type</td>
</tr>
<tr class="even">
<td><code>EVIDENCE_LEVEL</code></td>
<td>Biomarker (tier 1/2): evidence level (A,B,C,D,E)</td>
</tr>
<tr class="odd">
<td><code>CLINICAL_SIGNIFICANCE</code></td>
<td>Biomarker (tier 1/2): drug sensitivity, poor/better outcome etc</td>
</tr>
<tr class="even">
<td><code>EVIDENCE_TYPE</code></td>
<td>Biomarker (tier 1/2): predictive/diagnostic/therapeutic</td>
</tr>
<tr class="odd">
<td><code>DISEASE_ONTOLOGY_ID</code></td>
<td>Biomarker (tier 1/2): associated cancer type (Disease Ontology)</td>
</tr>
<tr class="even">
<td><code>EVIDENCE_DIRECTION</code></td>
<td>Biomarker (tier 1/2): supports/does not support</td>
</tr>
<tr class="odd">
<td><code>DESCRIPTION</code></td>
<td>Biomarker (tier 1/2): description</td>
</tr>
<tr class="even">
<td><code>VARIANT_ORIGIN</code></td>
<td>Biomarker (tier 1/2): variant origin (germline/somatic)</td>
</tr>
<tr class="odd">
<td><code>BIOMARKER_MAPPING</code></td>
<td>Biomarker (tier 1/2): accuracy of genomic mapping
(exact,codon,exon)</td>
</tr>
<tr class="even">
<td><code>CITATIO</code></td>
<td>Biomarker (tier 1/2): supporting literature</td>
</tr>
<tr class="odd">
<td><code>THERAPEUTIC_CONTEXT</code></td>
<td>Biomarker (tier 1/2): associated drugs</td>
</tr>
<tr class="even">
<td><code>RATING</code></td>
<td>Biomarker (tier 1/2): trust rating from 1 to 5 (CIVIC)</td>
</tr>
<tr class="odd">
<td><code>GENE_NAME</code></td>
<td>gene name description (Entrez/NCBI)</td>
</tr>
<tr class="even">
<td><code>PROTEIN_DOMAIN</code></td>
<td>PFAM protein domain</td>
</tr>
<tr class="odd">
<td><code>PROTEIN_FEATURE</code></td>
<td>UniProt feature overlapping variant site</td>
</tr>
<tr class="even">
<td><code>CDS_CHANGE</code></td>
<td>Coding sequence change</td>
</tr>
<tr class="odd">
<td><code>MUTATION_HOTSPOT</code></td>
<td>Known cancer mutation hotspot</td>
</tr>
<tr class="even">
<td><code>MUTATION_HOTSPOT_CANCERTYPE</code></td>
<td>Hotspot-associated cancer types</td>
</tr>
<tr class="odd">
<td><code>TCGA_FREQUENCY</code></td>
<td>Frequency of variant in TCGA cohorts</td>
</tr>
<tr class="even">
<td><code>ICGC_PCAWG_OCCURRENCE</code></td>
<td>Frequency of variant in ICGC-PCAWG cohorts</td>
</tr>
<tr class="odd">
<td><code>DOCM_LITERATURE</code></td>
<td>Literature links - DoCM</td>
</tr>
<tr class="even">
<td><code>DOCM_DISEASE</code></td>
<td>Associated diseases - DoCM</td>
</tr>
<tr class="odd">
<td><code>OPENTARGETS_RANK</code></td>
<td>Strength (max across all phenotypes) of phenotype association
according to the Open Targets Platform</td>
</tr>
<tr class="even">
<td><code>OPENTARGETS_ASSOCIATIONS</code></td>
<td>Phenotype associations with the gene retrieved from the Open Targets
Platform</td>
</tr>
<tr class="odd">
<td><code>INTOGEN_DRIVER_MUT</code></td>
<td>predicted driver mutation - IntOGen</td>
</tr>
<tr class="even">
<td><code>CONSEQUENCE</code></td>
<td>VEP consequence (primary transcript)</td>
</tr>
<tr class="odd">
<td><code>HGVSc</code></td>
<td>from VEP</td>
</tr>
<tr class="even">
<td><code>HGVSp</code></td>
<td>from VEP</td>
</tr>
<tr class="odd">
<td><code>NCBI_REFSEQ</code></td>
<td>Transcript accession ID(s) (NCBI RefSeq)</td>
</tr>
<tr class="even">
<td><code>ONCOGENE</code></td>
<td>Predicted as proto-oncogene from CancerMine/NCG</td>
</tr>
<tr class="odd">
<td><code>CANCERGENE_SUPPORT</code></td>
<td>Links to underlying publications (CancerMine) that support
oncogenic/tumor suppressive role of gene</td>
</tr>
<tr class="even">
<td><code>TUMOR_SUPPRESSOR</code></td>
<td>Predicted as tumor suppressor gene from CancerMine/NCG</td>
</tr>
<tr class="odd">
<td><code>PREDICTED_EFFECT</code></td>
<td>Effect predictions from dbNSFP</td>
</tr>
<tr class="even">
<td><code>VEP_ALL_CSQ</code></td>
<td>All VEP transcript block consequences</td>
</tr>
<tr class="odd">
<td><code>DBSNP</code></td>
<td>dbSNP rsID</td>
</tr>
<tr class="even">
<td><code>COSMIC</code></td>
<td>Cosmic mutation IDs</td>
</tr>
<tr class="odd">
<td><code>CLINVAR</code></td>
<td>ClinVar variant origin and associated phenotypes</td>
</tr>
<tr class="even">
<td><code>CANCER_ASSOCIATIONS</code></td>
<td>Gene-associated cancer types from DisGenet</td>
</tr>
<tr class="odd">
<td><code>TARGETED_DRUGS</code></td>
<td>Targeted drugs from Open Targets Platform /ChEMBL</td>
</tr>
<tr class="even">
<td><code>KEGG_PATHWAY</code></td>
<td>Gene-associated pathways from KEGG</td>
</tr>
<tr class="odd">
<td><code>CALL_CONFIDENCE</code></td>
<td>Variant confidence (as set by user in input VCF)</td>
</tr>
<tr class="even">
<td><code>DP_TUMOR</code></td>
<td>Variant sequencing depth in tumor (as set by user in input VCF)</td>
</tr>
<tr class="odd">
<td><code>AF_TUMOR</code></td>
<td>Variant allelic fraction in tumor (as set by user in input VCF)</td>
</tr>
<tr class="even">
<td><code>DP_CONTROL</code></td>
<td>Variant sequencing depth in control sample (as set by user in input
VCF)</td>
</tr>
<tr class="odd">
<td><code>AF_CONTROL</code></td>
<td>Variant allelic fraction in control sample (as set by user in input
VCF)</td>
</tr>
<tr class="even">
<td><code>GENOMIC_CHANGE</code></td>
<td>Variant ID</td>
</tr>
<tr class="odd">
<td><code>GENOME_VERSION</code></td>
<td>Genome assembly</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="example-reports">Example reports<a class="anchor" aria-label="anchor" href="#example-reports"></a>
</h4>
<p><a href="https://doi.org/10.5281/zenodo.5045309" class="external-link"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.5045309.svg" alt="DOI"></a></p>
<p>The HTML reports have been tested using the following browsers:</p>
<ul>
<li>Safari (Version 14.1.1 (16611.2.7.1.4))</li>
<li>Mozilla Firefox (83.0)</li>
<li>Google Chrome (Version 90.0.4430.212 (Official Build) (x86_64))</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="json">JSON<a class="anchor" aria-label="anchor" href="#json"></a>
</h3>
<p>A compressed JSON file that stores all the essential content of the
report is provided.</p>
<p>This file will easen the process of extracting particular parts of
the report for further analysis or integration with other workflows. The
JSON contains two main objects, <em>metadata</em> and <em>content</em>,
where the former contains information about the settings, data versions,
and the latter contains the various sections of the report.</p>
</div>
<div class="section level3">
<h3 id="snvsindels">SNVs/InDels<a class="anchor" aria-label="anchor" href="#snvsindels"></a>
</h3>
<div class="section level4">
<h4 id="variant-call-format---vcf">1. Variant call format - VCF<a class="anchor" aria-label="anchor" href="#variant-call-format---vcf"></a>
</h4>
<p>A VCF file containing annotated, somatic calls (single nucleotide
variants and insertion/deletions) is generated with the following naming
convention:</p>
<ul>
<li>
<code>&lt;sample_id&gt;.pcgr_acmg.&lt;genome_assembly&gt;.vcf.gz</code>
<ul>
<li>The <strong>sample_id</strong> is provided as input by the user, and
reflects a unique identifier of the tumor-normal sample pair to be
analyzed. Following common standards, the annotated VCF file is
compressed with <a href="http://www.htslib.org/doc/bgzip.html" class="external-link">bgzip</a>
and indexed with <a href="http://www.htslib.org/doc/tabix.html" class="external-link">tabix</a>. Below follows a
description of all annotations/tags present in the VCF INFO column after
processing with the PCGR annotation pipeline:</li>
</ul>
</li>
</ul>
<div class="section level5">
<h5 id="vep-consequence-annotations">
<em>VEP consequence annotations</em><a class="anchor" aria-label="anchor" href="#vep-consequence-annotations"></a>
</h5>
<table class="table">
<colgroup>
<col width="27%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>Tag</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>CSQ</code></td>
<td>Complete consequence annotations from VEP. Format (separated by a
<code>|</code>): <code>Allele</code>, <code>Consequence</code>,
<code>IMPACT</code>, <code>SYMBOL</code>, <code>Gene</code>,
<code>Feature_type</code>, <code>Feature</code>, <code>BIOTYPE</code>,
<code>EXON</code>, <code>INTRON</code>, <code>HGVSc</code>,
<code>HGVSp</code>, <code>cDNA_position</code>,
<code>CDS_position</code>, <code>Protein_position</code>,
<code>Amino_acids</code>, <code>Codons</code>,
<code>Existing_variation</code>, <code>ALLELE_NUM</code>,
<code>DISTANCE</code>, <code>STRAND</code>, <code>FLAGS</code>,
<code>PICK</code>, <code>VARIANT_CLASS</code>,
<code>SYMBOL_SOURCE</code>, <code>HGNC_ID</code>,
<code>CANONICAL</code>, <code>MANE_SELECT</code>,
<code>MANE_PLUS_CLINICAL</code>, <code>TSL</code>, <code>APPRIS</code>,
<code>CCDS</code>, <code>ENSP</code>, <code>SWISSPROT</code>,
<code>TREMBL</code>, <code>UNIPARC</code>, <code>UNIPROT_ISOFORM</code>,
<code>RefSeq</code>, <code>DOMAINS</code>, <code>HGVS_OFFSET</code>,
<code>AF</code>, <code>AFR_AF</code>, <code>AMR_AF</code>,
<code>EAS_AF</code>, <code>EUR_AF</code>, <code>SAS_AF</code>,
<code>gnomAD_AF</code>, <code>gnomAD_AFR_AF</code>,
<code>gnomAD_AMR_AF</code>, <code>gnomAD_ASJ_AF</code>,
<code>gnomAD_EAS_AF</code>, <code>gnomAD_FIN_AF</code>,
<code>gnomAD_NFE_AF</code>, <code>gnomAD_OTH_AF</code>,
<code>gnomAD_SAS_AF</code>, <code>CLIN_SIG</code>, <code>SOMATIC</code>,
<code>PHENO</code>, <code>CHECK_REF</code>, <code>MOTIF_NAME</code>,
<code>MOTIF_POS</code>, <code>HIGH_INF_POS</code>,
<code>MOTIF_SCORE_CHANGE</code>, <code>TRANSCRIPTION_FACTORS</code>,
<code>NearestExonJB</code>
</td>
</tr>
<tr class="even">
<td><code>Consequence</code></td>
<td>Impact modifier for the consequence type (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>Gene</code></td>
<td>Ensembl stable ID of affected gene (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>Feature_type</code></td>
<td>Type of feature. Currently one of Transcript, RegulatoryFeature,
MotifFeature (picked by VEPâ€™s <code>--flag_pick_allele</code>
option)</td>
</tr>
<tr class="odd">
<td><code>Feature</code></td>
<td>Ensembl stable ID of feature (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>cDNA_position</code></td>
<td>Relative position of base pair in cDNA sequence (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>CDS_position</code></td>
<td>Relative position of base pair in coding sequence (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>CDS_CHANGE</code></td>
<td>Coding, transcript-specific sequence annotation (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>AMINO_ACID_START</code></td>
<td>Protein position indicating absolute start of amino acid altered
(fetched from <code>Protein_position</code>)</td>
</tr>
<tr class="even">
<td><code>AMINO_ACID_END</code></td>
<td>Protein position indicating absolute end of amino acid altered
(fetched from <code>Protein_position</code>)</td>
</tr>
<tr class="odd">
<td><code>Protein_position</code></td>
<td>Relative position of amino acid in protein (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>Amino_acids</code></td>
<td>Only given if the variant affects the protein-coding sequence
(picked by VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>Codons</code></td>
<td>The alternative codons with the variant base in upper case (picked
by VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>IMPACT</code></td>
<td>Impact modifier for the consequence type (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>VARIANT_CLASS</code></td>
<td>Sequence Ontology variant class (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>SYMBOL</code></td>
<td>Gene symbol (picked by VEPâ€™s <code>--flag_pick_allele</code>
option)</td>
</tr>
<tr class="odd">
<td><code>SYMBOL_ENTREZ</code></td>
<td>Official gene symbol as provided by NCBIâ€™s Entrez gene</td>
</tr>
<tr class="even">
<td><code>SYMBOL_SOURCE</code></td>
<td>The source of the gene symbol (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>STRAND</code></td>
<td>The DNA strand (1 or -1) on which the transcript/feature lies
(picked by VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>ENSP</code></td>
<td>The Ensembl protein identifier of the affected transcript (picked by
VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>FLAGS</code></td>
<td>Transcript quality flags: <code>cds_start_NF</code>: CDS 5â€™,
incomplete <code>cds_end_NF</code>: CDS 3â€™ incomplete (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>SWISSPROT</code></td>
<td>Best match UniProtKB/Swiss-Prot accession of protein product (picked
by VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>TREMBL</code></td>
<td>Best match UniProtKB/TrEMBL accession of protein product (picked by
VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>UNIPARC</code></td>
<td>Best match UniParc accession of protein product (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>HGVSc</code></td>
<td>The HGVS coding sequence name (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>HGVSp</code></td>
<td>The HGVS protein sequence name (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>HGVSp_short</code></td>
<td>The HGVS protein sequence name, short version (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>HGVS_OFFSET</code></td>
<td>Indicates by how many bases the HGVS notations for this variant have
been shifted (picked by VEPâ€™s <code>--flag_pick_allele</code>
option)</td>
</tr>
<tr class="odd">
<td><code>NearestExonJB</code></td>
<td>VEP plugin that finds nearest exon junction for a coding sequence
variant. Format: Ensembl exon identifier+distanceto exon
boundary+boundary type(start/end)+exon length</td>
</tr>
<tr class="even">
<td><code>MOTIF_NAME</code></td>
<td>The source and identifier of a transcription factor binding profile
aligned at this position (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>MOTIF_POS</code></td>
<td>The relative position of the variation in the aligned TFBP (picked
by VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>HIGH_INF_POS</code></td>
<td>A flag indicating if the variant falls in a high information
position of a transcription factor binding profile (TFBP) (picked by
VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>MOTIF_SCORE_CHANGE</code></td>
<td>The difference in motif score of the reference and variant sequences
for the TFBP (picked by VEPâ€™s <code>--flag_pick_allele</code>
option)</td>
</tr>
<tr class="even">
<td><code>CELL_TYPE</code></td>
<td>List of cell types and classifications for regulatory feature
(picked by VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>CANONICAL</code></td>
<td>A flag indicating if the transcript is denoted as the canonical
transcript for this gene (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>CCDS</code></td>
<td>The CCDS identifier for this transcript, where applicable (picked by
VEPâ€™s <code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>INTRON</code></td>
<td>The intron number (out of total number) (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>EXON</code></td>
<td>The exon number (out of total number) (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>LAST_EXON</code></td>
<td>Logical indicator for last exon of transcript (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>LAST_INTRON</code></td>
<td>Logical indicator for last intron of transcript (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>INTRON_POSITION</code></td>
<td>Relative position of intron variant to nearest exon/intron junction
(NearestExonJB VEP plugin)</td>
</tr>
<tr class="even">
<td><code>EXON_POSITION</code></td>
<td>Relative position of exon variant to nearest intron/exon junction
(NearestExonJB VEP plugin)</td>
</tr>
<tr class="odd">
<td><code>DISTANCE</code></td>
<td>Shortest distance from variant to transcript (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="even">
<td><code>BIOTYPE</code></td>
<td>Biotype of transcript or regulatory feature (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>TSL</code></td>
<td>Transcript support level (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)&gt;</td>
</tr>
<tr class="even">
<td><code>PUBMED</code></td>
<td>PubMed ID(s) of publications that cite existing variant - VEP</td>
</tr>
<tr class="odd">
<td><code>PHENO</code></td>
<td>Indicates if existing variant is associated with a phenotype,
disease or trait - VEP</td>
</tr>
<tr class="even">
<td><code>GENE_PHENO</code></td>
<td>Indicates if overlapped gene is associated with a phenotype, disease
or trait - VEP</td>
</tr>
<tr class="odd">
<td><code>ALLELE_NUM</code></td>
<td>Allele number from input; 0 is reference, 1 is first alternate etc -
VEP</td>
</tr>
<tr class="even">
<td><code>REFSEQ_MATCH</code></td>
<td>The RefSeq transcript match status; contains a number of flags
indicating whether this RefSeq transcript matches the underlying
reference sequence and/or an Ensembl transcript (picked by VEPâ€™s
<code>--flag_pick_allele</code> option)</td>
</tr>
<tr class="odd">
<td><code>PICK</code></td>
<td>Indicates if this block of consequence data was picked by VEPâ€™s
<code>--flag_pick_allele</code> option</td>
</tr>
<tr class="even">
<td><code>VEP_ALL_CONSEQUENCE</code></td>
<td>All transcript consequences
(<code>Consequence:SYMBOL:Feature_type:Feature:BIOTYPE</code>) -
VEP</td>
</tr>
<tr class="odd">
<td><code>EXONIC_STATUS</code></td>
<td>Indicates if variant consequence type is â€˜exonicâ€™ or â€˜nonexonicâ€™. We
here define â€˜exonicâ€™ as any variant with either of the following
consequences: <code>stop_gained / stop_lost</code>,
<code>start_lost</code>, <code>frameshift_variant</code>,
<code>missense_variant</code>, <code>splice_donor_variant</code>,
<code>splice_acceptor_variant</code>,
<code>inframe_insertion / inframe_deletion</code>,
<code>synonymous_variant</code>, <code>protein_altering</code>
</td>
</tr>
<tr class="even">
<td><code>CODING_STATUS</code></td>
<td>Indicates if primary variant consequence type is â€˜codingâ€™ or
â€˜noncodingâ€™ (wrt. protein-alteration). â€˜codingâ€™ variants are here
defined as those with an â€˜exonicâ€™ status, with the exception of
synonymous variants</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="gene-information">
<em>Gene information</em><a class="anchor" aria-label="anchor" href="#gene-information"></a>
</h5>
<table class="table">
<colgroup>
<col width="27%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>Tag</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>ENTREZ_ID</code></td>
<td>
<a href="http://www.ncbi.nlm.nih.gov/gene" class="external-link">Entrez</a> gene
identifier</td>
</tr>
<tr class="even">
<td><code>APPRIS</code></td>
<td>Principal isoform flags according to the <a href="http://appris.bioinfo.cnio.es/#/downloads" class="external-link">APPRIS principal
isoform database</a>
</td>
</tr>
<tr class="odd">
<td><code>MANE_SELECT</code></td>
<td>Indicating if the transcript is the MANE Select or MANE Plus
Clinical transcript for the gene (picked by VEPâ€™s
<code>--flag_pick_allele_gene</code> option)</td>
</tr>
<tr class="even">
<td><code>UNIPROT_ID</code></td>
<td>
<a href="http://www.uniprot.org" class="external-link">UniProt</a> identifier</td>
</tr>
<tr class="odd">
<td><code>UNIPROT_ACC</code></td>
<td>
<a href="http://www.uniprot.org" class="external-link">UniProt</a> accession(s)</td>
</tr>
<tr class="even">
<td><code>ENSEMBL_GENE_ID</code></td>
<td>Ensembl gene identifier for VEPâ€™s picked transcript
(<em>ENSGXXXXXXX</em>)</td>
</tr>
<tr class="odd">
<td><code>ENSEMBL_TRANSCRIPT_ID</code></td>
<td>Ensembl transcript identifier for VEPâ€™s picked transcript
(<em>ENSTXXXXXX</em>)</td>
</tr>
<tr class="even">
<td><code>ENSEMBL_PROTEIN_ID</code></td>
<td>Ensembl corresponding protein identifier for VEPâ€™s picked
transcript</td>
</tr>
<tr class="odd">
<td><code>REFSEQ_MRNA</code></td>
<td>Corresponding RefSeq transcript(s) identifier for VEPâ€™s picked
transcript (<em>NM_XXXXX</em>)</td>
</tr>
<tr class="even">
<td><code>TRANSCRIPT_MANE_SELECT</code></td>
<td>MANE select transcript identifer: one high-quality representative
transcript per protein-coding gene that is well-supported by
experimental data and represents the biology of the gene</td>
</tr>
<tr class="odd">
<td><code>TRANSCRIPT_MANE_PLUS_CLINICAL</code></td>
<td>transcripts chosen to supplement MANE Select when needed for
clinical variant reporting</td>
</tr>
<tr class="even">
<td><code>GENCODE_TAG</code></td>
<td>tag for gencode transcript (basic etc)</td>
</tr>
<tr class="odd">
<td><code>GENCODE_TRANSCRIPT_TYPE</code></td>
<td>type of transcript (protein-coding etc.)</td>
</tr>
<tr class="even">
<td><code>CORUM_ID</code></td>
<td>Associated protein complexes (identifiers) from <a href="http://mips.helmholtz-muenchen.de/corum/" class="external-link">CORUM</a>
</td>
</tr>
<tr class="odd">
<td><code>TUMOR_SUPPRESSOR</code></td>
<td>Indicates whether gene is predicted as a tumor suppressor gene, from
Network of Cancer Genes (NCG) &amp; the CancerMine text-mining
resource</td>
</tr>
<tr class="even">
<td><code>TUMOR_SUPPRESSOR_EVIDENCE</code></td>
<td>Underlying evidence for gene being a tumor suppressor. Format:
<code>NCG:&lt;TRUE|FALSE&gt;&amp;CancerMine:&lt;LC|MC|HC&gt;:num_citations</code>
</td>
</tr>
<tr class="odd">
<td><code>ONCOGENE</code></td>
<td>Indicates whether gene is predicted as an oncogene, from Network of
Cancer Genes (NCG) &amp; the CancerMine text-mining resource</td>
</tr>
<tr class="even">
<td><code>ONCOGENE_EVIDENCE</code></td>
<td>Underlying evidence for gene being an oncogene. Format:
<code>NCG:&lt;TRUE|FALSE&gt;&amp;CancerMine:&lt;LC|MC|HC&gt;:num_citations</code>
</td>
</tr>
<tr class="odd">
<td><code>INTOGEN_DRIVER</code></td>
<td>Gene is predicted as a cancer driver in the <a href="https://www.intogen.org/downloads" class="external-link">IntoGen Cancer Drivers
Database</a>
</td>
</tr>
<tr class="even">
<td><code>TCGA_DRIVER</code></td>
<td>Gene is predicted as a cancer driver in the <a href="https://www.ncbi.nlm.nih.gov/pubmed/29625053" class="external-link">TCGA pan-cancer
analysis of cancer driver genes and mutations</a>
</td>
</tr>
<tr class="odd">
<td><code>PROB_EXAC_LOF_INTOLERANT</code></td>
<td>
<code>dbNSFP_gene</code>: the probability of being loss-of-function
intolerant (intolerant of both heterozygous and homozygous lof variants)
based on ExAC r0.3 data</td>
</tr>
<tr class="even">
<td><code>PROB_EXAC_LOF_INTOLERANT_HOM</code></td>
<td>
<code>dbNSFP_gene</code>: the probability of being intolerant of
homozygous, but not heterozygous lof variants based on ExAC r0.3
data</td>
</tr>
<tr class="odd">
<td><code>PROB_EXAC_LOF_TOLERANT_NULL</code></td>
<td>
<code>dbNSFP_gene</code>: the probability of being tolerant of both
heterozygous and homozygous lof variants based on ExAC r0.3 data</td>
</tr>
<tr class="even">
<td><code>PROB_EXAC_NONTCGA_LOF_INTOLERANT</code></td>
<td>
<code>dbNSFP_gene</code>: the probability of being loss-of-function
intolerant (intolerant of both heterozygous and homozygous lof variants)
based on ExAC r0.3 nonTCGA subset</td>
</tr>
<tr class="odd">
<td><code>PROB_EXAC_NONTCGA_LOF_INTOLERANT_HOM</code></td>
<td>
<code>dbNSFP_gene</code>: the probability of being intolerant of
homozygous, but not heterozygous lof variants based on ExAC r0.3 nonTCGA
subset</td>
</tr>
<tr class="even">
<td><code>PROB_EXAC_NONTCGA_LOF_TOLERANT_NULL</code></td>
<td>
<code>dbNSFP_gene</code>: the probability of being tolerant of both
heterozygous and homozygous lof variants based on ExAC r0.3 nonTCGA
subset</td>
</tr>
<tr class="odd">
<td><code>PROB_GNOMAD_LOF_INTOLERANT</code></td>
<td>
<code>dbNSFP_gene</code>: the probability of being loss-of-function
intolerant (intolerant of both heterozygous and homozygous lof variants
based on gnomAD 2.1 data</td>
</tr>
<tr class="even">
<td><code>PROB_GNOMAD_LOF_INTOLERANT_HOM</code></td>
<td>
<code>dbNSFP_gene</code>: the probability of being intolerant of
homozygous, but not heterozygous lof variants based on gnomAD 2.1
data</td>
</tr>
<tr class="odd">
<td><code>PROB_GNOMAD_LOF_TOLERANT_NULL</code></td>
<td>
<code>dbNSFP_gene</code>: the probability of being tolerant of both
heterozygous and homozygous lof variants based on gnomAD 2.1 data</td>
</tr>
<tr class="even">
<td><code>PROB_HAPLOINSUFFICIENCY</code></td>
<td>
<code>dbNSFP_gene</code>: Estimated probability of
haploinsufficiency of the gene (from <a href="http://dx.doi.org/10.1371/journal.pgen.1001154" class="external-link uri">http://dx.doi.org/10.1371/journal.pgen.1001154</a>)</td>
</tr>
<tr class="odd">
<td><code>ESSENTIAL_GENE_CRISPR</code></td>
<td>
<code>dbNSFP_gene</code>: Essential (E) or Non-essential
phenotype-changing (N) based on large scale CRISPR experiments. from <a href="http://dx.doi.org/10.1126/science.aac7041" class="external-link uri">http://dx.doi.org/10.1126/science.aac7041</a>
</td>
</tr>
<tr class="even">
<td><code>ESSENTIAL_GENE_CRISPR2</code></td>
<td>
<code>dbNSFP_gene</code>: Essential (E), context-Specific essential
(S), or Non-essential phenotype-changing (N) based on large scale CRISPR
experiments. from <a href="http://dx.doi.org/10.1016/j.cell.2015.11.015" class="external-link uri">http://dx.doi.org/10.1016/j.cell.2015.11.015</a>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="variant-effect-and-protein-coding-information">
<em>Variant effect and protein-coding information</em><a class="anchor" aria-label="anchor" href="#variant-effect-and-protein-coding-information"></a>
</h5>
<table class="table">
<colgroup>
<col width="27%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>Tag</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>MUTATION_HOTSPOT</code></td>
<td>mutation hotspot codon in <a href="http://cancerhotspots.org/" class="external-link">cancerhotspots.org</a>. Format:
<code>gene_symbol | codon | q-value</code>
</td>
</tr>
<tr class="even">
<td><code>MUTATION_HOTSPOT_TRANSCRIPT</code></td>
<td>hotspot-associated transcripts (Ensembl transcript ID)</td>
</tr>
<tr class="odd">
<td><code>MUTATION_HOTSPOT_CANCERTYPE</code></td>
<td>hotspot-associated cancer types (from cancerhotspots.org)</td>
</tr>
<tr class="even">
<td><code>UNIPROT_FEATURE</code></td>
<td>Overlapping protein annotations from <a href="http://www.uniprot.org" class="external-link">UniProt KB</a>
</td>
</tr>
<tr class="odd">
<td><code>PFAM_DOMAIN</code></td>
<td>Pfam domain identifier (from VEP)</td>
</tr>
<tr class="even">
<td><code>INTOGEN_DRIVER_MUT</code></td>
<td>Indicates if existing variant is predicted as driver mutation from
IntoGen Catalog of Driver Mutations</td>
</tr>
<tr class="odd">
<td><code>PUTATIVE_DRIVER_MUTATION</code></td>
<td>Variant is predicted as driver mutation in the <a href="https://www.ncbi.nlm.nih.gov/pubmed/29625053" class="external-link">TCGA pan-cancer
analysis of cancer driver genes and mutations</a>
</td>
</tr>
<tr class="even">
<td><code>EFFECT_PREDICTIONS</code></td>
<td>All predictions of effect of variant on protein function and
pre-mRNA splicing from <a href="https://sites.google.com/site/jpopgen/dbNSFP" class="external-link">database of
non-synonymous functional predictions - dbNSFP v4.2</a>. Predicted
effects are provided by different sources/algorithms (separated by
<code>&amp;</code>), <code>T</code> = Tolerated, <code>N</code> =
Neutral, <code>D</code> = Damaging: 1.<a href="https://sift.bii.a-star.edu.sg/" class="external-link">SIFT</a>, 2.<a href="http://www.mutationtaster.org/" class="external-link">MutationTaster</a> (data release
Nov 2015), 3.<a href="http://mutationassessor.org/" class="external-link">MutationAssessor</a>
(release 3), 4.<a href="http://fathmm.biocompute.org.uk" class="external-link">FATHMM</a>
(v2.3), 5.<a href="http://provean.jcvi.org/index.php" class="external-link">PROVEAN</a> (v1.1
Jan 2015), 6.<a href="http://fathmm.biocompute.org.uk/fathmmMKL.htm" class="external-link">FATHMM_MKL</a>,
7.<a href="https://www.nature.com/articles/s41588-018-0167-z" class="external-link">PRIMATEAI</a>,
8.<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5570203/" class="external-link">DEOGEN2</a>,
9.<a href="https://www.biorxiv.org/content/10.1101/2021.04.09.438706v1" class="external-link">DBNSFP_CONSENSUS_RNN</a>
(Ensembl/consensus prediction, based on deep learning), 10.<a href="http://nar.oxfordjournals.org/content/42/22/13534" class="external-link">SPLICE_SITE_EFFECT_ADA</a>
(Ensembl/consensus prediction of splice-altering SNVs, based on adaptive
boosting), 11.<a href="http://nar.oxfordjournals.org/content/42/22/13534" class="external-link">SPLICE_SITE_EFFECT_RF</a>
(Ensembl/consensus prediction of splice-altering SNVs, based on random
forest), 12.<a href="http://bejerano.stanford.edu/MCAP" class="external-link">M-CAP</a>, 13.<a href="http://mutpred.mutdb.org" class="external-link">MutPred</a>, 14.<a href="http://mendel.stanford.edu/SidowLab/downloads/gerp/" class="external-link">GERP</a>,
15.<a href="https://doi.org/10.1002/humu.23158" class="external-link">BayesDel</a>, 16.<a href="https://doi.org/10.1093/nar/gkaa288" class="external-link">LIST-S2</a>, 17.<a href="https://www.nature.com/articles/s41467-017-00443-5" class="external-link">ALoFT</a>
</td>
</tr>
<tr class="odd">
<td><code>DBNSFP_BAYESDEL_ADDAF</code></td>
<td>predicted effect from BayesDel (dbNSFP)</td>
</tr>
<tr class="even">
<td><code>DBNSFP_LIST_S2</code></td>
<td>predicted effect from LIST-S2 (dbNSFP)</td>
</tr>
<tr class="odd">
<td><code>DBNSFP_SIFT</code></td>
<td>predicted effect from SIFT (dbNSFP)</td>
</tr>
<tr class="even">
<td><code>DBNSFP_PROVEAN</code></td>
<td>predicted effect from PROVEAN (dbNSFP)</td>
</tr>
<tr class="odd">
<td><code>DBNSFP_MUTATIONTASTER</code></td>
<td>predicted effect from MUTATIONTASTER (dbNSFP)</td>
</tr>
<tr class="even">
<td><code>DBNSFP_MUTATIONASSESSOR</code></td>
<td>predicted effect from MUTATIONASSESSOR (dbNSFP)</td>
</tr>
<tr class="odd">
<td><code>DBNSFP_M_CAP</code></td>
<td>predicted effect from M-CAP (dbNSFP)</td>
</tr>
<tr class="even">
<td><code>DBNSFP_ALOFTPRED</code></td>
<td>predicted effect from ALoFT (dbNSFP)</td>
</tr>
<tr class="odd">
<td><code>DBNSFP_MUTPRED</code></td>
<td>score from MUTPRED (dbNSFP)</td>
</tr>
<tr class="even">
<td><code>DBNSFP_FATHMM</code></td>
<td>predicted effect from FATHMM (dbNSFP)</td>
</tr>
<tr class="odd">
<td><code>DBNSFP_PRIMATEAI</code></td>
<td>predicted effect from PRIMATEAI (dbNSFP)</td>
</tr>
<tr class="even">
<td><code>DBNSFP_DEOGEN2</code></td>
<td>predicted effect from DEOGEN2 (dbNSFP)</td>
</tr>
<tr class="odd">
<td><code>DBNSFP_GERP</code></td>
<td>evolutionary constraint measure from GERP (dbNSFP)</td>
</tr>
<tr class="even">
<td><code>DBNSFP_FATHMM_MKL</code></td>
<td>predicted effect from FATHMM-mkl (dbNSFP)</td>
</tr>
<tr class="odd">
<td><code>DBNSFP_META_RNN</code></td>
<td>predicted effect from ensemble prediction (deep learning -
dbNSFP)</td>
</tr>
<tr class="even">
<td><code>DBNSFP_SPLICE_SITE_RF</code></td>
<td>predicted effect of splice site disruption, using random forest
(dbscSNV)</td>
</tr>
<tr class="odd">
<td><code>DBNSFP_SPLICE_SITE_ADA</code></td>
<td>predicted effect of splice site disruption, using boosting
(dbscSNV)</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="variant-frequenciesannotations-in-germlinesomatic-databases">
<em>Variant frequencies/annotations in germline/somatic
databases</em><a class="anchor" aria-label="anchor" href="#variant-frequenciesannotations-in-germlinesomatic-databases"></a>
</h5>
<table class="table">
<colgroup>
<col width="27%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>Tag</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>AFR_AF_GNOMAD</code></td>
<td>African/American germline allele frequency (<a href="http://gnomad.broadinstitute.org/" class="external-link">gnomAD release 2</a>)</td>
</tr>
<tr class="even">
<td><code>AMR_AF_GNOMAD</code></td>
<td>American germline allele frequency (<a href="http://gnomad.broadinstitute.org/" class="external-link">gnomAD release 2</a>)</td>
</tr>
<tr class="odd">
<td><code>GLOBAL_AF_GNOMAD</code></td>
<td>Adjusted global germline allele frequency (<a href="http://gnomad.broadinstitute.org/" class="external-link">gnomAD release 2</a>)</td>
</tr>
<tr class="even">
<td><code>SAS_AF_GNOMAD</code></td>
<td>South Asian germline allele frequency (<a href="http://gnomad.broadinstitute.org/" class="external-link">gnomAD release 2</a>)</td>
</tr>
<tr class="odd">
<td><code>EAS_AF_GNOMAD</code></td>
<td>East Asian germline allele frequency (<a href="http://gnomad.broadinstitute.org/" class="external-link">gnomAD release 2</a>)</td>
</tr>
<tr class="even">
<td><code>FIN_AF_GNOMAD</code></td>
<td>Finnish germline allele frequency (<a href="http://gnomad.broadinstitute.org/" class="external-link">gnomAD release 2</a>)</td>
</tr>
<tr class="odd">
<td><code>NFE_AF_GNOMAD</code></td>
<td>Non-Finnish European germline allele frequency (<a href="http://gnomad.broadinstitute.org/" class="external-link">gnomAD release 2</a>)</td>
</tr>
<tr class="even">
<td><code>OTH_AF_GNOMAD</code></td>
<td>Other germline allele frequency (<a href="http://gnomad.broadinstitute.org/" class="external-link">gnomAD release 2</a>)</td>
</tr>
<tr class="odd">
<td><code>ASJ_AF_GNOMAD</code></td>
<td>Ashkenazi Jewish allele frequency (<a href="http://gnomad.broadinstitute.org/" class="external-link">gnomAD release 2</a>)</td>
</tr>
<tr class="even">
<td><code>AFR_AF_1KG</code></td>
<td>
<a href="http://www.1000genomes.org" class="external-link">1000G Project - phase 3</a>
germline allele frequency for samples from AFR (African)</td>
</tr>
<tr class="odd">
<td><code>AMR_AF_1KG</code></td>
<td>
<a href="http://www.1000genomes.org" class="external-link">1000G Project - phase 3</a>
germline allele frequency for samples from AMR (Ad Mixed American)</td>
</tr>
<tr class="even">
<td><code>EAS_AF_1KG</code></td>
<td>
<a href="http://www.1000genomes.org" class="external-link">1000G Project - phase 3</a>
germline allele frequency for samples from EAS (East Asian)</td>
</tr>
<tr class="odd">
<td><code>EUR_AF_1KG</code></td>
<td>
<a href="http://www.1000genomes.org" class="external-link">1000G Project - phase 3</a>
germline allele frequency for samples from EUR (European)</td>
</tr>
<tr class="even">
<td><code>SAS_AF_1KG</code></td>
<td>
<a href="http://www.1000genomes.org" class="external-link">1000G Project - phase 3</a>
germline allele frequency for samples from SAS (South Asian)</td>
</tr>
<tr class="odd">
<td><code>GLOBAL_AF_1KG</code></td>
<td>
<a href="http://www.1000genomes.org" class="external-link">1000G Project - phase 3</a>
germline allele frequency for all 1000G project samples (global)</td>
</tr>
<tr class="even">
<td><code>DBSNPRSID</code></td>
<td>
<a href="http://www.ncbi.nlm.nih.gov/SNP/" class="external-link">dbSNP</a> reference ID,
as provided by VEP</td>
</tr>
<tr class="odd">
<td><code>COSMIC_MUTATION_ID</code></td>
<td>Mutation identifier in <a href="http://cancer.sanger.ac.uk/cancergenome/projects/cosmic/" class="external-link">Catalog
of somatic mutations in cancer</a> database, as provided by VEP</td>
</tr>
<tr class="even">
<td><code>TCGA_PANCANCER_COUNT</code></td>
<td>Raw variant count across all TCGA tumor types</td>
</tr>
<tr class="odd">
<td><code>TCGA_FREQUENCY</code></td>
<td>Frequency of variant across TCGA tumor types. Format:
<code>tumortype| percent affected|affected cases|total cases</code>
</td>
</tr>
<tr class="even">
<td><code>ICGC_PCAWG_OCCURRENCE</code></td>
<td>Mutation occurrence in [ICGC</td>
</tr>
<tr class="odd">
<td><code>ICGC_PCAWG_AFFECTED_DONORS</code></td>
<td>Number of donors with the current mutation in [ICGC</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="clinical-associations">
<em>Clinical associations</em><a class="anchor" aria-label="anchor" href="#clinical-associations"></a>
</h5>
<table class="table">
<colgroup>
<col width="27%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>Tag</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>CLINVAR_MSID</code></td>
<td>
<a href="http://www.ncbi.nlm.nih.gov/clinvar" class="external-link">ClinVar</a> Measure
Set/Variant ID</td>
</tr>
<tr class="even">
<td><code>CLINVAR_ALLELE_ID</code></td>
<td>
<a href="http://www.ncbi.nlm.nih.gov/clinvar" class="external-link">ClinVar</a> allele
ID</td>
</tr>
<tr class="odd">
<td><code>CLINVAR_PMID</code></td>
<td>Associated Pubmed IDs for variant in <a href="http://www.ncbi.nlm.nih.gov/clinvar" class="external-link">ClinVar</a> - germline
state-of-origin</td>
</tr>
<tr class="even">
<td><code>CLINVAR_HGVSP</code></td>
<td>Protein variant expression using HGVS nomenclature</td>
</tr>
<tr class="odd">
<td><code>CLINVAR_PMID_SOMATIC</code></td>
<td>Associated Pubmed IDs for variant in <a href="http://www.ncbi.nlm.nih.gov/clinvar" class="external-link">ClinVar</a> - somatic
state-of-origin</td>
</tr>
<tr class="even">
<td><code>CLINVAR_CLNSIG</code></td>
<td>Clinical significance for variant in <a href="http://www.ncbi.nlm.nih.gov/clinvar" class="external-link">ClinVar</a> - germline
state-of-origin</td>
</tr>
<tr class="odd">
<td><code>CLINVAR_CLNSIG_SOMATIC</code></td>
<td>Clinical significance for variant in <a href="http://www.ncbi.nlm.nih.gov/clinvar" class="external-link">ClinVar</a> - somatic
state-of-origin</td>
</tr>
<tr class="even">
<td><code>CLINVAR_MEDGEN_CUI</code></td>
<td>Associated <a href="https://www.ncbi.nlm.nih.gov/medgen/" class="external-link">MedGen</a>
concept identifiers (<em>CUIs</em>) - germline state-of-origin</td>
</tr>
<tr class="odd">
<td><code>CLINVAR_MEDGEN_CUI_SOMATIC</code></td>
<td>Associated <a href="https://www.ncbi.nlm.nih.gov/medgen/" class="external-link">MedGen</a>
concept identifiers (<em>CUIs</em>) - somatic state-of-origin</td>
</tr>
<tr class="even">
<td><code>CLINVAR_VARIANT_ORIGIN</code></td>
<td>Origin of variant (somatic, germline, de novo etc.) for variant in
<a href="http://www.ncbi.nlm.nih.gov/clinvar" class="external-link">ClinVar</a>
</td>
</tr>
<tr class="odd">
<td><code>CLINVAR_REVIEW_STATUS_STARS</code></td>
<td>Rating of the <a href="http://www.ncbi.nlm.nih.gov/clinvar" class="external-link">ClinVar</a> variant (0-4
stars) with respect to level of review</td>
</tr>
<tr class="even">
<td><code>DOCM_PMID</code></td>
<td>Associated Pubmed IDs for variant in <a href="http://docm.genome.wustl.edu" class="external-link">Database of Curated
Mutations</a>
</td>
</tr>
<tr class="odd">
<td><code>OPENTARGETS_DISEASE_ASSOCS</code></td>
<td>Associations between protein targets and disease based on multiple
lines of evidence (mutations, affected pathways, GWAS, literature etc).
Format: <code>CUI:EFO_ID:IS_DIRECT:OVERALL_SCORE</code>
</td>
</tr>
<tr class="even">
<td><code>OPENTARGETS_TRACTABILITY_COMPOUND</code></td>
<td>Confidence for the existence of a modulator (small molecule) that
interacts with the target to elicit a desired biological effect</td>
</tr>
<tr class="odd">
<td><code>OPENTARGETS_TRACTABILITY_ANTIBODY</code></td>
<td>Confidence for the existence of a modulator (antibody) that
interacts with the target to elicit a desired biological effect</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="other">
<em>Other</em><a class="anchor" aria-label="anchor" href="#other"></a>
</h5>
<table class="table">
<colgroup>
<col width="27%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>Tag</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>CHEMBL_COMPOUND_ID</code></td>
<td>antineoplastic drugs targeting the encoded protein (from <a href="https://www.targetvalidation.org/" class="external-link">Open Targets Platform</a>,
drugs are listed as <a href="https://www.ebi.ac.uk/chembl/" class="external-link">ChEMBL</a>
compound identifiers)</td>
</tr>
<tr class="even">
<td>
<code>CIVIC_ID</code>, <code>CIVIC_ID_SEGMENT</code>
</td>
<td>Variant/segment (exon, codon) identifiers in the <a href="http://civic.genome.wustl.edu" class="external-link">CIViC database</a>
</td>
</tr>
<tr class="odd">
<td>
<code>CGI_ID</code>, <code>CGI_ID_SEGMENT</code>
</td>
<td>Variant/segment (exon, codon) identifier in the <a href="https://www.cancergenomeinterpreter.org/biomarkers" class="external-link">Cancer Genome
Interpreter Cancer Biomarkers Database</a>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level4">
<h4 id="tab-separated-values-tsv">2. Tab-separated values (TSV)<a class="anchor" aria-label="anchor" href="#tab-separated-values-tsv"></a>
</h4>
<p>We provide a tab-separated values file with most important
annotations for SNVs/InDels. The file has the following naming
convention:</p>
<ul>
<li><code>&lt;sample_id&gt;.pcgr_acmg.&lt;genome_assembly&gt;.snvs_indels.tiers.tsv</code></li>
</ul>
<p>The SNVs/InDels are organized into different <strong>tiers</strong>
(as defined above for the HTML report)</p>
<p>The following variables are included in the tiered TSV file (VCF tags
issued by the user (<code>--preserved_info_tags</code>) will be appended
at the end):</p>
<table class="table">
<colgroup>
<col width="43%">
<col width="56%">
</colgroup>
<thead><tr class="header">
<th>Variable</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1. <code>CHROM</code>
</td>
<td>Chromosome</td>
</tr>
<tr class="even">
<td>2. <code>POS</code>
</td>
<td>Position (VCF-based)</td>
</tr>
<tr class="odd">
<td>3. <code>REF</code>
</td>
<td>Reference allele</td>
</tr>
<tr class="even">
<td>4. <code>ALT</code>
</td>
<td>Alternate allele</td>
</tr>
<tr class="odd">
<td>5. <code>GENOMIC_CHANGE</code>
</td>
<td>Identifier for variant at the genome (VCF) level,
e.g.Â <code>1:g.152382569A&gt;G</code>. Format:
<code>&lt;chrom&gt;:g.&lt;position&gt;&lt;ref_allele&gt;&lt;alt_allele&gt;</code>
</td>
</tr>
<tr class="even">
<td>6. <code>GENOME_VERSION</code>
</td>
<td>Assembly version, e.g.Â GRCh37</td>
</tr>
<tr class="odd">
<td>7. <code>VCF_SAMPLE_ID</code>
</td>
<td>Sample identifier</td>
</tr>
<tr class="even">
<td>8. <code>VARIANT_CLASS</code>
</td>
<td>Variant type, e.g.Â SNV/insertion/deletion</td>
</tr>
<tr class="odd">
<td>9. <code>SYMBOL</code>
</td>
<td>Gene symbol</td>
</tr>
<tr class="even">
<td>10. <code>GENE_NAME</code>
</td>
<td>Gene description</td>
</tr>
<tr class="odd">
<td>11. <code>CCDS</code>
</td>
<td>CCDS identifier</td>
</tr>
<tr class="even">
<td>12. <code>CANONICAL</code>
</td>
<td>indication of canonical transcript</td>
</tr>
<tr class="odd">
<td>13. <code>ENTREZ_ID</code>
</td>
<td>Entrez gene identifier</td>
</tr>
<tr class="even">
<td>14. <code>UNIPROT_ID</code>
</td>
<td>UniProt protein identifier</td>
</tr>
<tr class="odd">
<td>15. <code>ENSEMBL_TRANSCRIPT_ID</code>
</td>
<td>Ensembl transcript identifier</td>
</tr>
<tr class="even">
<td>16. <code>ENSEMBL_GENE_ID</code>
</td>
<td>Ensembl gene identifier</td>
</tr>
<tr class="odd">
<td>17. <code>REFSEQ_MRNA</code>
</td>
<td>RefSeq mRNA identifier</td>
</tr>
<tr class="even">
<td>18. <code>REFSEQ_PEPTIDE</code>
</td>
<td>RefSeq peptide identifier</td>
</tr>
<tr class="odd">
<td>19. <code>TRANSCRIPT_MANE_SELECT</code>
</td>
<td>MANE select transcript identifier</td>
</tr>
<tr class="even">
<td>20. <code>ONCOGENE</code>
</td>
<td>Gene is predicted/classified as an oncogene (CancerMine/NCG)</td>
</tr>
<tr class="odd">
<td>21. <code>ONCOGENE_EVIDENCE</code>
</td>
<td>Underlying evidence for oncogene prediction/classification</td>
</tr>
<tr class="even">
<td>22. <code>TUMOR_SUPPRESSOR</code>
</td>
<td>Gene is predicted/classified as tumor suppressor
(CancerMine/NCG)</td>
</tr>
<tr class="odd">
<td>23. <code>TUMOR_SUPPRESSOR_EVIDENCE</code>
</td>
<td>Underlying evidence for tumor suppressor
prediction/classification</td>
</tr>
<tr class="even">
<td>24. <code>CONSEQUENCE</code>
</td>
<td>Variant consequence (as defined above for VCF output:
Consequence)</td>
</tr>
<tr class="odd">
<td>25. <code>PROTEIN_CHANGE</code>
</td>
<td>Protein change (HGVSp without reference accession)</td>
</tr>
<tr class="even">
<td>26. <code>PROTEIN_DOMAIN</code>
</td>
<td>Protein domain description (Pfam)</td>
</tr>
<tr class="odd">
<td>27. <code>CODING_STATUS</code>
</td>
<td>Coding variant status wrt. protein alteration (<code>coding</code>
or <code>noncoding</code>)</td>
</tr>
<tr class="even">
<td>28. <code>EXONIC_STATUS</code>
</td>
<td>Exonic variant status (<code>exonic</code> or
<code>nonexonic</code>)</td>
</tr>
<tr class="odd">
<td>29. <code>CDS_CHANGE</code>
</td>
<td>composite VEP-based variable for coding change, format:
<code>Consequence:Feature:cDNA_position:EXON:HGVSp_short</code>
</td>
</tr>
<tr class="even">
<td>30. <code>HGVSp</code>
</td>
<td></td>
</tr>
<tr class="odd">
<td>31. <code>HGVSc</code>
</td>
<td></td>
</tr>
<tr class="even">
<td>32. <code>REGULATORY_ANNOTATION</code>
</td>
<td>Regulatory annotations (promoter/enhancer/TF_binding site overlap
etc.) from VEP (optional)</td>
</tr>
<tr class="odd">
<td>33. <code>EFFECT_PREDICTIONS</code>
</td>
<td>as defined above for VCF</td>
</tr>
<tr class="even">
<td>34. <code>MUTATION_HOTSPOT</code>
</td>
<td>mutation hotspot codon in cancerhotspots.org. Format:
<code>gene_symbol | codon | q-value</code>
</td>
</tr>
<tr class="odd">
<td>35. <code>MUTATION_HOTSPOT_TRANSCRIPT</code>
</td>
<td>hotspot-associated transcripts (Ensembl transcript ID)</td>
</tr>
<tr class="even">
<td>36. <code>MUTATION_HOTSPOT_CANCERTYPE</code>
</td>
<td>hotspot-associated cancer types (from cancerhotspots.org)</td>
</tr>
<tr class="odd">
<td>37. <code>PUTATIVE_DRIVER_MUTATION</code>
</td>
<td>Indicates if variant is predicted as driver mutation from TCGAâ€™s
PanCancer study of cancer driver mutation</td>
</tr>
<tr class="even">
<td>38. <code>CHASMPLUS_DRIVER</code>
</td>
<td>Driver mutation predicted by CHASMplus algorithm</td>
</tr>
<tr class="odd">
<td>39. <code>CHASMPLUS_TTYPE</code>
</td>
<td>Tumor type for which mutation is predicted as driver by
CHASMplus</td>
</tr>
<tr class="even">
<td>40. <code>VEP_ALL_CSQ</code>
</td>
<td>all VEP transcript block consequences</td>
</tr>
<tr class="odd">
<td>41. <code>DBSNPRSID</code>
</td>
<td>dbSNP reference cluster ID</td>
</tr>
<tr class="even">
<td>42. <code>COSMIC_MUTATION_ID</code>
</td>
<td>COSMIC mutation ID</td>
</tr>
<tr class="odd">
<td>43. <code>TCGA_PANCANCER_COUNT</code>
</td>
<td>Raw variant count across all TCGA tumor types</td>
</tr>
<tr class="even">
<td>44. <code>TCGA_FREQUENCY</code>
</td>
<td>Frequency of variant across TCGA tumor types. Format:
<code>tumortype | percent affected | affected cases | total cases</code>
</td>
</tr>
<tr class="odd">
<td>45. <code>ICGC_PCAWG_OCCURRENCE</code>
</td>
<td>Mutation occurrence in ICGC-PCAWG by project:
<code>project_code | affected_donors | tested_donors | frequency</code>
</td>
</tr>
<tr class="even">
<td>46. <code>CHEMBL_COMPOUND_ID</code>
</td>
<td>Compounds (as ChEMBL IDs) that are targeted towards the encoded
protein (from Open Targets Platform)</td>
</tr>
<tr class="odd">
<td>47. <code>CHEMBL_COMPOUND_TERMS</code>
</td>
<td>Compounds (as drug names) that are targeted towards the encoded
protein (from Open Targets Platform)</td>
</tr>
<tr class="even">
<td>48. <code>SIMPLEREPEATS_HIT</code>
</td>
<td>Variant overlaps UCSC <em>simpleRepeat</em> sequence repeat
track</td>
</tr>
<tr class="odd">
<td>49. <code>WINMASKER_HIT</code>
</td>
<td>Variant overlaps UCSC <em>windowmaskerSdust</em> sequence repeat
track</td>
</tr>
<tr class="even">
<td>50. <code>OPENTARGETS_RANK</code>
</td>
<td>OpenTargets association score (between 0 and 1) for gene (maximum
across cancer phenotypes)</td>
</tr>
<tr class="odd">
<td>51. <code>CLINVAR</code>
</td>
<td>ClinVar association: variant origin and associated traits</td>
</tr>
<tr class="even">
<td>52. <code>CLINVAR_CLNSIG</code>
</td>
<td>clinical significance of ClinVar variant</td>
</tr>
<tr class="odd">
<td>53. <code>GLOBAL_AF_GNOMAD</code>
</td>
<td>global germline allele frequency in gnomAD</td>
</tr>
<tr class="even">
<td>54. <code>GLOBAL_AF_1KG</code>
</td>
<td>1000G Project - phase 3, germline allele frequency</td>
</tr>
<tr class="odd">
<td>55. <code>NCER_PERCENTILE</code>
</td>
<td>Genome-wide percentile score (10bp bins) from the non-coding
Essential Regulation (ncER) algorithm</td>
</tr>
<tr class="even">
<td>56. <code>CALL_CONFIDENCE</code>
</td>
<td>confidence indicator for somatic variant</td>
</tr>
<tr class="odd">
<td>57. <code>DP_TUMOR</code>
</td>
<td>sequencing depth at variant site (tumor sample)</td>
</tr>
<tr class="even">
<td>58. <code>AF_TUMOR</code>
</td>
<td>allelic fraction of alternate allele (tumor sample)</td>
</tr>
<tr class="odd">
<td>59. <code>DP_CONTROL</code>
</td>
<td>sequencing depth at variant site (control sample)</td>
</tr>
<tr class="even">
<td>60. <code>AF_CONTROL</code>
</td>
<td>allelic fraction of alternate allele (control sample)</td>
</tr>
<tr class="odd">
<td>61. <code>TIER</code>
</td>
<td></td>
</tr>
<tr class="even">
<td>62. <code>TIER_DESCRIPTION</code>
</td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="mutational-signature-contributions">3. Mutational signature contributions<a class="anchor" aria-label="anchor" href="#mutational-signature-contributions"></a>
</h4>
<p>We provide a tab-separated values file with information about
mutational signatures detected in the tumor sample. The file has the
following naming convention:</p>
<ul>
<li><code>&lt;sample_id&gt;.pcgr_acmg.&lt;genome_assembly&gt;.mutational_signatures.tsv</code></li>
</ul>
<p>The format of the TSV file is the following:</p>
<table class="table">
<colgroup>
<col width="43%">
<col width="56%">
</colgroup>
<thead><tr class="header">
<th>Variable</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1. <code>signature_id</code>
</td>
<td>identifier for signature</td>
</tr>
<tr class="even">
<td>2. <code>sample_id</code>
</td>
<td>sample identifier</td>
</tr>
<tr class="odd">
<td>3. <code>prop_signature</code>
</td>
<td>relative contribution of mutational signature</td>
</tr>
<tr class="even">
<td>4. <code>group</code>
</td>
<td>keyword for signature aetiology</td>
</tr>
<tr class="odd">
<td>5. <code>all_reference_signatures</code>
</td>
<td>logical indicating if all reference signatures were used for
reconstruction/inference</td>
</tr>
<tr class="even">
<td>6. <code>tumor_type</code>
</td>
<td>tumor type (used for retrieval of reference signatures)</td>
</tr>
<tr class="odd">
<td>7. <code>reference_collection</code>
</td>
<td>collection used for reference signatures</td>
</tr>
<tr class="even">
<td>8. <code>reference_signatures</code>
</td>
<td>signatures present in reference collection</td>
</tr>
<tr class="odd">
<td>9. <code>fitting_accuracy</code>
</td>
<td>accuracy of mutational signature fitting</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="copy-number-aberrations">Copy number aberrations<a class="anchor" aria-label="anchor" href="#copy-number-aberrations"></a>
</h3>
<div class="section level4">
<h4 id="tab-separated-values-tsv-1">1. Tab-separated values (TSV)<a class="anchor" aria-label="anchor" href="#tab-separated-values-tsv-1"></a>
</h4>
<p>Copy number segments are intersected with the genomic coordinates of
all transcripts from <a href="https://www.gencodegenes.org/releases/current.html" class="external-link">GENCODEâ€™s
basic gene annotation</a>. In addition, PCGR attaches cancer-relevant
annotations for the affected transcripts. The naming convention of the
compressed TSV file is as follows:</p>
<ul>
<li>
<code>&lt;sample_id&gt;.pcgr_acmg.&lt;genome_assembly&gt;.cna_segments.tsv.gz</code>
<ul>
<li>
<strong>NOTE</strong>: This file is organized according to the
<em>affected transcripts</em> (i.e.Â one line/record per affected
transcript).</li>
</ul>
</li>
</ul>
<p>The format of the compressed TSV file is the following:</p>
<table class="table">
<colgroup>
<col width="43%">
<col width="56%">
</colgroup>
<thead><tr class="header">
<th>Variable</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1. <code>chrom</code>
</td>
<td>chromosome</td>
</tr>
<tr class="even">
<td>2. <code>segment_start</code>
</td>
<td>start of copy number segment</td>
</tr>
<tr class="odd">
<td>3. <code>segment_end</code>
</td>
<td>end of copy number segment</td>
</tr>
<tr class="even">
<td>4. <code>segment_id</code>
</td>
<td>segment identifier</td>
</tr>
<tr class="odd">
<td>5. <code>segment_length_Mb</code>
</td>
<td>length of segment in Mb</td>
</tr>
<tr class="even">
<td>6. <code>event_type</code>
</td>
<td>focal or broad (covering more than 25% of chromosome arm)</td>
</tr>
<tr class="odd">
<td>7. <code>cytoband(s)</code>
</td>
<td>cytobands covered by segment</td>
</tr>
<tr class="even">
<td>8. <code>LogR</code>
</td>
<td>Copy log-ratio</td>
</tr>
<tr class="odd">
<td>9. <code>sample_id</code>
</td>
<td>Sample identifier</td>
</tr>
<tr class="even">
<td>10. <code>ensembl_gene_id</code>
</td>
<td>Ensembl gene identifier</td>
</tr>
<tr class="odd">
<td>11. <code>symbol</code>
</td>
<td>Gene symbol</td>
</tr>
<tr class="even">
<td>12. <code>ensembl_transcript_id</code>
</td>
<td>Ensembl transcript identifier</td>
</tr>
<tr class="odd">
<td>13. <code>transcript_start</code>
</td>
<td>chromosomal position of transcript start</td>
</tr>
<tr class="even">
<td>14. <code>transcript_end</code>
</td>
<td>chromosomal position of transcript end</td>
</tr>
<tr class="odd">
<td>15. <code>transcript_overlap_percent</code>
</td>
<td>percent of transcript length covered by CN segment</td>
</tr>
<tr class="even">
<td>16. <code>name</code>
</td>
<td>gene name description</td>
</tr>
<tr class="odd">
<td>17. <code>biotype</code>
</td>
<td>type of gene</td>
</tr>
<tr class="even">
<td>18. <code>tumor_suppressor</code>
</td>
<td>tumor suppressor gene status (CancerMine literature database/Network
of Cancer Genes)</td>
</tr>
<tr class="odd">
<td>19. <code>oncogene</code>
</td>
<td>oncogene status (CancerMine literature database/Network of Cancer
Genes)</td>
</tr>
<tr class="even">
<td>20. <code>intogen_drivers</code>
</td>
<td>predicted driver gene status (IntoGen Cancer Drivers Database)</td>
</tr>
<tr class="odd">
<td>21. <code>chembl_compound_id</code>
</td>
<td>anti-cancer drugs that are targeted towards the encoded protein
(from Open Targets Platform, drugs are listed as ChEMBL compound
identifiers)</td>
</tr>
<tr class="even">
<td>22. <code>gencode_tag</code>
</td>
<td></td>
</tr>
<tr class="odd">
<td>23. <code>gencode_release</code>
</td>
<td></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
