<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pcgrr">
<title>Running • PCGR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running">
<meta property="og:description" content="pcgrr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.10.15</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Intro</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/running.html">Running</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/input.html">Input files</a>
    <a class="dropdown-item" href="../articles/output.html">Output files</a>
    <a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a>
    <a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a>
    <a class="dropdown-item" href="../articles/developers.html">Developer notes</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/faq.html">FAQ</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Running</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/HEAD/../vignettes/running.Rmd" class="external-link"><code>../vignettes/running.Rmd</code></a></small>
      <div class="d-none name"><code>running.Rmd</code></div>
    </div>

    
    
<p>The PCGR software comes with a range of options to configure the
analyses performed and to customize the output report.</p>
<div class="section level2">
<h2 id="key-settings">Key settings<a class="anchor" aria-label="anchor" href="#key-settings"></a>
</h2>
<p>Below, we will outline key settings that are important with respect
to the sequencing assay that has been conducted.</p>
<div class="section level3">
<h3 id="whole-genome-whole-exome-or-targeted-sequencing-assay">Whole-genome, whole-exome or targeted sequencing assay<a class="anchor" aria-label="anchor" href="#whole-genome-whole-exome-or-targeted-sequencing-assay"></a>
</h3>
<p>By default, PCGR expects that the input VCF comes from whole-exome
sequencing. This can be adjusted with the <code>assay</code> option,
which takes three alternative values:</p>
<ul>
<li><code>--assay &lt;WGS|WES|TARGETED&gt;</code></li>
</ul>
<p>If the input VCF comes from a targeted sequencing assay/panel
(i.e. <code>--assay TARGETED</code>), the target size of the assay/panel
should be adjusted accordingly, which is critical for tumor mutational
burden estimation. Assay target size, reflecting the <em>protein-coding
portion of the assayed genomic region</em> in megabases, can be
configured through the following option:</p>
<ul>
<li><code>--target_size_mb &lt;value&gt;</code></li>
</ul>
</div>
<div class="section level3">
<h3 id="tumor-control-vs--tumor-only">Tumor-control vs. tumor-only<a class="anchor" aria-label="anchor" href="#tumor-control-vs--tumor-only"></a>
</h3>
<p>By default, PCGR expects that the input VCF contains somatic variants
identified from a tumor-control sequencing setup. This implies that the
VCF contains information with respect to variant allelic depth/support
both for the tumor sample and the corresponding control sample.</p>
<p>If the input VCF comes from a <strong>tumor-only</strong> assay, turn
on the <code>--tumor_only</code> option. In this mode, PCGR conducts a
set of successive filtering steps on the raw input set of variants,
aiming to exclude the majority of germline variants from the tumor-only
input set. In addition to default filtering applied against variants
found in the gnomAD/1000 Genomes Project databases (population-specific
allele frequency thresholds can be configured, see below), additional
filtering procedures can be explicitly set, i.e.:</p>
<ul>
<li>
<code>--exclude_dbsnp_nonsomatic</code> (recommended)</li>
<li><code>--exclude_likely_het_germline</code></li>
<li><code>--exclude_likely_hom_germline</code></li>
<li><code>--exclude_nonexonic</code></li>
</ul>
<p>If the user has provided a <a href="input.html#panel-of-normals-pon-vcf">panel-of-normals VCF file</a>
as input(<code>--pon_vcf</code>), one may also opt to remove variants
found in that variant collection using a dedicated option:</p>
<ul>
<li><code>--exclude_pon</code></li>
</ul>
<hr>
<p><strong>IMPORTANT NOTE</strong>: Note that a number of the analyses
available in PCGR is not particularly well-suited for tumor-only input,
e.g. mutational signature analysis and MSI status prediction.</p>
</div>
<div class="section level3">
<h3 id="sample-properties">Sample properties<a class="anchor" aria-label="anchor" href="#sample-properties"></a>
</h3>
<p>A few properties of the tumor sample can be fed to the report,
currently only used for display/reporting:</p>
<ul>
<li>
<code>--tumor_ploidy &lt;value&gt;</code> - ploidy estimate</li>
<li>
<code>--tumor_purity &lt;value&gt;</code> - purity estimate</li>
<li>
<code>--cell_line</code> - cell line sample</li>
</ul>
</div>
<div class="section level3">
<h3 id="allelic-support">Allelic support<a class="anchor" aria-label="anchor" href="#allelic-support"></a>
</h3>
<p>Proper designation of INFO tags in the input VCF that denote variant
depth/allelic fraction is critical to make the report as comprehensive
as possible. See the <a href="input.html">input section</a> on how to
accomplish this.</p>
<p>Several options are used to inform PCGR about tags in the VCF INFO
field that encodes this information:</p>
<ul>
<li><code>--tumor_dp_tag &lt;value&gt;</code></li>
<li><code>--tumor_af_tag &lt;value&gt;</code></li>
<li><code>--control_dp_tag &lt;value&gt;</code></li>
<li><code>--control_af_tag &lt;value&gt;</code></li>
<li><code>--call_conf_tag &lt;value&gt;</code></li>
</ul>
<p>If these tags are set correctly, one may set thresholds (sequencing
depth and/or allelic fraction) on the variants to be included in the
report, i.e. through:</p>
<ul>
<li><code>--tumor_dp_min &lt;value&gt;</code></li>
<li><code>--tumor_af_min &lt;value&gt;</code></li>
<li><code>--control_dp_min &lt;value&gt;</code></li>
<li><code>--control_af_max &lt;value&gt;</code></li>
</ul>
</div>
<div class="section level3">
<h3 id="tumor-site">Tumor site<a class="anchor" aria-label="anchor" href="#tumor-site"></a>
</h3>
<p>The primary tissue/site of the tumor sample that is analyzed can and
should be be provided, through the following option:</p>
<ul>
<li><code>--tumor_site &lt;value&gt;</code></li>
</ul>
<p>This option takes a value between 0 and 30, reflecting the main
primary sites/tissues for human cancers. Note that the default value is
<strong>0</strong> (implies <strong>Any</strong> tissue/site), which
essentially prohibits the presence of <a href="variant_classification.html">tier 1 variants</a> occurring in the
report.</p>
</div>
</div>
<div class="section level2">
<h2 id="all-options">All options<a class="anchor" aria-label="anchor" href="#all-options"></a>
</h2>
<p>A tumor sample report is generated by running the
<strong>pcgr</strong> command, which takes the following arguments and
options:</p>
<pre class="text"><code>usage:
    pcgr -h [options]
    --input_vcf &lt;INPUT_VCF&gt;
    --pcgr_dir &lt;PCGR_DIR&gt;
    --output_dir &lt;OUTPUT_DIR&gt;
    --genome_assembly &lt;GENOME_ASSEMBLY&gt;
    --sample_id &lt;SAMPLE_ID&gt;

Personal Cancer Genome Reporter (PCGR) workflow for clinical interpretation of somatic nucleotide variants and copy number aberration segments

Required arguments:
  --input_vcf INPUT_VCF - VCF input file with somatic variants in tumor sample, SNVs/InDels
  --pcgr_dir PCGR_DIR - PCGR base directory with accompanying data directory, e.g. ~/pcgr-0.9.4
  --output_dir OUTPUT_DIR - Output directory
  --genome_assembly {grch37,grch38} - Human genome assembly build: grch37 or grch38
  --sample_id SAMPLE_ID - Tumor sample/cancer genome identifier - prefix for output files

vcfanno options:
  --vcfanno_n_proc VCFANNO_N_PROC - Number of vcfanno processes (option '-p' in vcfanno), default: 4

VEP options:
  --vep_n_forks VEP_N_FORKS - Number of forks (option '--fork' in VEP), default: 4
  --vep_buffer_size VEP_BUFFER_SIZE - Variant buffer size (variants read into memory simultaneously, option '--buffer_size' in VEP)
                                      set lower to reduce memory usage, default: 100
  --vep_pick_order VEP_PICK_ORDER - Comma-separated string of ordered transcript/variant properties for selection of primary variant consequence
                                    (option '--pick_order' in VEP), default: canonical,appris,biotype,ccds,rank,tsl,length,mane
  --vep_no_intergenic   - Skip intergenic variants during processing (option '--no_intergenic' in VEP), default: False
  --vep_regulatory      - Add VEP regulatory annotations (option '--regulatory' )or non-coding interpretation, default: False
  --vep_gencode_all     - Consider all GENCODE transcripts with Variant Effect Predictor (VEP) (option '--gencode_basic' in VEP is used by default in PCGR) default: False

Tumor mutational burden (TMB) and MSI options:
  --target_size_mb TARGET_SIZE_MB - For mutational burden analysis - approximate protein-coding target size in Mb of sequencing assay (default: 34 (WES/WGS))
  --estimate_tmb        - Estimate tumor mutational burden from the total number of somatic mutations and target region size, default: False
  --estimate_msi_status - Predict microsatellite instability status from patterns of somatic mutations/indels, default: False
  --tmb_algorithm {all_coding,nonsyn} - Method for calculation of TMB, all coding variants (Chalmers et al., Genome Medicine, 2017), or non-synonymous variants only, default: all_coding

Mutational signature options:
  --estimate_signatures - Estimate relative contributions of reference mutational signatures in query sample and detect potential kataegis events, default: False
  --min_mutations_signatures MIN_MUTATIONS_SIGNATURES - Minimum number of SNVs required for reconstruction of mutational signatures (SBS) by MutationalPatterns (default: 200, minimum n = 100)
  --all_reference_signatures - Use all reference mutational signatures (SBS, n = 67) in signature reconstruction rather than only those already attributed to the tumor type (default: False)
  --include_artefact_signatures - Include sequencing artefacts in the collection of reference signatures (default: False

Tumor-only options:
  --tumor_only          - Input VCF comes from tumor-only sequencing, calls will be filtered for variants of germline origin, (default: False)
  --cell_line           - Input VCF comes from tumor cell line sequencing (requires --tumor_only), calls will be filtered for variants of germline origin, (default: False)
  --pon_vcf PON_VCF     - VCF file with germline calls from Panel of Normals (PON) - i.e. blacklisted variants, (default: None)
  --maf_onekg_eur MAF_ONEKG_EUR - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - European pop, default: 0.002)
  --maf_onekg_amr MAF_ONEKG_AMR - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - Ad Mixed American pop, default: 0.002)
  --maf_onekg_afr MAF_ONEKG_AFR - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - African pop, default: 0.002)
  --maf_onekg_eas MAF_ONEKG_EAS - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - East Asian pop, default: 0.002)
  --maf_onekg_sas MAF_ONEKG_SAS - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - South Asian pop, default: 0.002)
  --maf_onekg_global MAF_ONEKG_GLOBAL - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - global pop, default: 0.002)
  --maf_gnomad_nfe MAF_GNOMAD_NFE - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - European (non-Finnish), default: 0.002)
  --maf_gnomad_asj MAF_GNOMAD_ASJ - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - Ashkenazi Jewish, default: 0.002)
  --maf_gnomad_fin MAF_GNOMAD_FIN - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - European (Finnish), default: 0.002)
  --maf_gnomad_oth MAF_GNOMAD_OTH - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - Other, default: 0.002)
  --maf_gnomad_amr MAF_GNOMAD_AMR - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - Latino/Admixed American, default: 0.002)
  --maf_gnomad_afr MAF_GNOMAD_AFR - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - African/African-American, default: 0.002)
  --maf_gnomad_eas MAF_GNOMAD_EAS - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - East Asian, default: 0.002)
  --maf_gnomad_sas MAF_GNOMAD_SAS - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - South Asian, default: 0.002)
  --maf_gnomad_global MAF_GNOMAD_GLOBAL - Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - global population, default: 0.002)
  --exclude_pon        - Exclude variants occurring in PoN (Panel of Normals, if provided as VCF (--pon_vcf), default: False)
  --exclude_likely_hom_germline - Exclude likely homozygous germline variants (100 pct allelic fraction for alternate allele in tumor, very unlikely somatic event, default: False)
  --exclude_likely_het_germline - Exclude likely heterozygous germline variants (40-60 pct allelic fraction, AND presence in dbSNP + gnomAD, AND not existing as somatic event in COSMIC/TCGA, default: False)
  --exclude_dbsnp_nonsomatic - Exclude variants found in dbSNP (only those that are NOT found in ClinVar(somatic origin)/DoCM/TCGA/COSMIC, defult: False)
  --exclude_nonexonic - Exclude non-exonic variants, default: False)

Allelic support options:
  --tumor_dp_tag TUMOR_DP_TAG - Specify VCF INFO tag for sequencing depth (tumor, must be Type=Integer, default: _NA_
  --tumor_af_tag TUMOR_AF_TAG - Specify VCF INFO tag for variant allelic fraction (tumor,  must be Type=Float, default: _NA_
  --control_dp_tag CONTROL_DP_TAG - Specify VCF INFO tag for sequencing depth (control, must be Type=Integer, default: _NA_
  --control_af_tag CONTROL_AF_TAG - Specify VCF INFO tag for variant allelic fraction (control, must be Type=Float, default: _NA_
  --call_conf_tag CALL_CONF_TAG - Specify VCF INFO tag for somatic variant call confidence (must be categorical, e.g. Type=String, default: _NA_
  --tumor_dp_min TUMOR_DP_MIN - If VCF INFO tag for sequencing depth (tumor) is specified and found, set minimum required depth for inclusion in report (default: 0)
  --tumor_af_min TUMOR_AF_MIN - If VCF INFO tag for variant allelic fraction (tumor) is specified and found, set minimum required AF for inclusion in report (default: 0)
  --control_dp_min CONTROL_DP_MIN - If VCF INFO tag for sequencing depth (control) is specified and found, set minimum required depth for inclusion in report (default: 0)
  --control_af_max CONTROL_AF_MAX - If VCF INFO tag for variant allelic fraction (control) is specified and found, set maximum tolerated AF for inclusion in report (default: 1)

Other options:
  --input_cna INPUT_CNA - Somatic copy number alteration segments (tab-separated values)
  --logr_gain LOGR_GAIN - Log ratio-threshold (minimum) for segments containing copy number gains/amplifications (default: 0.8)
  --logr_homdel LOGR_HOMDEL - Log ratio-threshold (maximum) for segments containing homozygous deletions (default: -0.8)
  --cna_overlap_pct CNA_OVERLAP_PCT - Mean percent overlap between copy number segment and gene transcripts for reporting of gains/losses in tumor suppressor genes/oncogenes, (default: 50)
  --tumor_site TSITE  - Optional integer code to specify primary tumor type/site of query sample, choose any of the following identifiers:
                        0 = Any, 1 = Adrenal Gland, 2 = Ampulla of Vater, 3 = Biliary Tract, 4 = Bladder/Urinary Tract, 5 = Bone, 6 = Breast, 7 = Cervix, 8 = CNS/Brain, 9 = Colon/Rectum, 10 = Esophagus/Stomach,
                        11 = Eye, 12 = Head and Neck, 13 = Kidney, 14 = Liver, 15 = Lung, 16 = Lymphoid, 17 = Myeloid, 18 = Ovary/Fallopian Tube, 19 = Pancreas, 20 = Peripheral Nervous System,
                        21 = Peritoneum, 22 = Pleura, 23 = Prostate, 24 = Skin, 25 = Soft Tissue, 26 = Testis, 27 = Thymus, 28 = Thyroid, 29 = Uterus, 30 = Vulva/Vagina,
                        (default: 0 - any tumor type)
  --tumor_purity TUMOR_PURITY - Estimated tumor purity (between 0 and 1, (default: None)
  --tumor_ploidy TUMOR_PLOIDY - Estimated tumor ploidy (default: None)
  --cpsr_report CPSR_REPORT - CPSR report file (Gzipped JSON - file ending with 'cpsr.&lt;genome_assembly&gt;.json.gz' -  germline report of patient's blood/control sample
  --vcf2maf             Generate a MAF file for input VCF using https://github.com/mskcc/vcf2maf (default: False)
  --show_noncoding      List non-coding (i.e. non protein-altering) variants in report, default: False
  --assay {WES,WGS,TARGETED} - Type of DNA sequencing assay performed for input data (VCF) default: WES
  --include_trials      (Beta) Include relevant ongoing or future clinical trials, focusing on studies with molecularly targeted interventions
  --preserved_info_tags PRESERVED_INFO_TAGS Comma-separated string of VCF INFO tags from query VCF that should be kept in PCGR output TSV file
  --report_theme {default,cerulean,journal,flatly,readable,spacelab,united,cosmo,lumen,paper,sandstone,simplex,yeti} - Visual report theme (rmarkdown)
  --report_nonfloating_toc - Do not float the table of contents (TOC) in output report (rmarkdown), default: False
  --force_overwrite     By default, the script will fail with an error if any output file already exists. You can force the overwrite of existing result files by using this flag, default: False
  --version             show program's version number and exit
  --basic               Run functional variant annotation on VCF through VEP/vcfanno, omit other analyses (i.e. Tier assignment/MSI/TMB/Signatures etc. and report generation (STEP 4), default: False
  --no_vcf_validate     Skip validation of input VCF with Ensembl's vcf-validator, default: False
  --docker_uid DOCKER_USER_ID - Docker user ID. default is the host system user ID. If you are experiencing permission errors, try setting this up to root (`--docker-uid root`)
  --no_docker           Run the PCGR workflow in a non-Docker mode (see install_no_docker/ folder for instructions)
  --debug               Print full Docker commands to log, default: False</code></pre>
</div>
<div class="section level2">
<h2 id="example-run">Example run<a class="anchor" aria-label="anchor" href="#example-run"></a>
</h2>
<p>The <em>examples</em> folder contains input VCF files from two tumor
samples sequenced within TCGA (<strong>GRCh37</strong> only). A report
for a colorectal tumor case can be generated by running the following
command in your terminal window:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> <span class="er">(</span><span class="ex">base</span><span class="kw">)</span> <span class="ex">conda</span> activate pcgr</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> <span class="er">(</span><span class="ex">pcgr</span><span class="kw">)</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">pcgr</span> <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--pcgr_dir</span> /Users/you/dir2/data <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--output_dir</span> /Users/you/dir3/pcgr_outputs <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--sample_id</span> tumor_sample.COAD <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--tumor_dp_tag</span> TDP <span class="dt">\</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--tumor_af_tag</span> TVAF <span class="dt">\</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--call_conf_tag</span> TAL <span class="dt">\</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">--genome_assembly</span> grch37 <span class="dt">\</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">--input_vcf</span> /Users/you/pcgr/examples/tumor_sample.COAD.vcf.gz <span class="dt">\</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">--tumor_site</span> 9 <span class="dt">\</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">--input_cna</span> /Users/you/pcgr/examples/tumor_sample.COAD.cna.tsv <span class="dt">\</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">--tumor_purity</span> 0.9 <span class="dt">\</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">--tumor_ploidy</span> 2.0 <span class="dt">\</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">--include_trials</span> <span class="dt">\</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">--assay</span> WES <span class="dt">\</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">--estimate_signatures</span> <span class="dt">\</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">--estimate_msi_status</span> <span class="dt">\</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">--estimate_tmb</span> <span class="dt">\</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">--no_docker</span> <span class="dt">\</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">--force_overwrite</span></span></code></pre></div>
<p>This command will run the Conda-based PCGR workflow and produce the
following output files in the <em>examples</em> folder:</p>
<ol style="list-style-type: decimal">
<li>
<strong>tumor_sample.COAD.pcgr_acmg.grch37.html</strong> - An
interactive HTML report for clinical interpretation (rmarkdown)</li>
<li>
<strong>tumor_sample.COAD.pcgr_acmg.grch37.flexdb.html</strong> - An
interactive HTML report for clinical interpretation (flexdashboard)</li>
<li>
<strong>tumor_sample.COAD.pcgr_acmg.grch37.vcf.gz (.tbi)</strong> -
Bgzipped VCF file with rich set of annotations for precision
oncology</li>
<li>
<strong>tumor_sample.COAD.pcgr_acmg.grch37.pass.vcf.gz
(.tbi)</strong> - Bgzipped VCF file with rich set of annotations for
precision oncology (PASS variants only)</li>
<li>
<strong>tumor_sample.COAD.pcgr_acmg.grch37.pass.tsv.gz</strong> -
Compressed vcf2tsv-converted file with rich set of annotations for
precision oncology</li>
<li>
<strong>tumor_sample.COAD.pcgr_acmg.grch37.snvs_indels.tiers.tsv</strong>
- Tab-separated values file with variants organized according to tiers
of functional relevance</li>
<li>
<strong>tumor_sample.COAD.pcgr_acmg.grch37.mutational_signatures.tsv</strong>
- Tab-separated values file with information on contribution of
mutational signatures</li>
<li>
<strong>tumor_sample.COAD.pcgr_acmg.grch37.json.gz</strong> -
Compressed JSON dump of HTML report content</li>
<li>
<strong>tumor_sample.COAD.pcgr_acmg.grch37.cna_segments.tsv.gz</strong>
- Compressed tab-separated values file with annotations of gene
transcripts that overlap with somatic copy number aberrations</li>
<li>
<strong>tumor_sample.COAD.cpsr_config.rds</strong> - PCGR
configuration object (RDS format), mostly for debugging purposes</li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>,
<a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
