<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pcgrr">
<title>Running • PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running">
<meta property="og:description" content="pcgrr">
<meta property="og:image" content="https://sigven.github.io/pcgr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-info" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/running.html">Running</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/input.html">Input files</a>
    <a class="dropdown-item" href="../articles/output.html">Output files</a>
    <a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a>
    <a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a>
    <a class="dropdown-item" href="../articles/developers.html">Developer notes</a>
    <a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/faq.html">FAQ</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Running</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/HEAD/vignettes/running.Rmd" class="external-link"><code>vignettes/running.Rmd</code></a></small>
      <div class="d-none name"><code>running.Rmd</code></div>
    </div>

    
    
<p>The PCGR software comes with a range of options to configure the
analyses performed and to customize the output report.</p>
<div class="section level2">
<h2 id="key-settings">Key settings<a class="anchor" aria-label="anchor" href="#key-settings"></a>
</h2>
<p>Below, we will outline key settings that are important with respect
to the sequencing assay that has been conducted.</p>
<div class="section level3">
<h3 id="whole-genome-whole-exome-or-targeted-sequencing-assay">Whole-genome, whole-exome or targeted sequencing assay<a class="anchor" aria-label="anchor" href="#whole-genome-whole-exome-or-targeted-sequencing-assay"></a>
</h3>
<p>By default, PCGR expects that the input VCF comes from whole-exome
sequencing. This can be adjusted with the <code>assay</code> option,
which takes three alternative values:</p>
<ul>
<li><code>--assay &lt;WGS|WES|TARGETED&gt;</code></li>
</ul>
<p>If the input VCF comes from a targeted sequencing assay/panel
(i.e. <code>--assay TARGETED</code>), the target size of the assay/panel
should be adjusted accordingly, which is critical for tumor mutational
burden estimation. Assay target size, reflecting the <em>protein-coding
portion of the assayed genomic region</em> in megabases, can be
configured through the following option:</p>
<ul>
<li><code>--effective_target_size_mb &lt;value&gt;</code></li>
</ul>
<p>Ideally, this should reflect the <em>callable</em> target size of the
assay, i.e. the size of the target regions for which coding variants is
sufficiently covered (i.e. with a certain sequencing depth) by the
sequencing assay. This can be estimated by the user based on the design
of the sequencing assay, or by using tools such as <a href="https://github.com/brentp/mosdepth">mosdepth</a>.</p>
</div>
<div class="section level3">
<h3 id="tumor-control-vs--tumor-only">Tumor-control vs. tumor-only<a class="anchor" aria-label="anchor" href="#tumor-control-vs--tumor-only"></a>
</h3>
<p>By default, PCGR expects that the input VCF contains somatic variants
identified from a tumor-control sequencing setup. This implies that the
VCF contains information with respect to variant allelic depth/support
both for the tumor sample and the corresponding control sample.</p>
<p>If the input VCF comes from a <strong>tumor-only</strong> assay, turn
on the <code>--tumor_only</code> option. In this mode, PCGR conducts a
set of successive filtering steps on the raw input set of variants,
aiming to exclude the majority of germline variants from the tumor-only
input set. In addition to default filtering applied against variants
found in the gnomAD database (population-specific allele frequency
thresholds can be configured, see below), additional filtering
procedures can be explicitly set, i.e.:</p>
<ul>
<li><code>--exclude_dbsnp_nonsomatic</code></li>
<li><code>--exclude_likely_het_germline</code></li>
<li><code>--exclude_likely_hom_germline</code></li>
<li><code>--exclude_nonexonic</code></li>
</ul>
<p>Users should note that these filters may occasionally remove true
somatic variants, and should thus be used with caution.</p>
<p>If the user has provided a <a href="input.html#panel-of-normals-vcf">panel-of-normals VCF file</a> as
input(<code>--pon_vcf</code>), one may also opt to remove variants found
in that variant collection using a dedicated option:</p>
<ul>
<li><code>--exclude_pon</code></li>
</ul>
<hr>
<p><strong>IMPORTANT NOTE 1</strong>: A number of the analyses available
in PCGR are not particularly well-suited for tumor-only input,
e.g. mutational signature analysis and MSI status prediction.
<strong>IMPORTANT NOTE 2</strong>: If you run PCGR on tumor-only WGS
assays, we strongly recommend that you pre-filter your VCF against the
exome, since we have frequently encountered that the large unfiltered
variant set from such assays will cause a crash in PCGR. See also <a href="https://github.com/sigven/pcgr/issues/178" class="external-link">this issue</a>.</p>
</div>
<div class="section level3">
<h3 id="sample-properties">Sample properties<a class="anchor" aria-label="anchor" href="#sample-properties"></a>
</h3>
<p>A few selected properties of the tumor sample that should be
determined prior to PCGR analysis can be provided, currently only used
for display/reporting:</p>
<ul>
<li>
<code>--tumor_ploidy &lt;value&gt;</code> - ploidy estimate</li>
<li>
<code>--tumor_purity &lt;value&gt;</code> - purity estimate</li>
</ul>
<div class="section level4">
<h4 id="tumor-site">Tumor site<a class="anchor" aria-label="anchor" href="#tumor-site"></a>
</h4>
<p>The primary tissue/site of the tumor sample that is analyzed can and
should be be provided, through the following option:</p>
<ul>
<li><code>--tumor_site &lt;value&gt;</code></li>
</ul>
<p>This option takes a value between 0 and 30, reflecting the main
primary sites/tissues for human cancers. Note that the default value is
<strong>0</strong> (implies <strong>Any</strong> tissue/site), which
essentially prohibits the presence of <a href="variant_classification.html">tier 1 variants</a> occurring in the
report. If you want to check the nature of cancer subtypes attributed to
each of the primary tissue/sites, take a look at the <a href="primary_tumor_sites.html">table with phenotype terms and
associated primary sites</a>.</p>
</div>
</div>
<div class="section level3">
<h3 id="allelic-support">Allelic support<a class="anchor" aria-label="anchor" href="#allelic-support"></a>
</h3>
<p>Proper designation of INFO tags in the input VCF that denote variant
depth/allelic fraction is critical to make the report as comprehensive
as possible. See the <a href="input.html">input section</a> on how to
accomplish this.</p>
<p>Several options are used to inform PCGR about tags in the VCF INFO
field that encodes this information:</p>
<ul>
<li><code>--tumor_dp_tag &lt;value&gt;</code></li>
<li><code>--tumor_af_tag &lt;value&gt;</code></li>
<li><code>--control_dp_tag &lt;value&gt;</code></li>
<li><code>--control_af_tag &lt;value&gt;</code></li>
</ul>
<p>If these tags are set correctly, one may set thresholds (sequencing
depth and/or allelic fraction) on the variants to be included in the
report, i.e. through:</p>
<ul>
<li><code>--tumor_dp_min &lt;value&gt;</code></li>
<li><code>--tumor_af_min &lt;value&gt;</code></li>
<li><code>--control_dp_min &lt;value&gt;</code></li>
<li><code>--control_af_max &lt;value&gt;</code></li>
</ul>
<div class="section level4">
<h4 id="tumor-mutational-burden-tmb">Tumor mutational burden (TMB)<a class="anchor" aria-label="anchor" href="#tumor-mutational-burden-tmb"></a>
</h4>
<p>If tags for allelic support is provided in the VCF and configured by
the user, users can configure the TMB calculation by setting minimum
requirements for sequencing coverage and allelic fraction, i.e.:</p>
<ul>
<li><code>--tmb_dp_min &lt;value&gt;</code></li>
<li><code>--tmb_af_min &lt;value&gt;</code></li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="large-input-sets-vcf">Large input sets (VCF)<a class="anchor" aria-label="anchor" href="#large-input-sets-vcf"></a>
</h3>
<p>For input VCF files with &gt; 500,000 variants, note that these will
be subject to filtering prior to the final step in PCGR (step 4:
reporting). For such scenarios, raw input sets will be filtered
according to consequence type, in the sense that
intergenic/intronic/upstream/downstream gene variants will be excluded
prior to reporting. This is a necessary step to avoid memory issues etc.
during processing with the PCGR R package.</p>
<p>If you have a large input VCF, and have sufficient memory capacity on
your compute platform, we also recommend to increase the <a href="https://www.ensembl.org/info/docs/tools/vep/script/vep_options.html#cacheopt" class="external-link">VEP
buffer size</a> (option <code>--vep_buffer_size</code>), as this will
speed up the VEP processing significantly.</p>
</div>
</div>
<div class="section level2">
<h2 id="all-options">All options<a class="anchor" aria-label="anchor" href="#all-options"></a>
</h2>
<p>A tumor sample report is generated by running the
<strong>pcgr</strong> command, which takes the following arguments and
options:</p>
<pre class="text"><code>usage: 
    pcgr -h [options] 
    --input_vcf &lt;INPUT_VCF&gt;
    --vep_dir &lt;VEP_DIR&gt;
    --refdata_dir &lt;REFDATA_DIR&gt;
    --output_dir &lt;OUTPUT_DIR&gt;
    --genome_assembly &lt;GENOME_ASSEMBLY&gt;
    --sample_id &lt;SAMPLE_ID&gt; 

Personal Cancer Genome Reporter (PCGR) workflow for clinical translation of tumor omics data (SNVs/InDels, CNA, RNA expression)

Required arguments:
  --input_vcf INPUT_VCF
                        VCF input file with somatic variants in tumor sample, SNVs/InDels
  --vep_dir VEP_DIR     Directory of VEP cache, e.g.  $HOME/.vep
  --refdata_dir REFDATA_DIR
                        Directory where PCGR reference data bundle was downloaded and unpacked
  --output_dir OUTPUT_DIR
                        Output directory
  --genome_assembly {grch37,grch38}
                        Human genome assembly build: grch37 or grch38
  --sample_id SAMPLE_ID
                        Tumor sample/cancer genome identifier - prefix for output files

Sequencing assay options:
  --assay {WGS,WES,TARGETED}
                        Type of DNA sequencing assay performed for input data (VCF), default: WES
  --effective_target_size_mb EFFECTIVE_TARGET_SIZE_MB
                        Effective target size in Mb (potentially limited by read depth) of sequencing assay (for TMB analysis) (default: 34 (WES/WGS))
  --tumor_only          Input VCF comes from tumor-only sequencing, calls will be filtered for variants of germline origin, (default: False)


Tumor sample options:
  --tumor_site TSITE    Optional integer code to specify primary tumor type/site of query sample,
                        choose any of the following identifiers:
                        0 = Any
                        1 = Adrenal Gland
                        2 = Ampulla of Vater
                        3 = Biliary Tract
                        4 = Bladder/Urinary Tract
                        5 = Bone
                        6 = Breast
                        7 = Cervix
                        8 = CNS/Brain
                        9 = Colon/Rectum
                        10 = Esophagus/Stomach
                        11 = Eye
                        12 = Head and Neck
                        13 = Kidney
                        14 = Liver
                        15 = Lung
                        16 = Lymphoid
                        17 = Myeloid
                        18 = Ovary/Fallopian Tube
                        19 = Pancreas
                        20 = Peripheral Nervous System
                        21 = Peritoneum
                        22 = Pleura
                        23 = Prostate
                        24 = Skin
                        25 = Soft Tissue
                        26 = Testis
                        27 = Thymus
                        28 = Thyroid
                        29 = Uterus
                        30 = Vulva/Vagina
                        (default: 0 - any tumor type)
  --tumor_purity TUMOR_PURITY
                        Estimated tumor purity (between 0 and 1) (default: None)
  --tumor_ploidy TUMOR_PLOIDY
                        Estimated tumor ploidy (default: None)

Allelic support options:
  --tumor_dp_tag TUMOR_DP_TAG
                        Specify VCF INFO tag for sequencing depth (tumor, must be Type=Integer, default: _NA_
  --tumor_af_tag TUMOR_AF_TAG
                        Specify VCF INFO tag for variant allelic fraction (tumor,  must be Type=Float, default: _NA_
  --control_dp_tag CONTROL_DP_TAG
                        Specify VCF INFO tag for sequencing depth (control, must be Type=Integer, default: _NA_
  --control_af_tag CONTROL_AF_TAG
                        Specify VCF INFO tag for variant allelic fraction (control, must be Type=Float, default: _NA_
  --call_conf_tag CALL_CONF_TAG
                        Specify VCF INFO tag for somatic variant call confidence (must be categorical, e.g. Type=String, default: _NA_
  --tumor_dp_min TUMOR_DP_MIN
                        If VCF INFO tag for sequencing depth (tumor) is specified and found, set minimum required depth for inclusion in report (default: 0)
  --tumor_af_min TUMOR_AF_MIN
                        If VCF INFO tag for variant allelic fraction (tumor) is specified and found, set minimum required AF for inclusion in report (default: 0)
  --control_dp_min CONTROL_DP_MIN
                        If VCF INFO tag for sequencing depth (control) is specified and found, set minimum required depth for inclusion in report (default: 0)
  --control_af_max CONTROL_AF_MAX
                        If VCF INFO tag for variant allelic fraction (control) is specified and found, set maximum tolerated AF for inclusion in report (default: 1)

Tumor-only filtering options:
  --pon_vcf PON_VCF     VCF file with germline calls from Panel of Normals (PON) - i.e. blacklisted variants, (default: None)
  --maf_gnomad_nfe MAF_GNOMAD_NFE
                        Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF &gt; pct, (gnomAD - European (non-Finnish), default: 0.002)
  --maf_gnomad_asj MAF_GNOMAD_ASJ
                        Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF &gt; pct, (gnomAD - Ashkenazi Jewish, default: 0.002)
  --maf_gnomad_fin MAF_GNOMAD_FIN
                        Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF &gt; pct, (gnomAD - European (Finnish), default: 0.002)
  --maf_gnomad_oth MAF_GNOMAD_OTH
                        Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF &gt; pct, (gnomAD - Other, default: 0.002)
  --maf_gnomad_amr MAF_GNOMAD_AMR
                        Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF &gt; pct, (gnomAD - Latino/Admixed American, default: 0.002)
  --maf_gnomad_afr MAF_GNOMAD_AFR
                        Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF &gt; pct, (gnomAD - African/African-American, default: 0.002)
  --maf_gnomad_eas MAF_GNOMAD_EAS
                        Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF &gt; pct, (gnomAD - East Asian, default: 0.002)
  --maf_gnomad_sas MAF_GNOMAD_SAS
                        Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF &gt; pct, (gnomAD - South Asian, default: 0.002)
  --maf_gnomad_global MAF_GNOMAD_GLOBAL
                        Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF &gt; pct, (gnomAD - global population, default: 0.002)
  --exclude_pon         Exclude variants occurring in PoN (Panel of Normals, if provided as VCF (--pon_vcf), default: False)
  --exclude_likely_hom_germline
                        Exclude likely homozygous germline variants (allelic fraction of 1.0 for alternate allele in tumor - very unlikely somatic event), default: False)
  --exclude_likely_het_germline
                        Exclude likely heterozygous germline variants (0.4-0.6 allelic fraction, AND presence in dbSNP + gnomAD, AND not existing as somatic record in COSMIC OR TCGA, default: False)
  --exclude_clinvar_germline
                        Exclude variants found in ClinVar (germline variant origin), defult: False)
  --exclude_dbsnp_nonsomatic
                        Exclude variants found in dbSNP (except for those present in ClinVar (somatic origin) OR TCGA OR COSMIC), defult: False)
  --exclude_nonexonic   Exclude non-exonic variants, default: False)

VEP options:
  --vep_n_forks VEP_N_FORKS
                        Number of forks (VEP option '--fork'), default: 4
  --vep_buffer_size VEP_BUFFER_SIZE
                        Variant buffer size (variants read into memory simultaneously, VEP option '--buffer_size')
                        - set lower to reduce memory usage, default: 500
  --vep_pick_order VEP_PICK_ORDER
                        Comma-separated string of ordered transcript/variant properties for selection of primary variant consequence
                        (option '--pick_order' in VEP), default: mane_select,mane_plus_clinical,canonical,appris,tsl,biotype,ccds,rank,length
  --vep_no_intergenic   Skip intergenic variants during variant annotation (VEP option '--no_intergenic' in VEP), default: False
  --vep_regulatory      Add VEP regulatory annotations (VEP option '--regulatory') or non-coding interpretation, default: False
  --vep_gencode_basic   Consider basic GENCODE transcript set only with Variant Effect Predictor (VEP) (VEP option '--gencode_basic').

Tumor mutational burden (TMB) and MSI options:
  --estimate_tmb        Estimate tumor mutational burden from the total number of somatic mutations and target region size, default: False
  --tmb_display {coding_and_silent,coding_non_silent,missense_only}
                        Type of TMB measure to show in report, default: coding_and_silent
  --tmb_dp_min TMB_DP_MIN
                        If VCF INFO tag for sequencing depth (tumor) is specified and found, set minimum required sequencing depth for TMB calculation: default: 0
  --tmb_af_min TMB_AF_MIN
                        If VCF INFO tag for allelic fraction (tumor) is specified and found, set minimum required allelic fraction for TMB calculation: default: 0
  --estimate_msi        Predict microsatellite instability status from patterns of somatic mutations/indels, default: False

Mutational signature options:
  --estimate_signatures
                        Estimate relative contributions of reference mutational signatures in query sample (re-fitting), default: False
  --min_mutations_signatures MIN_MUTATIONS_SIGNATURES
                        Minimum number of SNVs required for re-fitting of mutational signatures (SBS) (default: 200, minimum n = 100)
  --all_reference_signatures
                        Use _all_ reference mutational signatures (SBS) during signature re-fitting rather than only those already attributed to the tumor type (default: False)
  --include_artefact_signatures
                        Include sequencing artefacts in the collection of reference signatures (default: False
  --prevalence_reference_signatures PREVALENCE_REFERENCE_SIGNATURES
                        Minimum tumor-type prevalence (in percent) of reference signatures to be included in refitting procedure (default: 0.1)

Somatic copy number alteration (CNA) data options:
  --input_cna INPUT_CNA
                        Somatic copy number alteration segments (tab-separated values)
  --n_copy_gain N_COPY_GAIN
                        Minimum number of total copy number for segments considered as gains/amplifications (default: 6)
  --cna_overlap_pct CNA_OVERLAP_PCT
                        Mean percent overlap between copy number segment and gene transcripts for reporting of gains/losses in tumor suppressor genes/oncogenes, (default: 50)

Bulk RNA-seq and RNA fusion data options:
  --input_rna_expression INPUT_RNA_EXP
                        File with bulk RNA expression counts (TPM) of transcripts in tumor (tab-separated values)
  --expression_sim      Compare expression profile of tumor sample to known expression profiles (default: False)
  --expression_sim_db EXPRESSION_SIM_DB
                        Comma-separated string of databases for used in RNA expression similarity analysis, default: tcga,depmap,treehouse

Other options:
  --vcf2maf             Generate a MAF file for input VCF using https://github.com/mskcc/vcf2maf (default: False)
  --vcfanno_n_proc VCFANNO_N_PROC
                        Number of vcfanno processes (option '-p' in vcfanno), default: 4
  --ignore_noncoding    Ignore non-coding (i.e. non protein-altering) variants in report, default: False
  --retained_info_tags RETAINED_INFO_TAGS
                        Comma-separated string of VCF INFO tags from query VCF that should be kept in PCGR output TSV file
  --force_overwrite     By default, the script will fail with an error if any output file already exists. You can force the overwrite of existing result files by using this flag, default: False
  --version             show program's version number and exit
  --no_reporting        Run functional variant annotation on VCF through VEP/vcfanno, omit other analyses (i.e. Tier assignment/MSI/TMB/Signatures etc. and report generation (STEP 4), default: False
  --debug               Print full commands to log
  --pcgrr_conda PCGRR_CONDA
                        pcgrr conda env name (default: pcgrr)</code></pre>
</div>
<div class="section level2">
<h2 id="example-run">Example run<a class="anchor" aria-label="anchor" href="#example-run"></a>
</h2>
<p>The <em>examples</em> folder contains input VCF files from two tumor
samples sequenced within TCGA. A molecular interpretation report for a
colorectal tumor sample can be generated by running the following
command in your terminal window (this assumes you have installed PCGR
through Conda, and that you have downloaded the necessary reference data
files, VEP cache, see <a href="installation.html">Installation</a>):</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">$</span> <span class="er">(</span><span class="ex">base</span><span class="kw">)</span> <span class="ex">conda</span> activate pcgr</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">$</span> <span class="er">(</span><span class="ex">pcgr</span><span class="kw">)</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="ex">pcgr</span> <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>    <span class="at">--refdata_dir</span> /Users/you/dir2/data <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>    <span class="at">--vep_dir</span> /Users/you/dir3/vep/.vep <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>    <span class="at">--output_dir</span> /Users/you/dir4/pcgr_outputs <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>    <span class="at">--sample_id</span> T001-COAD <span class="dt">\</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>    <span class="at">--genome_assembly</span> grch37 <span class="dt">\</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>    <span class="at">--input_vcf</span> /Users/you/pcgr/examples/T001-COAD.grch37.vcf.gz <span class="dt">\</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>    <span class="at">--tumor_dp_tag</span> TDP <span class="dt">\</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>    <span class="at">--tumor_af_tag</span> TVAF <span class="dt">\</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>    <span class="at">--tumor_site</span> 9 <span class="dt">\</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>    <span class="at">--input_cna</span> /Users/you/pcgr/examples/T001-COAD.cna.tsv <span class="dt">\</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>    <span class="at">--tumor_purity</span> 0.9 <span class="dt">\</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>    <span class="at">--tumor_ploidy</span> 2.0 <span class="dt">\</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a>    <span class="at">--assay</span> WES <span class="dt">\</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a>    <span class="at">--vcf2maf</span> <span class="dt">\</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>    <span class="at">--estimate_signatures</span> <span class="dt">\</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>    <span class="at">--estimate_msi</span> <span class="dt">\</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a>    <span class="at">--estimate_tmb</span> <span class="dt">\</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>    <span class="at">--force_overwrite</span></span></code></pre></div>
<p>This command will run the Conda-based PCGR workflow and produce the
following files in the <em>/Users/you/dir4/pcgr_outputs</em> folder:</p>
<table class="table">
<colgroup>
<col width="13%">
<col width="27%">
<col width="59%">
</colgroup>
<thead><tr class="header">
<th>N</th>
<th>File</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.html</strong></td>
<td>An interactive HTML report for clinical interpretation
(quarto-based)</td>
</tr>
<tr class="even">
<td>2</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.xlsx</strong></td>
<td>An excel workbook with multiple sheets of annotations (Assay &amp;
sample info/SNVs &amp; InDels/CNAs/biomarkers/TMB/MSI), suitable for
aggregation analysis across multiple samples</td>
</tr>
<tr class="odd">
<td>3</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.vcf.gz (.tbi)</strong></td>
<td>Bgzipped VCF file with rich set of variant annotations to support
interpretation</td>
</tr>
<tr class="even">
<td>4</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.pass.vcf.gz
(.tbi)</strong></td>
<td>Bgzipped VCF file with rich set of variant annotations to support
interpretation (PASS variants only)</td>
</tr>
<tr class="odd">
<td>5</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.pass.tsv.gz</strong></td>
<td>Compressed vcf2tsv-converted file with rich set of variant
annotations to support interpretation</td>
</tr>
<tr class="even">
<td>6</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.conf.yaml</strong></td>
<td>PCGR configuration data file (YAML), as generated by pre-reporting
annotation workflow</td>
</tr>
<tr class="odd">
<td>7</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.cna_segments.tsv.gz</strong></td>
<td>Tab-separated values file with raw copy number segments, per
affected transcript</td>
</tr>
<tr class="even">
<td>8</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.cna_gene_ann.tsv.gz</strong></td>
<td>Tab-separated values file with genes subject to copy number
alteration, including actionability assessment</td>
</tr>
<tr class="odd">
<td>9</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.tmb.tsv</strong></td>
<td>Tab-separated values file with information on tumor mutational
burden (TMB) estimates</td>
</tr>
<tr class="even">
<td>10</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.maf</strong></td>
<td>Annotated SNVs/InDels, converted to the Mutation Annotation Format
(MAF)</td>
</tr>
<tr class="odd">
<td>11</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.msigs.tsv.gz</strong></td>
<td>Tab-separated values file with information on the contribution of
mutational signatures in the tumor sample</td>
</tr>
<tr class="even">
<td>12</td>
<td><strong>&lt;sample_id&gt;.pcgr.grch37.snv_indel_ann.tsv.gz</strong></td>
<td>Tab-separated values file with key SNV/InDel variant annotations,
including oncogenicity and clinical actionability assessment</td>
</tr>
</tbody>
</table>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
