<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pcgrr">
<title>Variant classification • PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Variant classification">
<meta property="og:description" content="pcgrr">
<meta property="og:image" content="https://sigven.github.io/pcgr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-info" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/running.html">Running</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/input.html">Input files</a>
    <a class="dropdown-item" href="../articles/output.html">Output files</a>
    <a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a>
    <a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a>
    <a class="dropdown-item" href="../articles/developers.html">Developer notes</a>
    <a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/faq.html">FAQ</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Variant classification</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/HEAD/../vignettes/variant_classification.Rmd" class="external-link"><code>../vignettes/variant_classification.Rmd</code></a></small>
      <div class="d-none name"><code>variant_classification.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="oncogenicity">Oncogenicity<a class="anchor" aria-label="anchor" href="#oncogenicity"></a>
</h2>
<p>PCGR evaluates somatic aberrations (SNV/InDels) for oncogenic
potential through an implementation of standard operating procedures
proposed by <a href="https://pubmed.ncbi.nlm.nih.gov/35101336/" class="external-link">VICC/CGC/ClinGen</a>.
Here, various properties of the variants and genes affected are assigned
criteria-specific scores, both negative and positive, pending on whether
the properties support an oncogenic or benign variant nature.
Criteria-specific scores are in turn aggregated towards an overall
oncogenicity score per variant.</p>
<p>Note that all properties/criteria provided in the SOP’s are
<em>not</em> readily implemented in PCGR, specifically the ones
requiring manual curation or expert review (i.e. experimental oncogenic
variant evidence, requring support from <em>in vitro</em> or <em>in
vivo</em> functional studies (criteria <em>OM1/OS1</em>)). This implies
that some variants interrogated by PCGR may not be classified as
oncogenic, even though they could be classified as such with more
functional evidence available. Considering the nature of our current
implementation, we have thus also adopted slightly different score
thresholds for variant classifications to those proposed originally by
<a href="https://pubmed.ncbi.nlm.nih.gov/35101336/" class="external-link">Horak et al.,
2022</a>. We are working to further improve the oncogenicity
classification in PCGR, and welcome feedback on this matter.</p>
<p>Note also that for somatic copy number aberrations, we showcase
potential oncogenic events as <strong>proto-oncogenes subject to
amplifications</strong> (where level of amplification is configurable by
the user), as well as<strong>tumor suppressor genes subject to
homozygous deletions</strong>.</p>
<p>The following criteria are currently used for oncogenicity
classification in PCGR:</p>
<ul>
<li>
<em>CLINGEN_VICC_SBVS1</em> - Very high MAF (&gt; 0.05 in gnomAD -
any five major continental pops)</li>
<li>
<em>CLINGEN_VICC_SBS1</em> - High MAF (&gt; 0.01 in gnomAD - any
five major continental pops)</li>
<li>
<em>CLINGEN_VICC_SBP1</em> - Insilico support a benign effect on the
gene or gene product (multiple lines of evidence (&gt;= 7 algorithms)
from dbNSFP)</li>
<li>
<em>CLINGEN_VICC_SBP2</em> - Silent and intronic changes outside of
the consensus splice site (VEP consequence)</li>
<li>
<em>CLINGEN_VICC_OVS1</em> - Null variant - in bona fide tumor
suppressor gene (predicted as LoF in tumor suppressors from
CGC/NCG/CancerMine)</li>
<li>
<em>CLINGEN_VICC_OS3</em> - Located in a mutation hotspot, &gt;= 50
samples with a variant at amino acid position, &gt;= 10 samples with
same amino acid change (cancerhotspots.org)</li>
<li>
<em>CLINGEN_VICC_OM1</em> - Located in a presumably critical site of
functional domain - here, this is implemented through indirect evidence
from overlap with known oncogenic and predictive (actionable)
biomarkers</li>
<li>
<em>CLINGEN_VICC_OM2</em> - Protein length changes from in-frame
dels/ins in known oncogene/tumor suppressor genes or stop-loss variants
in a tumor suppressor gene (tumor suppressors/oncogenes from
CGC/NCG/CancerMine)</li>
<li>
<em>CLINGEN_VICC_OM3</em> - Located in a mutation hotspot, &lt; 50
samples with a variant at amino acid position, &gt;= 10 samples with
same amino acid change (cancerhotspots.org)</li>
<li>
<em>CLINGEN_VICC_OP1</em> - Insilico support a damaging effect on
the gene or gene product (multiple lines of evidence (&gt;= 7
algorithms) from dbNSFP)</li>
<li>
<em>CLINGEN_VICC_OP3</em> - Located in a mutation hotspot, &lt; 10
samples with the same amino acid change (cancerhotspots.org)</li>
<li>
<em>CLINGEN_VICC_OP4</em> - Absent from controls (gnomAD) / very low
MAF ( &lt; 0.0001 in all five major subpopulations)</li>
</ul>
</div>
<div class="section level2">
<h2 id="actionability">Actionability<a class="anchor" aria-label="anchor" href="#actionability"></a>
</h2>
<p>PCGR prioritizes and evaluates variants according to clinical
actionability. Currently, PCGR implements its tier classification
framework along the proposed ACMG/AMP guidelines, as detailed below.</p>
<p>This tier model attempts to adopt concensus recommendations by ACMG,
as outlined in <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5707196/" class="external-link">Li et al.,
2017</a>:</p>
<ul>
<li>
<em>Tier 1: Variants of strong clinical significance</em> -
constitutes variants linked to predictive, prognostic, or diagnostic
evidence items in the <a href="http://civic.genome.wustl.edu" class="external-link">CIViC
database</a> or the <a href="https://www.cancergenomeinterpreter.org/biomarkers" class="external-link">Cancer
Biomarkers Database</a> that are
<ul>
<li>A: found within the same tumor type/class as specified by the user,
AND</li>
<li>B: of strong clinical evidence (i.e. part of guidelines, validated
or discovered in late clinical trials)</li>
</ul>
</li>
<li>
<em>Tier 2: Variants of potential clinical significance</em> -
constitutes other variants linked to predictive, prognostic, or
diagnostic evidence items in the <a href="http://civic.genome.wustl.edu" class="external-link">CIViC database</a> or the <a href="https://www.cancergenomeinterpreter.org/biomarkers" class="external-link">Cancer
Biomarkers Database</a> that are either
<ul>
<li>A: of strong clinical evidence in other tumor types/classes than the
one specified by the user, OR</li>
<li>B: of weak clinical evidence (early trials, case reports etc.) in
the same tumor type/class as specified by the user</li>
</ul>
</li>
<li>
<em>Tier 3: Variants of uncertain clinical significance</em> -
includes other coding variants found in proto-oncogenes or tumor
suppressor genes (SNVs and indels only)</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
