<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Variant classification • PCGR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Variant classification">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-7HRLE8GQRE"></script><script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){dataLayer.push(arguments);}
   gtag('js', new Date());

   gtag('config', 'G-7HRLE8GQRE');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PCGR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/running.html">Running</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/input.html">Input files</a></li>
    <li><a class="dropdown-item" href="../articles/output.html">Output files</a></li>
    <li><a class="dropdown-item" href="../articles/variant_classification.html">Variant classification</a></li>
    <li><a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="../articles/developers.html">Developer notes</a></li>
    <li><a class="dropdown-item" href="../articles/primary_tumor_sites.html">Primary tumor sites</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/pcgr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Variant classification</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/pcgr/blob/main/vignettes/variant_classification.Rmd" class="external-link"><code>vignettes/variant_classification.Rmd</code></a></small>
      <div class="d-none name"><code>variant_classification.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="oncogenicity">Oncogenicity<a class="anchor" aria-label="anchor" href="#oncogenicity"></a>
</h2>
<p>PCGR evaluates somatic aberrations (SNV/InDels) for oncogenic
potential through an implementation of standard operating procedures
proposed by <a href="https://pubmed.ncbi.nlm.nih.gov/35101336/" class="external-link">ClinGen/CGC/VICC</a>.
Here, various properties of the variants and genes affected are assigned
criteria-specific scores, both negative and positive, pending on whether
the properties support an oncogenic or benign variant nature.
Criteria-specific scores are in turn aggregated towards an overall
oncogenicity score per variant.</p>
<p>Note that all properties/criteria provided in the SOP’s are
<em>not</em> readily implemented in PCGR, specifically the ones
requiring manual curation or expert review (i.e. experimental oncogenic
variant evidence, requiring support from <em>in vitro</em> or <em>in
vivo</em> functional studies (criteria <em>OS2</em>)). Also, the current
source for existing oncogenicity classifications (ClinVar), which is
needed for implementation of criteria <em>OS1</em> and <em>OM3</em>, is
considerably limited. Considering the current limitations, some
oncogenic variants are likely to be missed and classified with uncertain
significance (VUS) by PCGR. We highlight conventional ClinVar
classifications (i.e. with respect to pathogenicity) alongside the
current oncogenicity classifications, this may likely add value to the
interpretation for uncertain cases. Furthermore, taking into the account
the nature of the current implementation, we have adopted slightly
different score thresholds for variant classifications to those proposed
originally by <a href="https://pubmed.ncbi.nlm.nih.gov/35101336/" class="external-link">Horak
et al., 2022</a>. We are working to further improve the oncogenicity
classification in PCGR, and welcome feedback on this matter.</p>
<p>Note also that for somatic copy number aberrations, we showcase
potential oncogenic events as <strong>proto-oncogenes subject to
amplifications</strong> (where level of amplification is configurable by
the user), as well as <strong>tumor suppressor genes subject to
homozygous deletions</strong>.</p>
<p>The following criteria/codes are currently used for variant
oncogenicity classification in PCGR (key resources/tools used for
implementation indicated in parentheses):</p>
<ul>
<li>
<em>ONCG_OVS1</em> - Null variant - predicted as LoF - in bona fide
tumor suppressor gene (VEP;CGC;CancerMine)</li>
<li>
<em>ONCG_OS1</em> - Same amino acid change as previously established
oncogenic variant - regardless of nucleotide change (ClinVar)</li>
<li>
<em>ONCG_OS3</em> - Located in a mutation hotspot with &gt;= 50
samples with variant at AA position, &gt;= 10 samples with same AA
change (cancerhotspots.org)</li>
<li>
<em>ONCG_OM1</em> - Presumably critical site of functional domain
(CIViC)</li>
<li>
<em>ONCG_OM2</em> - Protein length changes from in-frame dels/ins in
known oncogene/tumor suppressor genes or stop-loss variants in a tumor
suppressor gene (VEP;CGC;CancerMine)</li>
<li>
<em>ONCG_OM3</em> - Missense variant at an amino acid residue where
a different missense variant determined to be oncogenic (using this
standard) has been documented (ClinVar)</li>
<li>
<em>ONCG_OM4</em> - Located in a mutation hotspot with &lt; 50
samples with variant at AA position, &gt;= 10 samples with same AA
change (cancerhotspots.org)</li>
<li>
<em>ONCG_OP1</em> - Multiple lines of computational evidence support
of a damaging variant effect on the gene or gene product (dbNSFP)</li>
<li>
<em>ONCG_OP3</em> - Located in a mutation hotspot with &lt; 10
samples with the same amino acid change (cancerhotspots.org)</li>
<li>
<em>ONCG_OP4</em> - Absent from controls (gnomAD) / very low MAF
(any five major gnomAD subpopulations) (gnomAD)</li>
<li>
<em>ONCG_SBVS1</em> - Very high MAF (any five major gnomAD
subpopulations) (gnomAD)</li>
<li>
<em>ONCG_SBS1</em> - High MAF (any five major gnomAD subpopulations)
(gnomAD)</li>
<li>
<em>ONCG_SBP1</em> - Multiple lines of computational evidence
support a benign variant effect on the gene or gene product
(dbNSFP)</li>
<li>
<em>ONCG_SBP2</em> - Silent and intronic changes outside of the
consensus splice site (VEP)</li>
</ul>
</div>
<div class="section level2">
<h2 id="actionability">Actionability<a class="anchor" aria-label="anchor" href="#actionability"></a>
</h2>
<p>PCGR prioritizes and evaluates variants according to clinical
actionability. Currently, PCGR implements its tier classification
framework along the proposed AMP/ASCO/CAP guidelines, as outlined in <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5707196/" class="external-link">Li et al.,
2017</a>:</p>
<ul>
<li>
<strong>Tier I: Variants of strong clinical significance</strong> -
constitutes aberrations linked to predictive, prognostic, or diagnostic
biomarkers in the <a href="https://civicdb.org" class="external-link">CIViC database</a> and
the <a href="https://www.cancergenomeinterpreter.org/biomarkers" class="external-link">Cancer
Biomarkers Database</a> that are
<ul>
<li>Found within the same tumor type/class as specified by the user,
<strong>AND</strong>
</li>
<li>Of strong clinical evidence (i.e. approved therapies, part of
guidelines, validated or discovered in late clinical trials (<a href="https://civic.readthedocs.io/en/latest/model/evidence/level.html" class="external-link">CIViC
evidence levels A/B</a>))</li>
</ul>
</li>
<li>
<strong>Tier II: Variants of potential clinical
significance</strong> - constitutes other aberrations linked to
predictive, prognostic, or diagnostic biomarkers in the <a href="https://civicdb.org" class="external-link">CIViC database</a> and the <a href="https://www.cancergenomeinterpreter.org/biomarkers" class="external-link">Cancer
Biomarkers Database</a> that are either
<ul>
<li>Of strong clinical evidence in other tumor types/classes than the
one specified by the user, <strong>OR</strong>
</li>
<li>Of weak clinical evidence (early trials, case reports etc. (<a href="https://civic.readthedocs.io/en/latest/model/evidence/level.html" class="external-link">CIViC
evidence levels C/D/E</a>))) in the same tumor type/class as specified
by the user</li>
</ul>
</li>
<li>
<strong>Tier III: Variants of uncertain clinical significance
(SNVs/InDels only)</strong> -
<ul>
<li>Other coding variants, not observed at significant allele
frequencies (gnomAD MAF &lt; 0.001), found in oncogenes or tumor
suppressor genes, yet <em>not</em> linked to any known predictive,
prognostic, or diagnostic biomarkers in the <a href="https://civicdb.org" class="external-link">CIViC database</a> and the <a href="https://www.cancergenomeinterpreter.org/biomarkers" class="external-link">Cancer
Biomarkers Database</a>
</li>
</ul>
</li>
</ul>
<p>In PCGR, we skip the classification of variants into the
AMP/ASCO/CAP-specified <em>Tier IV</em> (benign/likely benign variants),
but rather take a more cautious approach. Specifically, for SNVs/indels
that do not fall into tier I, II, or III, we classify them into <em>Tier
V: Other coding variants</em>, which includes protein-coding variants in
non-cancer related genes, as well as <em>Tier VI: Other non-coding
variants</em>, which includes synonymous variants, intronic variants,
and other variants in non-coding regions.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>, <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
